{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "881cf12c",
   "metadata": {},
   "source": [
    "# Merge multiple tables for the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41e1d949",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e6244add",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_data='/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/scRNA/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "144dbc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_tables='/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c237118",
   "metadata": {},
   "source": [
    "## Endocrine markers (postnatal, embryo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d13d40bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load table parts\n",
    "res={}\n",
    "res['postnatal_DE']=pd.read_excel(\n",
    "    path_data+'combined/celltypes/DEedgeR_summaries_ctParsed.xlsx',sheet_name=None)\n",
    "res['embryonic_DE']=pd.read_excel(\n",
    "    path_data+'combined/celltypes_embryo/DEedgeR_summaries_ctParsed.xlsx',sheet_name=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "12756c00",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'postnatal_DE': {'alpha':                      gene     logFC           padj gene_symbol\n",
       "  0      ENSMUSG00000052551  3.687634   2.860448e-54      Adarb2\n",
       "  1      ENSMUSG00000000394  3.684738  1.817751e-271         Gcg\n",
       "  2      ENSMUSG00000060969  3.610245  1.063880e-202        Irx1\n",
       "  3      ENSMUSG00000032523  3.560820  2.163253e-201       Hhatl\n",
       "  4      ENSMUSG00000027895  3.255302   2.120431e-91       Kcnc4\n",
       "  ...                   ...       ...            ...         ...\n",
       "  15978  ENSMUSG00000027187 -1.079029   0.000000e+00         Cat\n",
       "  15979  ENSMUSG00000027984 -1.136794   0.000000e+00        Hadh\n",
       "  15980  ENSMUSG00000053617 -1.184759  3.133599e-259    Sh3pxd2a\n",
       "  15981  ENSMUSG00000047407 -1.204728  6.294808e-276       Tgif1\n",
       "  15982  ENSMUSG00000020623 -1.272446  6.521274e-121      Map2k6\n",
       "  \n",
       "  [15983 rows x 4 columns],\n",
       "  'beta':                      gene     logFC           padj gene_symbol\n",
       "  0      ENSMUSG00000005268  5.224285  2.405878e-287        Prlr\n",
       "  1      ENSMUSG00000044988  5.211260  1.458425e-255        Ucn3\n",
       "  2      ENSMUSG00000099162  5.171442  3.856100e-134     Gm27168\n",
       "  3      ENSMUSG00000009246  4.958923  5.067859e-193       Trpm5\n",
       "  4      ENSMUSG00000031384  4.907128  6.360954e-107        Asb9\n",
       "  ...                   ...       ...            ...         ...\n",
       "  15978  ENSMUSG00000072949 -2.036016  9.089100e-286       Acot1\n",
       "  15979  ENSMUSG00000031563 -2.307052   0.000000e+00        Wwc2\n",
       "  15980  ENSMUSG00000024197 -2.367655   0.000000e+00       Plin3\n",
       "  15981  ENSMUSG00000043336 -2.854150   0.000000e+00     Filip1l\n",
       "  15982  ENSMUSG00000053007 -2.872495  8.025695e-231       Creb5\n",
       "  \n",
       "  [15983 rows x 4 columns],\n",
       "  'gamma':                      gene     logFC           padj gene_symbol\n",
       "  0      ENSMUSG00000017316  3.999618  2.149552e-308         Ppy\n",
       "  1      ENSMUSG00000031430  2.876311  3.046279e-100       Vsig1\n",
       "  2      ENSMUSG00000063672  2.685109  8.962613e-158      Nkx6-3\n",
       "  3      ENSMUSG00000097850  2.264223  6.230412e-128    Mir205hg\n",
       "  4      ENSMUSG00000041552  1.832270  1.319095e-123      Ptchd1\n",
       "  ...                   ...       ...            ...         ...\n",
       "  15978  ENSMUSG00000031712 -1.905402   1.930760e-74        Il15\n",
       "  15979  ENSMUSG00000041696 -1.977261  1.014724e-108      Rasl12\n",
       "  15980  ENSMUSG00000042826 -2.180555   5.548893e-43       Fgf11\n",
       "  15981  ENSMUSG00000041075 -2.411146   4.823158e-53        Fzd7\n",
       "  15982  ENSMUSG00000043439 -2.918702   1.220065e-99        Epop\n",
       "  \n",
       "  [15983 rows x 4 columns],\n",
       "  'delta':                      gene     logFC           padj gene_symbol\n",
       "  0      ENSMUSG00000002930  7.114148  2.610209e-218     Ppp1r17\n",
       "  1      ENSMUSG00000004366  6.983684   0.000000e+00         Sst\n",
       "  2      ENSMUSG00000067578  6.831414  3.730726e-269       Cbln4\n",
       "  3      ENSMUSG00000051136  6.413016  8.593626e-285        Ghsr\n",
       "  4      ENSMUSG00000022577  6.386551  5.836892e-320        Ly6h\n",
       "  ...                   ...       ...            ...         ...\n",
       "  15978  ENSMUSG00000041939 -1.338510   1.423883e-95         Mvk\n",
       "  15979  ENSMUSG00000028266 -1.547366   0.000000e+00        Lmo4\n",
       "  15980  ENSMUSG00000038349 -1.677550  2.033204e-196       Plcl1\n",
       "  15981  ENSMUSG00000019194 -2.021996  1.998609e-162       Scn1b\n",
       "  15982  ENSMUSG00000066278 -2.329224  5.089735e-279      Vps37b\n",
       "  \n",
       "  [15983 rows x 4 columns]},\n",
       " 'embryonic_DE': {'Alpha':                      gene     logFC           padj gene_symbol\n",
       "  0      ENSMUSG00000023964  7.044834   2.333485e-40       Calcr\n",
       "  1      ENSMUSG00000000394  4.466205  6.380238e-162         Gcg\n",
       "  2      ENSMUSG00000000392  3.239607   9.871047e-11         Fap\n",
       "  3      ENSMUSG00000051062  3.194766   8.013153e-04       Fbll1\n",
       "  4      ENSMUSG00000109125  3.158323   1.217595e-35     Gm45159\n",
       "  ...                   ...       ...            ...         ...\n",
       "  14362  ENSMUSG00000022836 -2.318480   9.183928e-07        Mylk\n",
       "  14363  ENSMUSG00000045775 -2.617580   1.220346e-46     Slc16a5\n",
       "  14364  ENSMUSG00000047854 -2.803694   1.253859e-18       Stx19\n",
       "  14365  ENSMUSG00000093662 -2.936315   3.696392e-30     Gm20649\n",
       "  14366  ENSMUSG00000030103 -3.720680   1.894412e-21     Bhlhe40\n",
       "  \n",
       "  [14367 rows x 4 columns],\n",
       "  'Beta':                      gene     logFC           padj    gene_symbol\n",
       "  0      ENSMUSG00000000215  6.963225   2.180709e-84           Ins2\n",
       "  1      ENSMUSG00000035804  6.403119  9.435797e-173           Ins1\n",
       "  2      ENSMUSG00000029819  6.332436  6.120832e-126            Npy\n",
       "  3      ENSMUSG00000036760  5.980911   1.067479e-05          Kcnk9\n",
       "  4      ENSMUSG00000086995  4.567796   6.670258e-03        Gm13544\n",
       "  ...                   ...       ...            ...            ...\n",
       "  14362  ENSMUSG00000097762 -3.661904   9.776838e-07  4732463B04Rik\n",
       "  14363  ENSMUSG00000029597 -3.933221   2.760694e-29            Sds\n",
       "  14364  ENSMUSG00000079685 -4.184818   5.964792e-10          Ulbp1\n",
       "  14365  ENSMUSG00000068877 -4.677352   2.904425e-47       Selenbp2\n",
       "  14366  ENSMUSG00000020102 -5.039276   8.967606e-23        Slc16a7\n",
       "  \n",
       "  [14367 rows x 4 columns],\n",
       "  'Delta':                      gene     logFC           padj gene_symbol\n",
       "  0      ENSMUSG00000004366  7.636632  1.015419e-129         Sst\n",
       "  1      ENSMUSG00000022887  6.043206   1.333360e-10       Masp1\n",
       "  2      ENSMUSG00000073421  4.887491   4.060510e-10      H2-Ab1\n",
       "  3      ENSMUSG00000038642  4.872158   3.695156e-05        Ctss\n",
       "  4      ENSMUSG00000027400  4.674240   1.246338e-03        Pdyn\n",
       "  ...                   ...       ...            ...         ...\n",
       "  14362  ENSMUSG00000042590 -7.833902   2.757987e-08       Ipo11\n",
       "  14363  ENSMUSG00000028786 -7.986881   2.025925e-24      Tmem54\n",
       "  14364  ENSMUSG00000015846 -8.034933   1.383605e-05        Rxra\n",
       "  14365  ENSMUSG00000075327 -8.228585   6.056173e-04       Zbtb2\n",
       "  14366  ENSMUSG00000027087 -8.539751   1.110266e-06       Itgav\n",
       "  \n",
       "  [14367 rows x 4 columns],\n",
       "  'Epsilon':                      gene     logFC          padj gene_symbol\n",
       "  0      ENSMUSG00000067261  7.542560  4.777554e-34       Foxd3\n",
       "  1      ENSMUSG00000064177  6.669233  6.172560e-77        Ghrl\n",
       "  2      ENSMUSG00000086096  6.529931  3.596753e-04     Gm12688\n",
       "  3      ENSMUSG00000054667  5.893363  1.206572e-38        Irs4\n",
       "  4      ENSMUSG00000039904  5.392943  3.699497e-03       Gpr37\n",
       "  ...                   ...       ...           ...         ...\n",
       "  14362  ENSMUSG00000019845 -6.395489  1.671676e-08       Tube1\n",
       "  14363  ENSMUSG00000100235 -6.407591  4.788051e-04     Gm28557\n",
       "  14364  ENSMUSG00000063446 -6.721722  3.553311e-16      Plppr1\n",
       "  14365  ENSMUSG00000020897 -6.873099  1.290140e-24       Aurkb\n",
       "  14366  ENSMUSG00000029999 -6.913000  2.401649e-05        Tgfa\n",
       "  \n",
       "  [14367 rows x 4 columns]}}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "23c0dc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'endocrine_markers.xlsx',engine='xlsxwriter') \n",
    "for age,res_sub1 in res.items():\n",
    "    for ct,res_sub2 in res_sub1.items():\n",
    "        res_sub2.to_excel(writer, sheet_name=age+'_'+ct.lower(),index=False)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "e273568d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/endocrine_markers.xlsx'"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'endocrine_markers.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1917224d",
   "metadata": {},
   "source": [
    "## Human diabetes comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "634d9f33",
   "metadata": {},
   "outputs": [],
   "source": [
    "res={}\n",
    "res['T1D_DE']=pd.read_table(path_data+'de_human/diabetes/T1D_vs_healthy.tsv')\n",
    "res['T2D_DE']=pd.read_table(path_data+'de_human/diabetes/T2D_vs_healthy.tsv')\n",
    "for dt,res_sub in pd.read_excel(\n",
    "    path_data+'de_human/diabetes/human_overlap_enrichment.xlsx',sheet_name=None,index_col=0).items():\n",
    "    res_sub.rename({'label':'gene_set'},axis=1,inplace=True)\n",
    "    res[f'{dt}_DE_enrichment']=res_sub\n",
    "res['mouse_genesets_DE']=pd.read_table(\n",
    "    path_data+'de_human/diabetes/mouse_models_genesets_lfcDvsH.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e8f0d45b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'T1D_DE':       gene_symbol          padj     logFC    dataset  n_group  n_ref  \\\n",
       " 0             CRP  2.653762e-04  5.404300   GSE83139        6     67   \n",
       " 1           FXYD3  9.586213e-05  6.465961   GSE83139        6     67   \n",
       " 2           ZNF91  2.759576e-12 -7.388538   GSE83139        6     67   \n",
       " 3            TSIX  4.123457e-12 -8.402358   GSE83139        6     67   \n",
       " 4            XIST  4.123457e-12 -8.673610   GSE83139        6     67   \n",
       " ...           ...           ...       ...        ...      ...    ...   \n",
       " 41687       KANTR  9.967668e-01  0.001006  GSE148073     1071   3158   \n",
       " 41688        GALT  9.970685e-01  0.000884  GSE148073     1071   3158   \n",
       " 41689       RPL24  9.986011e-01  0.000072  GSE148073     1071   3158   \n",
       " 41690        BFAR  9.986823e-01 -0.000186  GSE148073     1071   3158   \n",
       " 41691  TMEM263-DT  9.996062e-01 -0.000075  GSE148073     1071   3158   \n",
       " \n",
       "        ratio_group  \n",
       " 0         1.000000  \n",
       " 1         1.000000  \n",
       " 2         1.000000  \n",
       " 3         1.000000  \n",
       " 4         1.000000  \n",
       " ...            ...  \n",
       " 41687     0.111111  \n",
       " 41688     0.226891  \n",
       " 41689     0.061625  \n",
       " 41690     0.443511  \n",
       " 41691     0.235294  \n",
       " \n",
       " [41692 rows x 7 columns],\n",
       " 'T2D_DE':       gene_symbol          padj     logFC   dataset  n_group  n_ref  \\\n",
       " 0            TRY6  1.087087e-10 -4.807086  GSE83139       38     67   \n",
       " 1           PRSS1  1.087087e-10 -4.266847  GSE83139       38     67   \n",
       " 2           CTRB2  2.541567e-08 -3.579158  GSE83139       38     67   \n",
       " 3           REG1B  6.727300e-08 -4.399756  GSE83139       38     67   \n",
       " 4           PRSS2  8.177571e-08 -3.875727  GSE83139       38     67   \n",
       " ...           ...           ...       ...       ...      ...    ...   \n",
       " 86131   LOC574080  9.997393e-01 -0.000345  GSE81608      296    207   \n",
       " 86132        EMSY  9.998183e-01 -0.000191  GSE81608      296    207   \n",
       " 86133      KCTD13  9.998207e-01  0.000154  GSE81608      296    207   \n",
       " 86134       PIAS1  9.998207e-01  0.000077  GSE81608      296    207   \n",
       " 86135    TMEM167B  9.998225e-01  0.000072  GSE81608      296    207   \n",
       " \n",
       "        ratio_group  \n",
       " 0         1.000000  \n",
       " 1         1.000000  \n",
       " 2         1.000000  \n",
       " 3         1.000000  \n",
       " 4         1.000000  \n",
       " ...            ...  \n",
       " 86131     0.385135  \n",
       " 86132     0.158784  \n",
       " 86133     0.016892  \n",
       " 86134     0.111486  \n",
       " 86135     0.449324  \n",
       " \n",
       " [86136 rows x 7 columns],\n",
       " 'T1D_DE_enrichment':                                                                                              gene_set  \\\n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                               GO:0042611_GOCC_MHC_PROTEIN_COMPLEX   \n",
       " GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENT...   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENT...   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX               GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX   \n",
       " R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY   R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY   \n",
       " ...                                                                                               ...   \n",
       " GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION                   GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION   \n",
       " GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS     GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS   \n",
       " GO:0031016_GOBP_PANCREAS_DEVELOPMENT                             GO:0031016_GOBP_PANCREAS_DEVELOPMENT   \n",
       " GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS   GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS   \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...  GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEO...   \n",
       " \n",
       "                                                             pval      fdr  \\\n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                 2.200000e-08  0.00015   \n",
       " GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  2.700000e-08  0.00015   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  2.000000e-07  0.00058   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX         2.000000e-07  0.00058   \n",
       " R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY   4.000000e-07  0.00091   \n",
       " ...                                                          ...      ...   \n",
       " GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION           4.100000e-03  0.24000   \n",
       " GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS    4.200000e-03  0.24000   \n",
       " GO:0031016_GOBP_PANCREAS_DEVELOPMENT                4.300000e-03  0.25000   \n",
       " GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS   4.300000e-03  0.25000   \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...  4.300000e-03  0.25000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                        32       22   \n",
       " GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...         32       23   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...         32        9   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX                32        9   \n",
       " R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY          32       11   \n",
       " ...                                                       ...      ...   \n",
       " GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION                  32      476   \n",
       " GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS           32       69   \n",
       " GO:0031016_GOBP_PANCREAS_DEVELOPMENT                       32       70   \n",
       " GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS          32      237   \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...         32      481   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                       4       22975   \n",
       " GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...        4       22975   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...        3       22975   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX               3       22975   \n",
       " R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY         3       22975   \n",
       " ...                                                     ...         ...   \n",
       " GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION                 4       22975   \n",
       " GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS          2       22975   \n",
       " GO:0031016_GOBP_PANCREAS_DEVELOPMENT                      2       22975   \n",
       " GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS         3       22975   \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...        4       22975   \n",
       " \n",
       "                                                                     hits  \\\n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                 B2M,CD74,HLA-A,HLA-B   \n",
       " GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  B2M,CD74,HLA-A,HLA-B   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       B2M,HLA-A,HLA-B   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX              B2M,HLA-A,HLA-B   \n",
       " R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY        B2M,HLA-A,HLA-B   \n",
       " ...                                                                  ...   \n",
       " GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION            B2M,CD74,EGR1,HLA-B   \n",
       " GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS                B2M,CD74   \n",
       " GO:0031016_GOBP_PANCREAS_DEVELOPMENT                           HES1,IGF2   \n",
       " GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS        CD74,EGR1,PSMB9   \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...   B2M,EGR1,LCN2,MUC13   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                 0.181818          32  \n",
       " GO:0019883_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.173913          32  \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.333333          32  \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX         0.333333          32  \n",
       " R-HSA-1236977_REACTOME_ENDOSOMAL_VACUOLAR_PATHWAY   0.272727          32  \n",
       " ...                                                      ...         ...  \n",
       " GO:0002521_GOBP_LEUKOCYTE_DIFFERENTIATION           0.008403          32  \n",
       " GO:0033077_GOBP_T_CELL_DIFFERENTIATION_IN_THYMUS    0.028986          32  \n",
       " GO:0031016_GOBP_PANCREAS_DEVELOPMENT                0.028571          32  \n",
       " GO:0042180_GOBP_CELLULAR_KETONE_METABOLIC_PROCESS   0.012658          32  \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...  0.008316          32  \n",
       " \n",
       " [198 rows x 10 columns],\n",
       " 'T2D_DE_enrichment':                                                                                              gene_set  \\\n",
       " GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS                   GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS   \n",
       " GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT             GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT               GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY               GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY   \n",
       " GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHET...  GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHE...   \n",
       " ...                                                                                               ...   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING           GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION  GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHE...   \n",
       " GO:0032094_GOBP_RESPONSE_TO_FOOD                                     GO:0032094_GOBP_RESPONSE_TO_FOOD   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS                 hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS   \n",
       " GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_C...  GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_...   \n",
       " \n",
       "                                                             pval     fdr  \\\n",
       " GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS           3.600000e-07  0.0041   \n",
       " GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT        4.800000e-06  0.0140   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT         5.500000e-06  0.0140   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         6.000000e-06  0.0140   \n",
       " GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHET...  6.800000e-06  0.0140   \n",
       " ...                                                          ...     ...   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING       2.400000e-03  0.2200   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION  2.600000e-03  0.2300   \n",
       " GO:0032094_GOBP_RESPONSE_TO_FOOD                    2.600000e-03  0.2300   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          2.600000e-03  0.2300   \n",
       " GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_C...  2.700000e-03  0.2400   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS                  59      206   \n",
       " GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT               59      111   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT                59      201   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY                59      446   \n",
       " GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHET...         59       17   \n",
       " ...                                                       ...      ...   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING              59       33   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION         59      123   \n",
       " GO:0032094_GOBP_RESPONSE_TO_FOOD                           59       34   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS                 59      124   \n",
       " GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_C...         59       35   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS                 7       26683   \n",
       " GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT              5       26683   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT               6       26683   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY               8       26683   \n",
       " GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHET...        3       26683   \n",
       " ...                                                     ...         ...   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING             2       26683   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION        3       26683   \n",
       " GO:0032094_GOBP_RESPONSE_TO_FOOD                          2       26683   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS                3       26683   \n",
       " GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_C...        2       26683   \n",
       " \n",
       "                                                                                       hits  \\\n",
       " GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS            BMP5,DKK3,EGR1,PCSK1N,PCSK2,RBP1,SPP1   \n",
       " GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT                 BMP5,DKK3,NKX2-2,ONECUT2,SIX3   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT            BMP5,MSN,NKX2-2,ONECUT2,SIX3,SPINT2   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         BMP5,CHGB,DKK3,GCG,MRAP2,NENF,NPY,SPP1   \n",
       " GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHET...                          BMP5,DKK3,EGR1   \n",
       " ...                                                                                    ...   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING                                    MRAP2,NPY   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION                     CDKN2A,FMN1,ONECUT2   \n",
       " GO:0032094_GOBP_RESPONSE_TO_FOOD                                                  NENF,NPY   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS                               CADM1,HLA-A,HLA-B   \n",
       " GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_C...                             HLA-A,HLA-B   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS           0.033981          59  \n",
       " GO:0035270_GOBP_ENDOCRINE_SYSTEM_DEVELOPMENT        0.045045          59  \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT         0.029851          59  \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         0.017937          59  \n",
       " GO:0010566_GOBP_REGULATION_OF_KETONE_BIOSYNTHET...  0.176471          59  \n",
       " ...                                                      ...         ...  \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING       0.060606          59  \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION  0.024390          59  \n",
       " GO:0032094_GOBP_RESPONSE_TO_FOOD                    0.058824          59  \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          0.024194          59  \n",
       " GO:0001914_GOBP_REGULATION_OF_T_CELL_MEDIATED_C...  0.057143          59  \n",
       " \n",
       " [130 rows x 10 columns],\n",
       " 'mouse_genesets_DE':     model                                           gene_set  me_logFC  \\\n",
       " 0     NOD        GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX  0.837413   \n",
       " 1   db/db        GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX -0.075227   \n",
       " 2    mSTZ        GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX  0.044751   \n",
       " 3     NOD     GO:0019730_GOBP_ANTIMICROBIAL_HUMORAL_RESPONSE  0.418911   \n",
       " 4   db/db     GO:0019730_GOBP_ANTIMICROBIAL_HUMORAL_RESPONSE  0.053417   \n",
       " 5    mSTZ     GO:0019730_GOBP_ANTIMICROBIAL_HUMORAL_RESPONSE  0.143032   \n",
       " 6     NOD  GO:0032104_GOBP_REGULATION_OF_RESPONSE_TO_EXTR... -0.057049   \n",
       " 7   db/db  GO:0032104_GOBP_REGULATION_OF_RESPONSE_TO_EXTR...  0.821465   \n",
       " 8    mSTZ  GO:0032104_GOBP_REGULATION_OF_RESPONSE_TO_EXTR...  0.346937   \n",
       " 9     NOD         GO:0030658_GOCC_TRANSPORT_VESICLE_MEMBRANE  0.258124   \n",
       " 10  db/db         GO:0030658_GOCC_TRANSPORT_VESICLE_MEMBRANE  0.142228   \n",
       " 11   mSTZ         GO:0030658_GOCC_TRANSPORT_VESICLE_MEMBRANE  0.191906   \n",
       " 12    NOD     GO:0031018_GOBP_ENDOCRINE_PANCREAS_DEVELOPMENT  0.014874   \n",
       " 13  db/db     GO:0031018_GOBP_ENDOCRINE_PANCREAS_DEVELOPMENT -0.272897   \n",
       " 14   mSTZ     GO:0031018_GOBP_ENDOCRINE_PANCREAS_DEVELOPMENT -0.164703   \n",
       " 15    NOD          GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS -0.076720   \n",
       " 16  db/db          GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS  0.191907   \n",
       " 17   mSTZ          GO:0042445_GOBP_HORMONE_METABOLIC_PROCESS  0.108625   \n",
       " 18    NOD          GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION -0.035559   \n",
       " 19  db/db          GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION  0.122277   \n",
       " 20   mSTZ          GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION  0.110364   \n",
       " 21    NOD                GO:0042254_GOBP_RIBOSOME_BIOGENESIS  0.007705   \n",
       " 22  db/db                GO:0042254_GOBP_RIBOSOME_BIOGENESIS  0.201538   \n",
       " 23   mSTZ                GO:0042254_GOBP_RIBOSOME_BIOGENESIS  0.116894   \n",
       " 24    NOD  GO:1900037_GOBP_REGULATION_OF_CELLULAR_RESPONS...  0.006861   \n",
       " 25  db/db  GO:1900037_GOBP_REGULATION_OF_CELLULAR_RESPONS...  0.161162   \n",
       " 26   mSTZ  GO:1900037_GOBP_REGULATION_OF_CELLULAR_RESPONS...  0.197043   \n",
       " 27    NOD  GO:0010498_GOBP_PROTEASOMAL_PROTEIN_CATABOLIC_...  0.009777   \n",
       " 28  db/db  GO:0010498_GOBP_PROTEASOMAL_PROTEIN_CATABOLIC_...  0.122873   \n",
       " 29   mSTZ  GO:0010498_GOBP_PROTEASOMAL_PROTEIN_CATABOLIC_...  0.189798   \n",
       " \n",
       "              pval  \n",
       " 0   1.987168e-139  \n",
       " 1    1.481093e-57  \n",
       " 2    6.545152e-07  \n",
       " 3   1.030383e-147  \n",
       " 4    5.807052e-65  \n",
       " 5    1.686833e-79  \n",
       " 6    1.930270e-04  \n",
       " 7    0.000000e+00  \n",
       " 8   5.942459e-288  \n",
       " 9   1.635492e-108  \n",
       " 10   0.000000e+00  \n",
       " 11  5.857748e-254  \n",
       " 12   1.318416e-01  \n",
       " 13   0.000000e+00  \n",
       " 14   1.365603e-83  \n",
       " 15   2.080443e-13  \n",
       " 16   0.000000e+00  \n",
       " 17   5.669616e-46  \n",
       " 18   5.499381e-09  \n",
       " 19   0.000000e+00  \n",
       " 20  4.088351e-103  \n",
       " 21   2.608601e-01  \n",
       " 22   0.000000e+00  \n",
       " 23  7.749837e-253  \n",
       " 24   6.573771e-01  \n",
       " 25   0.000000e+00  \n",
       " 26  6.406475e-164  \n",
       " 27   7.295625e-02  \n",
       " 28   0.000000e+00  \n",
       " 29  3.083857e-197  }"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "07947ae8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'human_diabetes_de.xlsx',engine='xlsxwriter') \n",
    "for name,res_sub1 in res.items():\n",
    "    res_sub1.to_excel(writer, sheet_name=name,index=False)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7fa070bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/human_diabetes_de.xlsx'"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'human_diabetes_de.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "530042f3",
   "metadata": {},
   "source": [
    "## GPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "554e189a",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Load table parts\n",
    "res={}\n",
    "res['GPs']=pd.read_table(path_data+'combined/moransi/sfintegrated/gene_hc_t2.4.tsv',\n",
    "                         index_col=0)\n",
    "res['GPs'].index.name='EID'\n",
    "for gp,res_sub in pd.read_excel(\n",
    "    path_data+'combined/moransi/sfintegrated/gene_hc_t2.4_enrichment.xlsx', \n",
    "    index_col=0,sheet_name=None).items():\n",
    "    res_sub.index.name='gene_set'\n",
    "    res_sub.drop('label',axis=1,inplace=True)\n",
    "    res[f'GP{gp}_enrichment']=res_sub\n",
    "res['sample_explained_var']=pd.read_table(\n",
    "    path_data+'combined/moransi/sfintegrated/explained_var/per_sample/explainedVar_GPsample_significanceSummary_parsed.tsv',\n",
    "    index_col=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "72ff632b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GPs':                     hc gene_symbol  rel_beta_expr  mean_expr_in_expr_cells  \\\n",
       " EID                                                                          \n",
       " ENSMUSG00000025907   1      Rb1cc1            1.0                 0.808519   \n",
       " ENSMUSG00000099032   2       Tcf24            1.0                 0.605948   \n",
       " ENSMUSG00000025935   2       Tram1            1.0                 1.637445   \n",
       " ENSMUSG00000042215  17        Bag2            1.0                 0.675964   \n",
       " ENSMUSG00000026131  13         Dst            1.0                 1.273423   \n",
       " ...                 ..         ...            ...                      ...   \n",
       " ENSMUSG00000064363  13      mt-Nd4            1.0                 3.757161   \n",
       " ENSMUSG00000064367   1      mt-Nd5            1.0                 2.021609   \n",
       " ENSMUSG00000064368   1      mt-Nd6            1.0                 0.684435   \n",
       " ENSMUSG00000064370  13     mt-Cytb            1.0                 4.252210   \n",
       " ENSMUSG00000051412   2       Vamp7            1.0                 0.746626   \n",
       " \n",
       "                     N_PMID  N_PMID_pancreas_notCancerNonendo  \n",
       " EID                                                           \n",
       " ENSMUSG00000025907   124.0                               1.0  \n",
       " ENSMUSG00000099032     7.0                               0.0  \n",
       " ENSMUSG00000025935    44.0                               3.0  \n",
       " ENSMUSG00000042215   169.0                               0.0  \n",
       " ENSMUSG00000026131   135.0                               1.0  \n",
       " ...                    ...                               ...  \n",
       " ENSMUSG00000064363    79.0                               6.0  \n",
       " ENSMUSG00000064367    70.0                               5.0  \n",
       " ENSMUSG00000064368    42.0                               4.0  \n",
       " ENSMUSG00000064370    39.0                               5.0  \n",
       " ENSMUSG00000051412     0.0                               0.0  \n",
       " \n",
       " [1687 rows x 6 columns],\n",
       " 'GP1_enrichment':                                                             pval     fdr  \\\n",
       " gene_set                                                                   \n",
       " GO:0048193_GOBP_GOLGI_VESICLE_TRANSPORT             1.500000e-07  0.0016   \n",
       " GO:0006888_GOBP_ENDOPLASMIC_RETICULUM_TO_GOLGI_...  4.200000e-07  0.0016   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  4.300000e-07  0.0016   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...  1.400000e-06  0.0039   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...  4.300000e-06  0.0094   \n",
       " R-HSA-9703465_REACTOME_SIGNALING_BY_FLT3_FUSION...  1.100000e-05  0.0210   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY        1.500000e-05  0.0230   \n",
       " R-HSA-9682385_REACTOME_FLT3_SIGNALING_IN_DISEASE    2.800000e-05  0.0380   \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...  7.600000e-05  0.0910   \n",
       " GO:0043001_GOBP_GOLGI_TO_PLASMA_MEMBRANE_PROTEI...  8.800000e-05  0.0960   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...  9.900000e-05  0.0980   \n",
       " GO:0048588_GOBP_DEVELOPMENTAL_CELL_GROWTH           1.300000e-04  0.1200   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION  1.900000e-04  0.1400   \n",
       " GO:0003955_GOMF_NAD_P_H_DEHYDROGENASE_QUINONE_A...  2.000000e-04  0.1400   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    2.100000e-04  0.1400   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                   2.200000e-04  0.1400   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             2.200000e-04  0.1400   \n",
       " GO:0006120_GOBP_MITOCHONDRIAL_ELECTRON_TRANSPOR...  2.900000e-04  0.1800   \n",
       " GO:0006893_GOBP_GOLGI_TO_PLASMA_MEMBRANE_TRANSPORT  3.200000e-04  0.1800   \n",
       " GO:0016655_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  3.200000e-04  0.1800   \n",
       " GO:0061951_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...  3.400000e-04  0.1800   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           3.600000e-04  0.1800   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...  3.900000e-04  0.1800   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION         4.000000e-04  0.1800   \n",
       " GO:0030126_GOCC_COPI_VESICLE_COAT                   4.600000e-04  0.2000   \n",
       " GO:0150116_GOBP_REGULATION_OF_CELL_SUBSTRATE_JU...  5.600000e-04  0.2300   \n",
       " GO:1990138_GOBP_NEURON_PROJECTION_EXTENSION         6.000000e-04  0.2300   \n",
       " GO:0010256_GOBP_ENDOMEMBRANE_SYSTEM_ORGANIZATION    6.200000e-04  0.2300   \n",
       " GO:0016887_GOMF_ATPASE_ACTIVITY                     6.200000e-04  0.2300   \n",
       " GO:0005200_GOMF_STRUCTURAL_CONSTITUENT_OF_CYTOS...  6.700000e-04  0.2400   \n",
       " GO:0090066_GOBP_REGULATION_OF_ANATOMICAL_STRUCT...  7.100000e-04  0.2500   \n",
       " GO:0015629_GOCC_ACTIN_CYTOSKELETON                  7.500000e-04  0.2500   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0048193_GOBP_GOLGI_VESICLE_TRANSPORT                    47      347   \n",
       " GO:0006888_GOBP_ENDOPLASMIC_RETICULUM_TO_GOLGI_...         47      191   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...         47      284   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...         47      144   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...         47      174   \n",
       " R-HSA-9703465_REACTOME_SIGNALING_BY_FLT3_FUSION...         47       18   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY               47      458   \n",
       " R-HSA-9682385_REACTOME_FLT3_SIGNALING_IN_DISEASE           47       24   \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...         47       91   \n",
       " GO:0043001_GOBP_GOLGI_TO_PLASMA_MEMBRANE_PROTEI...         47       35   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...         47      188   \n",
       " GO:0048588_GOBP_DEVELOPMENTAL_CELL_GROWTH                  47      200   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION         47      115   \n",
       " GO:0003955_GOMF_NAD_P_H_DEHYDROGENASE_QUINONE_A...         47       46   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                           47      118   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                          47      351   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                    47      120   \n",
       " GO:0006120_GOBP_MITOCHONDRIAL_ELECTRON_TRANSPOR...         47       52   \n",
       " GO:0006893_GOBP_GOLGI_TO_PLASMA_MEMBRANE_TRANSPORT         47       54   \n",
       " GO:0016655_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...         47       54   \n",
       " GO:0061951_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...         47       55   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION                  47      136   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...         47      253   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION                47      393   \n",
       " GO:0030126_GOCC_COPI_VESICLE_COAT                          47       13   \n",
       " GO:0150116_GOBP_REGULATION_OF_CELL_SUBSTRATE_JU...         47       65   \n",
       " GO:1990138_GOBP_NEURON_PROJECTION_EXTENSION                47      156   \n",
       " GO:0010256_GOBP_ENDOMEMBRANE_SYSTEM_ORGANIZATION           47      427   \n",
       " GO:0016887_GOMF_ATPASE_ACTIVITY                            47      427   \n",
       " GO:0005200_GOMF_STRUCTURAL_CONSTITUENT_OF_CYTOS...         47       69   \n",
       " GO:0090066_GOBP_REGULATION_OF_ANATOMICAL_STRUCT...         47      438   \n",
       " GO:0015629_GOCC_ACTIN_CYTOSKELETON                         47      443   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0048193_GOBP_GOLGI_VESICLE_TRANSPORT                   9       18965   \n",
       " GO:0006888_GOBP_ENDOPLASMIC_RETICULUM_TO_GOLGI_...        7       18965   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...        8       18965   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...        6       18965   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...        6       18965   \n",
       " R-HSA-9703465_REACTOME_SIGNALING_BY_FLT3_FUSION...        3       18965   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY              8       18965   \n",
       " R-HSA-9682385_REACTOME_FLT3_SIGNALING_IN_DISEASE          3       18965   \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...        4       18965   \n",
       " GO:0043001_GOBP_GOLGI_TO_PLASMA_MEMBRANE_PROTEI...        3       18965   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...        5       18965   \n",
       " GO:0048588_GOBP_DEVELOPMENTAL_CELL_GROWTH                 5       18965   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION        4       18965   \n",
       " GO:0003955_GOMF_NAD_P_H_DEHYDROGENASE_QUINONE_A...        3       18965   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                          4       18965   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                         6       18965   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                   4       18965   \n",
       " GO:0006120_GOBP_MITOCHONDRIAL_ELECTRON_TRANSPOR...        3       18965   \n",
       " GO:0006893_GOBP_GOLGI_TO_PLASMA_MEMBRANE_TRANSPORT        3       18965   \n",
       " GO:0016655_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...        3       18965   \n",
       " GO:0061951_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...        3       18965   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION                 4       18965   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...        5       18965   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION               6       18965   \n",
       " GO:0030126_GOCC_COPI_VESICLE_COAT                         2       18965   \n",
       " GO:0150116_GOBP_REGULATION_OF_CELL_SUBSTRATE_JU...        3       18965   \n",
       " GO:1990138_GOBP_NEURON_PROJECTION_EXTENSION               4       18965   \n",
       " GO:0010256_GOBP_ENDOMEMBRANE_SYSTEM_ORGANIZATION          6       18965   \n",
       " GO:0016887_GOMF_ATPASE_ACTIVITY                           6       18965   \n",
       " GO:0005200_GOMF_STRUCTURAL_CONSTITUENT_OF_CYTOS...        3       18965   \n",
       " GO:0090066_GOBP_REGULATION_OF_ANATOMICAL_STRUCT...        6       18965   \n",
       " GO:0015629_GOCC_ACTIN_CYTOSKELETON                        6       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0048193_GOBP_GOLGI_VESICLE_TRANSPORT             Arcn1,Copa,Mia2,Golga4,Golgb1,Macf1,Mia3,Sptbn...   \n",
       " GO:0006888_GOBP_ENDOPLASMIC_RETICULUM_TO_GOLGI_...          Arcn1,Copa,Mia2,Golgb1,Mia3,Sptbn1,Trip11   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...     Arcn1,Copa,Gfpt1,Golgb1,Mia2,Mia3,Sptbn1,Uggt1   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...                 Arcn1,Copa,Golgb1,Mia2,Mia3,Sptbn1   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...                 Arcn1,Copa,Golgb1,Mia2,Mia3,Sptbn1   \n",
       " R-HSA-9703465_REACTOME_SIGNALING_BY_FLT3_FUSION...                               Golgb1,Sptbn1,Trip11   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY          Cd2ap,Copa,Eif3a,Macf1,Pclo,Rock2,Sorbs2,Sptbn1   \n",
       " R-HSA-9682385_REACTOME_FLT3_SIGNALING_IN_DISEASE                                 Golgb1,Sptbn1,Trip11   \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...                           Arcn1,Copa,Golgb1,Sptbn1   \n",
       " GO:0043001_GOBP_GOLGI_TO_PLASMA_MEMBRANE_PROTEI...                                Golga4,Macf1,Sptbn1   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...                     Arcn1,Copa,Golga4,Kif1b,Trip11   \n",
       " GO:0048588_GOBP_DEVELOPMENTAL_CELL_GROWTH                             Golga4,Macf1,Sorbs2,Spag9,Usp9x   \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION                              Macf1,Rock2,Utrn,Wnt4   \n",
       " GO:0003955_GOMF_NAD_P_H_DEHYDROGENASE_QUINONE_A...                              mt-Nd4l,mt-Nd5,mt-Nd6   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                                        mt-Atp8,mt-Nd4l,mt-Nd5,mt-Nd6   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                                    Abca1,Abcc8,Cd2ap,Mia2,Mia3,Pclo   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                                 mt-Atp8,mt-Nd4l,mt-Nd5,mt-Nd6   \n",
       " GO:0006120_GOBP_MITOCHONDRIAL_ELECTRON_TRANSPOR...                              mt-Nd4l,mt-Nd5,mt-Nd6   \n",
       " GO:0006893_GOBP_GOLGI_TO_PLASMA_MEMBRANE_TRANSPORT                                Golga4,Macf1,Sptbn1   \n",
       " GO:0016655_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...                              mt-Nd4l,mt-Nd5,mt-Nd6   \n",
       " GO:0061951_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...                                Golga4,Macf1,Sptbn1   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION                               mt-Atp8,mt-Nd4l,mt-Nd5,mt-Nd6   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...                 Gfpt1,mt-Nd4l,mt-Nd5,mt-Nd6,Rb1cc1   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION                     Cd2ap,Cyfip2,Rock2,Sorbs2,Sptbn1,Wnt4   \n",
       " GO:0030126_GOCC_COPI_VESICLE_COAT                                                          Arcn1,Copa   \n",
       " GO:0150116_GOBP_REGULATION_OF_CELL_SUBSTRATE_JU...                                   Macf1,Rock2,Wnt4   \n",
       " GO:1990138_GOBP_NEURON_PROJECTION_EXTENSION                                  Golga4,Macf1,Spag9,Usp9x   \n",
       " GO:0010256_GOBP_ENDOMEMBRANE_SYSTEM_ORGANIZATION                Abca1,Golgb1,Hook1,Mia3,Sptbn1,Trip11   \n",
       " GO:0016887_GOMF_ATPASE_ACTIVITY                                     Abca1,Abcc8,Atrx,Kif1b,Macf1,Nav2   \n",
       " GO:0005200_GOMF_STRUCTURAL_CONSTITUENT_OF_CYTOS...                                Cd2ap,Sorbs2,Sptbn1   \n",
       " GO:0090066_GOBP_REGULATION_OF_ANATOMICAL_STRUCT...            Cyfip2,Golga4,Macf1,Rb1cc1,Rock2,Sptbn1   \n",
       " GO:0015629_GOCC_ACTIN_CYTOSKELETON                              Cd2ap,Filip1,Macf1,Sorbs2,Sptbn1,Utrn   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0048193_GOBP_GOLGI_VESICLE_TRANSPORT             0.025937          47  \n",
       " GO:0006888_GOBP_ENDOPLASMIC_RETICULUM_TO_GOLGI_...  0.036649          47  \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  0.028169          47  \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...  0.041667          47  \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...  0.034483          47  \n",
       " R-HSA-9703465_REACTOME_SIGNALING_BY_FLT3_FUSION...  0.166667          47  \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY        0.017467          47  \n",
       " R-HSA-9682385_REACTOME_FLT3_SIGNALING_IN_DISEASE    0.125000          47  \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...  0.043956          47  \n",
       " GO:0043001_GOBP_GOLGI_TO_PLASMA_MEMBRANE_PROTEI...  0.085714          47  \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...  0.026596          47  \n",
       " GO:0048588_GOBP_DEVELOPMENTAL_CELL_GROWTH           0.025000          47  \n",
       " GO:0001952_GOBP_REGULATION_OF_CELL_MATRIX_ADHESION  0.034783          47  \n",
       " GO:0003955_GOMF_NAD_P_H_DEHYDROGENASE_QUINONE_A...  0.065217          47  \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    0.033898          47  \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                   0.017094          47  \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             0.033333          47  \n",
       " GO:0006120_GOBP_MITOCHONDRIAL_ELECTRON_TRANSPOR...  0.057692          47  \n",
       " GO:0006893_GOBP_GOLGI_TO_PLASMA_MEMBRANE_TRANSPORT  0.055556          47  \n",
       " GO:0016655_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  0.055556          47  \n",
       " GO:0061951_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...  0.054545          47  \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           0.029412          47  \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...  0.019763          47  \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION         0.015267          47  \n",
       " GO:0030126_GOCC_COPI_VESICLE_COAT                   0.153846          47  \n",
       " GO:0150116_GOBP_REGULATION_OF_CELL_SUBSTRATE_JU...  0.046154          47  \n",
       " GO:1990138_GOBP_NEURON_PROJECTION_EXTENSION         0.025641          47  \n",
       " GO:0010256_GOBP_ENDOMEMBRANE_SYSTEM_ORGANIZATION    0.014052          47  \n",
       " GO:0016887_GOMF_ATPASE_ACTIVITY                     0.014052          47  \n",
       " GO:0005200_GOMF_STRUCTURAL_CONSTITUENT_OF_CYTOS...  0.043478          47  \n",
       " GO:0090066_GOBP_REGULATION_OF_ANATOMICAL_STRUCT...  0.013699          47  \n",
       " GO:0015629_GOCC_ACTIN_CYTOSKELETON                  0.013544          47  ,\n",
       " 'GP2_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0032527_GOBP_PROTEIN_EXIT_FROM_ENDOPLASMIC_R...  5.100000e-14   \n",
       " GO:0072599_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...  2.700000e-13   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  3.000000e-13   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  2.500000e-12   \n",
       " GO:0035967_GOBP_CELLULAR_RESPONSE_TO_TOPOLOGICA...  4.000000e-12   \n",
       " ...                                                          ...   \n",
       " R-HSA-400511_REACTOME_SYNTHESIS_SECRETION_AND_I...  2.500000e-03   \n",
       " GO:0051656_GOBP_ESTABLISHMENT_OF_ORGANELLE_LOCA...  2.800000e-03   \n",
       " GO:0030136_GOCC_CLATHRIN_COATED_VESICLE             2.900000e-03   \n",
       " GO:2000300_GOBP_REGULATION_OF_SYNAPTIC_VESICLE_...  3.000000e-03   \n",
       " R-HSA-3000480_REACTOME_SCAVENGING_BY_CLASS_A_RE...  3.000000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0032527_GOBP_PROTEIN_EXIT_FROM_ENDOPLASMIC_R...  5.500000e-10        111   \n",
       " GO:0072599_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...  1.100000e-09        111   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  1.100000e-09        111   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  6.900000e-09        111   \n",
       " GO:0035967_GOBP_CELLULAR_RESPONSE_TO_TOPOLOGICA...  8.600000e-09        111   \n",
       " ...                                                          ...        ...   \n",
       " R-HSA-400511_REACTOME_SYNTHESIS_SECRETION_AND_I...  2.100000e-01        111   \n",
       " GO:0051656_GOBP_ESTABLISHMENT_OF_ORGANELLE_LOCA...  2.300000e-01        111   \n",
       " GO:0030136_GOCC_CLATHRIN_COATED_VESICLE             2.400000e-01        111   \n",
       " GO:2000300_GOBP_REGULATION_OF_SYNAPTIC_VESICLE_...  2.400000e-01        111   \n",
       " R-HSA-3000480_REACTOME_SCAVENGING_BY_CLASS_A_RE...  2.400000e-01        111   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0032527_GOBP_PROTEIN_EXIT_FROM_ENDOPLASMIC_R...       43       10   \n",
       " GO:0072599_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...       34        9   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...      268       17   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...      179       14   \n",
       " GO:0035967_GOBP_CELLULAR_RESPONSE_TO_TOPOLOGICA...      149       13   \n",
       " ...                                                     ...      ...   \n",
       " R-HSA-400511_REACTOME_SYNTHESIS_SECRETION_AND_I...       13        2   \n",
       " GO:0051656_GOBP_ESTABLISHMENT_OF_ORGANELLE_LOCA...      408        8   \n",
       " GO:0030136_GOCC_CLATHRIN_COATED_VESICLE                 166        5   \n",
       " GO:2000300_GOBP_REGULATION_OF_SYNAPTIC_VESICLE_...       49        3   \n",
       " R-HSA-3000480_REACTOME_SCAVENGING_BY_CLASS_A_RE...       14        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0032527_GOBP_PROTEIN_EXIT_FROM_ENDOPLASMIC_R...       18965   \n",
       " GO:0072599_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...       18965   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...       18965   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...       18965   \n",
       " GO:0035967_GOBP_CELLULAR_RESPONSE_TO_TOPOLOGICA...       18965   \n",
       " ...                                                        ...   \n",
       " R-HSA-400511_REACTOME_SYNTHESIS_SECRETION_AND_I...       18965   \n",
       " GO:0051656_GOBP_ESTABLISHMENT_OF_ORGANELLE_LOCA...       18965   \n",
       " GO:0030136_GOCC_CLATHRIN_COATED_VESICLE                  18965   \n",
       " GO:2000300_GOBP_REGULATION_OF_SYNAPTIC_VESICLE_...       18965   \n",
       " R-HSA-3000480_REACTOME_SCAVENGING_BY_CLASS_A_RE...       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0032527_GOBP_PROTEIN_EXIT_FROM_ENDOPLASMIC_R...  Derl1,Derl3,Edem1,Herpud1,Hsp90b1,Sar1a,Sel1l,...   \n",
       " GO:0072599_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...  Edem1,Herpud1,Rab3gap1,Sec63,Spcs3,Srp72,Srpr,...   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  Atp2a3,Calr,Derl1,Derl3,Edem1,Herpud1,Hsp90b1,...   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  Bag3,Calr,Derl1,Derl3,Edem1,Herpud1,Hsp90b1,Hs...   \n",
       " GO:0035967_GOBP_CELLULAR_RESPONSE_TO_TOPOLOGICA...  Bag3,Calr,Derl1,Derl3,Edem1,Herpud1,Hsp90b1,Hs...   \n",
       " ...                                                                                               ...   \n",
       " R-HSA-400511_REACTOME_SYNTHESIS_SECRETION_AND_I...                                        Pcsk1,Spcs3   \n",
       " GO:0051656_GOBP_ESTABLISHMENT_OF_ORGANELLE_LOCA...  Arfgap3,Armcx3,Lman1,Mcfd2,Myo5a,Sar1a,Uso1,Vamp7   \n",
       " GO:0030136_GOCC_CLATHRIN_COATED_VESICLE                              Btbd8,Inpp5f,Scarb2,Tgoln1,Vamp7   \n",
       " GO:2000300_GOBP_REGULATION_OF_SYNAPTIC_VESICLE_...                                 P2ry1,Rab3gap1,Syp   \n",
       " R-HSA-3000480_REACTOME_SCAVENGING_BY_CLASS_A_RE...                                       Calr,Hsp90b1   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0032527_GOBP_PROTEIN_EXIT_FROM_ENDOPLASMIC_R...  0.232558         111  \n",
       " GO:0072599_GOBP_ESTABLISHMENT_OF_PROTEIN_LOCALI...  0.264706         111  \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  0.063433         111  \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  0.078212         111  \n",
       " GO:0035967_GOBP_CELLULAR_RESPONSE_TO_TOPOLOGICA...  0.087248         111  \n",
       " ...                                                      ...         ...  \n",
       " R-HSA-400511_REACTOME_SYNTHESIS_SECRETION_AND_I...  0.153846         111  \n",
       " GO:0051656_GOBP_ESTABLISHMENT_OF_ORGANELLE_LOCA...  0.019608         111  \n",
       " GO:0030136_GOCC_CLATHRIN_COATED_VESICLE             0.030120         111  \n",
       " GO:2000300_GOBP_REGULATION_OF_SYNAPTIC_VESICLE_...  0.061224         111  \n",
       " R-HSA-3000480_REACTOME_SCAVENGING_BY_CLASS_A_RE...  0.142857         111  \n",
       " \n",
       " [134 rows x 9 columns],\n",
       " 'GP3_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0150034_GOCC_DISTAL_AXON                         3.900000e-09  0.000042   \n",
       " GO:0044297_GOCC_CELL_BODY                           1.600000e-07  0.000870   \n",
       " GO:0030427_GOCC_SITE_OF_POLARIZED_GROWTH            8.600000e-06  0.031000   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                       2.300000e-05  0.063000   \n",
       " R-HSA-381426_REACTOME_REGULATION_OF_INSULIN_LIK...  4.000000e-05  0.086000   \n",
       " GO:0005775_GOCC_VACUOLAR_LUMEN                      5.000000e-05  0.091000   \n",
       " GO:0007212_GOBP_DOPAMINE_RECEPTOR_SIGNALING_PAT...  6.700000e-05  0.100000   \n",
       " GO:0006656_GOBP_PHOSPHATIDYLCHOLINE_BIOSYNTHETI...  7.700000e-05  0.100000   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         8.300000e-05  0.100000   \n",
       " GO:0008654_GOBP_PHOSPHOLIPID_BIOSYNTHETIC_PROCESS   2.300000e-04  0.250000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0150034_GOCC_DISTAL_AXON                               228      266   \n",
       " GO:0044297_GOCC_CELL_BODY                                 228      497   \n",
       " GO:0030427_GOCC_SITE_OF_POLARIZED_GROWTH                  228      174   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                             228      268   \n",
       " R-HSA-381426_REACTOME_REGULATION_OF_INSULIN_LIK...        228      105   \n",
       " GO:0005775_GOCC_VACUOLAR_LUMEN                            228      140   \n",
       " GO:0007212_GOBP_DOPAMINE_RECEPTOR_SIGNALING_PAT...        228       36   \n",
       " GO:0006656_GOBP_PHOSPHATIDYLCHOLINE_BIOSYNTHETI...        228       37   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN               228      262   \n",
       " GO:0008654_GOBP_PHOSPHOLIPID_BIOSYNTHETIC_PROCESS         228      250   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0150034_GOCC_DISTAL_AXON                              18       18965   \n",
       " GO:0044297_GOCC_CELL_BODY                                22       18965   \n",
       " GO:0030427_GOCC_SITE_OF_POLARIZED_GROWTH                 11       18965   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                            13       18965   \n",
       " R-HSA-381426_REACTOME_REGULATION_OF_INSULIN_LIK...        8       18965   \n",
       " GO:0005775_GOCC_VACUOLAR_LUMEN                            9       18965   \n",
       " GO:0007212_GOBP_DOPAMINE_RECEPTOR_SIGNALING_PAT...        5       18965   \n",
       " GO:0006656_GOBP_PHOSPHATIDYLCHOLINE_BIOSYNTHETI...        5       18965   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN              12       18965   \n",
       " GO:0008654_GOBP_PHOSPHOLIPID_BIOSYNTHETIC_PROCESS        11       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0150034_GOCC_DISTAL_AXON                         Basp1,Calb1,Cck,Dscam,Exoc7,Fez1,Hap1,Mapt,Nrs...   \n",
       " GO:0044297_GOCC_CELL_BODY                           Alcam,Arg1,Asic2,Calb1,Cck,Cd200,Dscam,Fez1,Gr...   \n",
       " GO:0030427_GOCC_SITE_OF_POLARIZED_GROWTH            Basp1,Dscam,Exoc7,Fez1,Hap1,Mapt,Nrsn1,Olfm1,P...   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                       Arg1,Fabp5,Fam3c,Fuca1,Ggh,Gpi1,Gusb,Lgals3bp,...   \n",
       " R-HSA-381426_REACTOME_REGULATION_OF_INSULIN_LIK...     Apoa2,Ckap4,Golm1,Nucb1,Pappa2,Prss23,Sdc2,Vgf   \n",
       " GO:0005775_GOCC_VACUOLAR_LUMEN                          Arg1,Chid1,Fabp5,Fuca1,Gc,Ggh,Gusb,Prss1,Sdc2   \n",
       " GO:0007212_GOBP_DOPAMINE_RECEPTOR_SIGNALING_PAT...                         Caly,Gna14,Gnas,Gng2,Ncstn   \n",
       " GO:0006656_GOBP_PHOSPHATIDYLCHOLINE_BIOSYNTHETI...                       Ache,Apoa2,Fabp3,Fabp5,Lpin2   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         Apoa2,Ckap4,Col27a1,Erlec1,Golm1,Nucb1,Os9,Prs...   \n",
       " GO:0008654_GOBP_PHOSPHOLIPID_BIOSYNTHETIC_PROCESS   Ache,Apoa2,Fabp3,Fabp5,Lpin2,Ocrl,Pgap4,Pigk,P...   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0150034_GOCC_DISTAL_AXON                         0.067669         228  \n",
       " GO:0044297_GOCC_CELL_BODY                           0.044266         228  \n",
       " GO:0030427_GOCC_SITE_OF_POLARIZED_GROWTH            0.063218         228  \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                       0.048507         228  \n",
       " R-HSA-381426_REACTOME_REGULATION_OF_INSULIN_LIK...  0.076190         228  \n",
       " GO:0005775_GOCC_VACUOLAR_LUMEN                      0.064286         228  \n",
       " GO:0007212_GOBP_DOPAMINE_RECEPTOR_SIGNALING_PAT...  0.138889         228  \n",
       " GO:0006656_GOBP_PHOSPHATIDYLCHOLINE_BIOSYNTHETI...  0.135135         228  \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         0.045802         228  \n",
       " GO:0008654_GOBP_PHOSPHOLIPID_BIOSYNTHETIC_PROCESS   0.044000         228  ,\n",
       " 'GP4_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  5.500000e-25   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  1.000000e-16   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...  1.500000e-15   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  2.700000e-14   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX   9.100000e-14   \n",
       " ...                                                          ...   \n",
       " GO:0016859_GOMF_CIS_TRANS_ISOMERASE_ACTIVITY        5.200000e-03   \n",
       " GO:0035567_GOBP_NON_CANONICAL_WNT_SIGNALING_PAT...  5.500000e-03   \n",
       " GO:0004180_GOMF_CARBOXYPEPTIDASE_ACTIVITY           5.600000e-03   \n",
       " R-HSA-983169_REACTOME_CLASS_I_MHC_MEDIATED_ANTI...  5.600000e-03   \n",
       " GO:0008047_GOMF_ENZYME_ACTIVATOR_ACTIVITY           5.700000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  6.000000e-21        209   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  5.600000e-13        209   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...  5.500000e-12        209   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  7.400000e-11        209   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX   2.000000e-10        209   \n",
       " ...                                                          ...        ...   \n",
       " GO:0016859_GOMF_CIS_TRANS_ISOMERASE_ACTIVITY        2.200000e-01        209   \n",
       " GO:0035567_GOBP_NON_CANONICAL_WNT_SIGNALING_PAT...  2.300000e-01        209   \n",
       " GO:0004180_GOMF_CARBOXYPEPTIDASE_ACTIVITY           2.400000e-01        209   \n",
       " R-HSA-983169_REACTOME_CLASS_I_MHC_MEDIATED_ANTI...  2.400000e-01        209   \n",
       " GO:0008047_GOMF_ENZYME_ACTIVATOR_ACTIVITY           2.400000e-01        209   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...      121       25   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...      284       26   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...      122       18   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...      268       23   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX        12        8   \n",
       " ...                                                     ...      ...   \n",
       " GO:0016859_GOMF_CIS_TRANS_ISOMERASE_ACTIVITY             32        3   \n",
       " GO:0035567_GOBP_NON_CANONICAL_WNT_SIGNALING_PAT...      145        6   \n",
       " GO:0004180_GOMF_CARBOXYPEPTIDASE_ACTIVITY                33        3   \n",
       " R-HSA-983169_REACTOME_CLASS_I_MHC_MEDIATED_ANTI...      350       10   \n",
       " GO:0008047_GOMF_ENZYME_ACTIVATOR_ACTIVITY               466       12   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...       18965   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...       18965   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...       18965   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...       18965   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX        18965   \n",
       " ...                                                        ...   \n",
       " GO:0016859_GOMF_CIS_TRANS_ISOMERASE_ACTIVITY             18965   \n",
       " GO:0035567_GOBP_NON_CANONICAL_WNT_SIGNALING_PAT...       18965   \n",
       " GO:0004180_GOMF_CARBOXYPEPTIDASE_ACTIVITY                18965   \n",
       " R-HSA-983169_REACTOME_CLASS_I_MHC_MEDIATED_ANTI...       18965   \n",
       " GO:0008047_GOMF_ENZYME_ACTIVATOR_ACTIVITY                18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  Dad1,Ddost,Derl2,Emc7,H13,Hsd17b12,Krtcap2,Orm...   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  Alg5,Arf4,Bet1,Cmas,Copz2,Ctsz,Dad1,Ddost,Derl...   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...  Bcap31,Bet1,Copz2,Ctsz,Ergic3,Erp44,Kdelr1,Lrp...   \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  Bcap31,Ccnd1,Derl2,Dnajb11,Dnajb9,Dnajc3,Edem2...   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX     Dad1,Ddost,Krtcap2,Ost4,Ostc,Rpn2,Tmem258,Tusc3   \n",
       " ...                                                                                               ...   \n",
       " GO:0016859_GOMF_CIS_TRANS_ISOMERASE_ACTIVITY                                        Fkbp11,Fkbp2,Ppib   \n",
       " GO:0035567_GOBP_NON_CANONICAL_WNT_SIGNALING_PAT...                Calm1,Gpc4,Psma5,Psmb4,Psmd12,Psmd6   \n",
       " GO:0004180_GOMF_CARBOXYPEPTIDASE_ACTIVITY                                                Cpe,Cpq,Ctsz   \n",
       " R-HSA-983169_REACTOME_CLASS_I_MHC_MEDIATED_ANTI...  Bcap31,Ctsl,Pdia3,Psma5,Psmb4,Psmd12,Psmd6,Sar...   \n",
       " GO:0008047_GOMF_ENZYME_ACTIVATOR_ACTIVITY           Adprh,Arhgap36,Arl1,Atp1b1,Calm1,Cks1b,Dad1,Dd...   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  0.206612         209  \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  0.091549         209  \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...  0.147541         209  \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  0.085821         209  \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX   0.666667         209  \n",
       " ...                                                      ...         ...  \n",
       " GO:0016859_GOMF_CIS_TRANS_ISOMERASE_ACTIVITY        0.093750         209  \n",
       " GO:0035567_GOBP_NON_CANONICAL_WNT_SIGNALING_PAT...  0.041379         209  \n",
       " GO:0004180_GOMF_CARBOXYPEPTIDASE_ACTIVITY           0.090909         209  \n",
       " R-HSA-983169_REACTOME_CLASS_I_MHC_MEDIATED_ANTI...  0.028571         209  \n",
       " GO:0008047_GOMF_ENZYME_ACTIVATOR_ACTIVITY           0.025751         209  \n",
       " \n",
       " [257 rows x 9 columns],\n",
       " 'GP5_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'GP6_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'GP7_enrichment':                                                             pval     fdr  \\\n",
       " gene_set                                                                   \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...  8.600000e-07  0.0094   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  2.800000e-06  0.0150   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...  5.400000e-06  0.0200   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...  1.100000e-05  0.0310   \n",
       " R-HSA-6811434_REACTOME_COPI_DEPENDENT_GOLGI_TO_...  5.500000e-05  0.1100   \n",
       " R-HSA-190872_REACTOME_TRANSPORT_OF_CONNEXONS_TO...  6.000000e-05  0.1100   \n",
       " R-HSA-389977_REACTOME_POST_CHAPERONIN_TUBULIN_F...  1.000000e-04  0.1100   \n",
       " R-HSA-2500257_REACTOME_RESOLUTION_OF_SISTER_CHR...  1.200000e-04  0.1100   \n",
       " R-HSA-389960_REACTOME_FORMATION_OF_TUBULIN_FOLD...  1.300000e-04  0.1100   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...  1.400000e-04  0.1100   \n",
       " R-HSA-8856688_REACTOME_GOLGI_TO_ER_RETROGRADE_T...  1.400000e-04  0.1100   \n",
       " R-HSA-9619483_REACTOME_ACTIVATION_OF_AMPK_DOWNS...  1.400000e-04  0.1100   \n",
       " hsa05020_KEGG_PRION_DISEASES                        1.500000e-04  0.1100   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS  1.500000e-04  0.1100   \n",
       " R-HSA-5663220_REACTOME_RHO_GTPASES_ACTIVATE_FOR...  1.500000e-04  0.1100   \n",
       " R-HSA-9668328_REACTOME_SEALING_OF_THE_NUCLEAR_E...  1.800000e-04  0.1200   \n",
       " R-HSA-190861_REACTOME_GAP_JUNCTION_ASSEMBLY         2.000000e-04  0.1300   \n",
       " R-HSA-389958_REACTOME_COOPERATION_OF_PREFOLDIN_...  2.100000e-04  0.1300   \n",
       " R-HSA-264876_REACTOME_INSULIN_PROCESSING            2.300000e-04  0.1300   \n",
       " R-HSA-9646399_REACTOME_AGGREPHAGY                   2.900000e-04  0.1600   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...  3.900000e-04  0.2000   \n",
       " R-HSA-2467813_REACTOME_SEPARATION_OF_SISTER_CHR...  4.000000e-04  0.2000   \n",
       " R-HSA-8955332_REACTOME_CARBOXYTERMINAL_POST_TRA...  4.400000e-04  0.2000   \n",
       " R-HSA-9609736_REACTOME_ASSEMBLY_AND_CELL_SURFAC...  4.400000e-04  0.2000   \n",
       " R-HSA-437239_REACTOME_RECYCLING_PATHWAY_OF_L1       4.900000e-04  0.2000   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...  4.900000e-04  0.2000   \n",
       " R-HSA-68877_REACTOME_MITOTIC_PROMETAPHASE           5.000000e-04  0.2000   \n",
       " hsa05130_KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFEC...  5.100000e-04  0.2000   \n",
       " R-HSA-3371497_REACTOME_HSP90_CHAPERONE_CYCLE_FO...  5.900000e-04  0.2200   \n",
       " R-HSA-6811436_REACTOME_COPI_INDEPENDENT_GOLGI_T...  6.200000e-04  0.2200   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...         16       91   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...         16      284   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...         16      144   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...         16      174   \n",
       " R-HSA-6811434_REACTOME_COPI_DEPENDENT_GOLGI_TO_...         16       90   \n",
       " R-HSA-190872_REACTOME_TRANSPORT_OF_CONNEXONS_TO...         16       14   \n",
       " R-HSA-389977_REACTOME_POST_CHAPERONIN_TUBULIN_F...         16       18   \n",
       " R-HSA-2500257_REACTOME_RESOLUTION_OF_SISTER_CHR...         16      116   \n",
       " R-HSA-389960_REACTOME_FORMATION_OF_TUBULIN_FOLD...         16       20   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...         16      122   \n",
       " R-HSA-8856688_REACTOME_GOLGI_TO_ER_RETROGRADE_T...         16      122   \n",
       " R-HSA-9619483_REACTOME_ACTIVATION_OF_AMPK_DOWNS...         16       21   \n",
       " hsa05020_KEGG_PRION_DISEASES                               16       22   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS         16       22   \n",
       " R-HSA-5663220_REACTOME_RHO_GTPASES_ACTIVATE_FOR...         16      126   \n",
       " R-HSA-9668328_REACTOME_SEALING_OF_THE_NUCLEAR_E...         16       24   \n",
       " R-HSA-190861_REACTOME_GAP_JUNCTION_ASSEMBLY                16       25   \n",
       " R-HSA-389958_REACTOME_COOPERATION_OF_PREFOLDIN_...         16       26   \n",
       " R-HSA-264876_REACTOME_INSULIN_PROCESSING                   16       27   \n",
       " R-HSA-9646399_REACTOME_AGGREPHAGY                          16       30   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...         16       35   \n",
       " R-HSA-2467813_REACTOME_SEPARATION_OF_SISTER_CHR...         16      175   \n",
       " R-HSA-8955332_REACTOME_CARBOXYTERMINAL_POST_TRA...         16       37   \n",
       " R-HSA-9609736_REACTOME_ASSEMBLY_AND_CELL_SURFAC...         16       37   \n",
       " R-HSA-437239_REACTOME_RECYCLING_PATHWAY_OF_L1              16       39   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...         16      188   \n",
       " R-HSA-68877_REACTOME_MITOTIC_PROMETAPHASE                  16      190   \n",
       " hsa05130_KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFEC...         16       40   \n",
       " R-HSA-3371497_REACTOME_HSP90_CHAPERONE_CYCLE_FO...         16       43   \n",
       " R-HSA-6811436_REACTOME_COPI_INDEPENDENT_GOLGI_T...         16       44   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...        4       18965   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...        5       18965   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...        4       18965   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...        4       18965   \n",
       " R-HSA-6811434_REACTOME_COPI_DEPENDENT_GOLGI_TO_...        3       18965   \n",
       " R-HSA-190872_REACTOME_TRANSPORT_OF_CONNEXONS_TO...        2       18965   \n",
       " R-HSA-389977_REACTOME_POST_CHAPERONIN_TUBULIN_F...        2       18965   \n",
       " R-HSA-2500257_REACTOME_RESOLUTION_OF_SISTER_CHR...        3       18965   \n",
       " R-HSA-389960_REACTOME_FORMATION_OF_TUBULIN_FOLD...        2       18965   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...        3       18965   \n",
       " R-HSA-8856688_REACTOME_GOLGI_TO_ER_RETROGRADE_T...        3       18965   \n",
       " R-HSA-9619483_REACTOME_ACTIVATION_OF_AMPK_DOWNS...        2       18965   \n",
       " hsa05020_KEGG_PRION_DISEASES                              2       18965   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS        2       18965   \n",
       " R-HSA-5663220_REACTOME_RHO_GTPASES_ACTIVATE_FOR...        3       18965   \n",
       " R-HSA-9668328_REACTOME_SEALING_OF_THE_NUCLEAR_E...        2       18965   \n",
       " R-HSA-190861_REACTOME_GAP_JUNCTION_ASSEMBLY               2       18965   \n",
       " R-HSA-389958_REACTOME_COOPERATION_OF_PREFOLDIN_...        2       18965   \n",
       " R-HSA-264876_REACTOME_INSULIN_PROCESSING                  2       18965   \n",
       " R-HSA-9646399_REACTOME_AGGREPHAGY                         2       18965   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...        2       18965   \n",
       " R-HSA-2467813_REACTOME_SEPARATION_OF_SISTER_CHR...        3       18965   \n",
       " R-HSA-8955332_REACTOME_CARBOXYTERMINAL_POST_TRA...        2       18965   \n",
       " R-HSA-9609736_REACTOME_ASSEMBLY_AND_CELL_SURFAC...        2       18965   \n",
       " R-HSA-437239_REACTOME_RECYCLING_PATHWAY_OF_L1             2       18965   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...        3       18965   \n",
       " R-HSA-68877_REACTOME_MITOTIC_PROMETAPHASE                 3       18965   \n",
       " hsa05130_KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFEC...        2       18965   \n",
       " R-HSA-3371497_REACTOME_HSP90_CHAPERONE_CYCLE_FO...        2       18965   \n",
       " R-HSA-6811436_REACTOME_COPI_INDEPENDENT_GOLGI_T...        2       18965   \n",
       " \n",
       "                                                                               hits  \\\n",
       " gene_set                                                                             \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...        Ins2,Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  Gmppb,Ins2,Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...        Ins2,Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...        Ins2,Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-6811434_REACTOME_COPI_DEPENDENT_GOLGI_TO_...             Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-190872_REACTOME_TRANSPORT_OF_CONNEXONS_TO...                   Tuba1a,Tubb4b   \n",
       " R-HSA-389977_REACTOME_POST_CHAPERONIN_TUBULIN_F...                   Tuba1a,Tubb4b   \n",
       " R-HSA-2500257_REACTOME_RESOLUTION_OF_SISTER_CHR...             Spc25,Tuba1a,Tubb4b   \n",
       " R-HSA-389960_REACTOME_FORMATION_OF_TUBULIN_FOLD...                   Tuba1a,Tubb4b   \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...                Hspa5,Ins2,Tmed9   \n",
       " R-HSA-8856688_REACTOME_GOLGI_TO_ER_RETROGRADE_T...             Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-9619483_REACTOME_ACTIVATION_OF_AMPK_DOWNS...                   Tuba1a,Tubb4b   \n",
       " hsa05020_KEGG_PRION_DISEASES                                             C8b,Hspa5   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS                   Tuba1a,Tubb4b   \n",
       " R-HSA-5663220_REACTOME_RHO_GTPASES_ACTIVATE_FOR...             Spc25,Tuba1a,Tubb4b   \n",
       " R-HSA-9668328_REACTOME_SEALING_OF_THE_NUCLEAR_E...                   Tuba1a,Tubb4b   \n",
       " R-HSA-190861_REACTOME_GAP_JUNCTION_ASSEMBLY                          Tuba1a,Tubb4b   \n",
       " R-HSA-389958_REACTOME_COOPERATION_OF_PREFOLDIN_...                   Tuba1a,Tubb4b   \n",
       " R-HSA-264876_REACTOME_INSULIN_PROCESSING                                Ins2,Pcsk2   \n",
       " R-HSA-9646399_REACTOME_AGGREPHAGY                                    Tuba1a,Tubb4b   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...                   Tuba1a,Tubb4b   \n",
       " R-HSA-2467813_REACTOME_SEPARATION_OF_SISTER_CHR...             Spc25,Tuba1a,Tubb4b   \n",
       " R-HSA-8955332_REACTOME_CARBOXYTERMINAL_POST_TRA...                   Tuba1a,Tubb4b   \n",
       " R-HSA-9609736_REACTOME_ASSEMBLY_AND_CELL_SURFAC...                   Tuba1a,Tubb4b   \n",
       " R-HSA-437239_REACTOME_RECYCLING_PATHWAY_OF_L1                        Tuba1a,Tubb4b   \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...             Tmed9,Tuba1a,Tubb4b   \n",
       " R-HSA-68877_REACTOME_MITOTIC_PROMETAPHASE                      Spc25,Tuba1a,Tubb4b   \n",
       " hsa05130_KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFEC...                   Tuba1a,Tubb4b   \n",
       " R-HSA-3371497_REACTOME_HSP90_CHAPERONE_CYCLE_FO...                   Tuba1a,Tubb4b   \n",
       " R-HSA-6811436_REACTOME_COPI_INDEPENDENT_GOLGI_T...                   Tuba1a,Tubb4b   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " R-HSA-6807878_REACTOME_COPI_MEDIATED_ANTEROGRAD...  0.043956          16  \n",
       " R-HSA-446203_REACTOME_ASPARAGINE_N_LINKED_GLYCO...  0.017606          16  \n",
       " R-HSA-199977_REACTOME_ER_TO_GOLGI_ANTEROGRADE_T...  0.027778          16  \n",
       " R-HSA-948021_REACTOME_TRANSPORT_TO_THE_GOLGI_AN...  0.022989          16  \n",
       " R-HSA-6811434_REACTOME_COPI_DEPENDENT_GOLGI_TO_...  0.033333          16  \n",
       " R-HSA-190872_REACTOME_TRANSPORT_OF_CONNEXONS_TO...  0.142857          16  \n",
       " R-HSA-389977_REACTOME_POST_CHAPERONIN_TUBULIN_F...  0.111111          16  \n",
       " R-HSA-2500257_REACTOME_RESOLUTION_OF_SISTER_CHR...  0.025862          16  \n",
       " R-HSA-389960_REACTOME_FORMATION_OF_TUBULIN_FOLD...  0.100000          16  \n",
       " GO:0005793_GOCC_ENDOPLASMIC_RETICULUM_GOLGI_INT...  0.024590          16  \n",
       " R-HSA-8856688_REACTOME_GOLGI_TO_ER_RETROGRADE_T...  0.024590          16  \n",
       " R-HSA-9619483_REACTOME_ACTIVATION_OF_AMPK_DOWNS...  0.095238          16  \n",
       " hsa05020_KEGG_PRION_DISEASES                        0.090909          16  \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS  0.090909          16  \n",
       " R-HSA-5663220_REACTOME_RHO_GTPASES_ACTIVATE_FOR...  0.023810          16  \n",
       " R-HSA-9668328_REACTOME_SEALING_OF_THE_NUCLEAR_E...  0.083333          16  \n",
       " R-HSA-190861_REACTOME_GAP_JUNCTION_ASSEMBLY         0.080000          16  \n",
       " R-HSA-389958_REACTOME_COOPERATION_OF_PREFOLDIN_...  0.076923          16  \n",
       " R-HSA-264876_REACTOME_INSULIN_PROCESSING            0.074074          16  \n",
       " R-HSA-9646399_REACTOME_AGGREPHAGY                   0.066667          16  \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...  0.057143          16  \n",
       " R-HSA-2467813_REACTOME_SEPARATION_OF_SISTER_CHR...  0.017143          16  \n",
       " R-HSA-8955332_REACTOME_CARBOXYTERMINAL_POST_TRA...  0.054054          16  \n",
       " R-HSA-9609736_REACTOME_ASSEMBLY_AND_CELL_SURFAC...  0.054054          16  \n",
       " R-HSA-437239_REACTOME_RECYCLING_PATHWAY_OF_L1       0.051282          16  \n",
       " R-HSA-6811442_REACTOME_INTRA_GOLGI_AND_RETROGRA...  0.015957          16  \n",
       " R-HSA-68877_REACTOME_MITOTIC_PROMETAPHASE           0.015789          16  \n",
       " hsa05130_KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFEC...  0.050000          16  \n",
       " R-HSA-3371497_REACTOME_HSP90_CHAPERONE_CYCLE_FO...  0.046512          16  \n",
       " R-HSA-6811436_REACTOME_COPI_INDEPENDENT_GOLGI_T...  0.045455          16  ,\n",
       " 'GP8_enrichment':                                           pval     fdr  signature  geneset  \\\n",
       " gene_set                                                                     \n",
       " GO:0008430_GOMF_SELENIUM_BINDING  3.500000e-07  0.0038         42        7   \n",
       " \n",
       "                                   overlap  background               hits  \\\n",
       " gene_set                                                                   \n",
       " GO:0008430_GOMF_SELENIUM_BINDING        3       18965  Dio1,Gpx3,Selenop   \n",
       " \n",
       "                                     recall  query_size  \n",
       " gene_set                                                \n",
       " GO:0008430_GOMF_SELENIUM_BINDING  0.428571          42  ,\n",
       " 'GP9_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             4.200000e-45   \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  1.100000e-37   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           4.200000e-37   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  2.600000e-36   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  8.800000e-33   \n",
       " ...                                                          ...   \n",
       " GO:0090383_GOBP_PHAGOSOME_ACIDIFICATION             2.600000e-03   \n",
       " GO:0005759_GOCC_MITOCHONDRIAL_MATRIX                2.600000e-03   \n",
       " GO:0031970_GOCC_ORGANELLE_ENVELOPE_LUMEN            2.600000e-03   \n",
       " R-HSA-1236975_REACTOME_ANTIGEN_PROCESSING_CROSS...  2.600000e-03   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY        2.700000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             4.500000e-41         58   \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  6.100000e-34         58   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           1.500000e-33         58   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  7.100000e-33         58   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  1.900000e-29         58   \n",
       " ...                                                          ...        ...   \n",
       " GO:0090383_GOBP_PHAGOSOME_ACIDIFICATION             2.500000e-01         58   \n",
       " GO:0005759_GOCC_MITOCHONDRIAL_MATRIX                2.500000e-01         58   \n",
       " GO:0031970_GOCC_ORGANELLE_ENVELOPE_LUMEN            2.500000e-01         58   \n",
       " R-HSA-1236975_REACTOME_ANTIGEN_PROCESSING_CROSS...  2.500000e-01         58   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY        2.500000e-01         58   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                 120       27   \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...      251       28   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION               136       24   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...      121       23   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...      168       23   \n",
       " ...                                                     ...      ...   \n",
       " GO:0090383_GOBP_PHAGOSOME_ACIDIFICATION                  25        2   \n",
       " GO:0005759_GOCC_MITOCHONDRIAL_MATRIX                    456        6   \n",
       " GO:0031970_GOCC_ORGANELLE_ENVELOPE_LUMEN                 89        3   \n",
       " R-HSA-1236975_REACTOME_ANTIGEN_PROCESSING_CROSS...       90        3   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY            458        6   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                  18965   \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...       18965   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION                18965   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...       18965   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...       18965   \n",
       " ...                                                        ...   \n",
       " GO:0090383_GOBP_PHAGOSOME_ACIDIFICATION                  18965   \n",
       " GO:0005759_GOCC_MITOCHONDRIAL_MATRIX                     18965   \n",
       " GO:0031970_GOCC_ORGANELLE_ENVELOPE_LUMEN                 18965   \n",
       " R-HSA-1236975_REACTOME_ANTIGEN_PROCESSING_CROSS...       18965   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY             18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             Atp5e,Atp5k,Atp5j2,Atp5l,Atp6v0e,Atp6v1f,Cox17...   \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  Atp5e,Atp5md,Atp5k,Atp5j2,Atp5l,Atp5mpl,Bax,Co...   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           Atp5e,Atp5k,Atp5j2,Atp5l,Cox5b,Cox6a1,Cox6b1,C...   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  Atp5e,Atp5k,Atp5j2,Atp5l,Cox5b,Cox6a1,Cox6b1,C...   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  Atp5e,Atp5k,Atp5j2,Atp5l,Cox5b,Cox6a1,Cox6b1,C...   \n",
       " ...                                                                                               ...   \n",
       " GO:0090383_GOBP_PHAGOSOME_ACIDIFICATION                                               Atp6v0e,Atp6v1f   \n",
       " GO:0005759_GOCC_MITOCHONDRIAL_MATRIX                         Atp5e,Mrpl30,Mrpl33,Mrpl52,Mrps21,Ndufa7   \n",
       " GO:0031970_GOCC_ORGANELLE_ENVELOPE_LUMEN                                          Cox17,Ndufb7,Timm8b   \n",
       " R-HSA-1236975_REACTOME_ANTIGEN_PROCESSING_CROSS...                                 S100a1,Sec61g,Sem1   \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY               Mrpl30,Mrpl33,Mrpl52,Mrps21,Ndufa7,Rpl36al   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             0.225000          58  \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  0.111554          58  \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           0.176471          58  \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  0.190083          58  \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  0.136905          58  \n",
       " ...                                                      ...         ...  \n",
       " GO:0090383_GOBP_PHAGOSOME_ACIDIFICATION             0.080000          58  \n",
       " GO:0005759_GOCC_MITOCHONDRIAL_MATRIX                0.013158          58  \n",
       " GO:0031970_GOCC_ORGANELLE_ENVELOPE_LUMEN            0.033708          58  \n",
       " R-HSA-1236975_REACTOME_ANTIGEN_PROCESSING_CROSS...  0.033333          58  \n",
       " GO:0005198_GOMF_STRUCTURAL_MOLECULE_ACTIVITY        0.013100          58  \n",
       " \n",
       " [115 rows x 9 columns],\n",
       " 'GP10_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  7.900000e-40   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           5.400000e-37   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  1.800000e-36   \n",
       " GO:0098800_GOCC_INNER_MITOCHONDRIAL_MEMBRANE_PR...  1.600000e-35   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  2.000000e-34   \n",
       " ...                                                          ...   \n",
       " GO:0033218_GOMF_AMIDE_BINDING                       2.000000e-03   \n",
       " GO:0008320_GOMF_PROTEIN_TRANSMEMBRANE_TRANSPORT...  2.200000e-03   \n",
       " GO:0051540_GOMF_METAL_CLUSTER_BINDING               2.200000e-03   \n",
       " GO:0006605_GOBP_PROTEIN_TARGETING                   2.500000e-03   \n",
       " GO:0030150_GOBP_PROTEIN_IMPORT_INTO_MITOCHONDRI...  2.500000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  8.600000e-36         79   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           2.900000e-33         79   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  6.500000e-33         79   \n",
       " GO:0098800_GOCC_INNER_MITOCHONDRIAL_MEMBRANE_PR...  4.300000e-32         79   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  4.300000e-31         79   \n",
       " ...                                                          ...        ...   \n",
       " GO:0033218_GOMF_AMIDE_BINDING                       2.000000e-01         79   \n",
       " GO:0008320_GOMF_PROTEIN_TRANSMEMBRANE_TRANSPORT...  2.300000e-01         79   \n",
       " GO:0051540_GOMF_METAL_CLUSTER_BINDING               2.300000e-01         79   \n",
       " GO:0006605_GOBP_PROTEIN_TARGETING                   2.500000e-01         79   \n",
       " GO:0030150_GOBP_PROTEIN_IMPORT_INTO_MITOCHONDRI...  2.500000e-01         79   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...      251       32   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION               136       26   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...      121       25   \n",
       " GO:0098800_GOCC_INNER_MITOCHONDRIAL_MEMBRANE_PR...      131       25   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...      168       26   \n",
       " ...                                                     ...      ...   \n",
       " GO:0033218_GOMF_AMIDE_BINDING                           313        6   \n",
       " GO:0008320_GOMF_PROTEIN_TRANSMEMBRANE_TRANSPORT...       17        2   \n",
       " GO:0051540_GOMF_METAL_CLUSTER_BINDING                    62        3   \n",
       " GO:0006605_GOBP_PROTEIN_TARGETING                       329        6   \n",
       " GO:0030150_GOBP_PROTEIN_IMPORT_INTO_MITOCHONDRI...       18        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...       18965   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION                18965   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...       18965   \n",
       " GO:0098800_GOCC_INNER_MITOCHONDRIAL_MEMBRANE_PR...       18965   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...       18965   \n",
       " ...                                                        ...   \n",
       " GO:0033218_GOMF_AMIDE_BINDING                            18965   \n",
       " GO:0008320_GOMF_PROTEIN_TRANSMEMBRANE_TRANSPORT...       18965   \n",
       " GO:0051540_GOMF_METAL_CLUSTER_BINDING                    18965   \n",
       " GO:0006605_GOBP_PROTEIN_TARGETING                        18965   \n",
       " GO:0030150_GOBP_PROTEIN_IMPORT_INTO_MITOCHONDRI...       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  Atp5g1,Atp5g2,Atp5g3,Atp5h,Atp5j,Atp5o,Cox4i1,...   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           Atp5g1,Atp5g2,Atp5g3,Atp5h,Atp5j,Atp5o,Chchd2,...   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  Atp5g1,Atp5g2,Atp5g3,Atp5h,Atp5j,Atp5o,Cox14,C...   \n",
       " GO:0098800_GOCC_INNER_MITOCHONDRIAL_MEMBRANE_PR...  Atp5g1,Atp5g2,Atp5g3,Atp5h,Atp5j,Atp5o,Cox4i1,...   \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  Atp5g1,Atp5g2,Atp5g3,Atp5h,Atp5j,Atp5o,Cox14,C...   \n",
       " ...                                                                                               ...   \n",
       " GO:0033218_GOMF_AMIDE_BINDING                                     Clta,Pfdn5,Srp14,Tm2d1,Tomm20,Vdac2   \n",
       " GO:0008320_GOMF_PROTEIN_TRANSMEMBRANE_TRANSPORT...                                     Timm17a,Tomm20   \n",
       " GO:0051540_GOMF_METAL_CLUSTER_BINDING                                              Cisd1,Iscu,Uqcrfs1   \n",
       " GO:0006605_GOBP_PROTEIN_TARGETING                            Atpif1,Ndufa13,Srp14,Timm17a,Tomm20,Ubl5   \n",
       " GO:0030150_GOBP_PROTEIN_IMPORT_INTO_MITOCHONDRI...                                     Timm17a,Tomm20   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0098798_GOCC_MITOCHONDRIAL_PROTEIN_CONTAININ...  0.127490          79  \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           0.191176          79  \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  0.206612          79  \n",
       " GO:0098800_GOCC_INNER_MITOCHONDRIAL_MEMBRANE_PR...  0.190840          79  \n",
       " R-HSA-1428517_REACTOME_THE_CITRIC_ACID_TCA_CYCL...  0.154762          79  \n",
       " ...                                                      ...         ...  \n",
       " GO:0033218_GOMF_AMIDE_BINDING                       0.019169          79  \n",
       " GO:0008320_GOMF_PROTEIN_TRANSMEMBRANE_TRANSPORT...  0.117647          79  \n",
       " GO:0051540_GOMF_METAL_CLUSTER_BINDING               0.048387          79  \n",
       " GO:0006605_GOBP_PROTEIN_TARGETING                   0.018237          79  \n",
       " GO:0030150_GOBP_PROTEIN_IMPORT_INTO_MITOCHONDRI...  0.111111          79  \n",
       " \n",
       " [109 rows x 9 columns],\n",
       " 'GP11_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0005839_GOCC_PROTEASOME_CORE_COMPLEX             6.000000e-20   \n",
       " GO:0010499_GOBP_PROTEASOMAL_UBIQUITIN_INDEPENDE...  2.100000e-19   \n",
       " GO:0006521_GOBP_REGULATION_OF_CELLULAR_AMINO_AC...  8.100000e-17   \n",
       " R-HSA-162909_REACTOME_HOST_INTERACTIONS_OF_HIV_...  8.400000e-17   \n",
       " hsa03050_KEGG_PROTEASOME                            4.200000e-16   \n",
       " ...                                                          ...   \n",
       " GO:1901505_GOMF_CARBOHYDRATE_DERIVATIVE_TRANSME...  6.500000e-03   \n",
       " GO:0032387_GOBP_NEGATIVE_REGULATION_OF_INTRACEL...  6.800000e-03   \n",
       " GO:0097345_GOBP_MITOCHONDRIAL_OUTER_MEMBRANE_PE...  6.800000e-03   \n",
       " R-HSA-71406_REACTOME_PYRUVATE_METABOLISM_AND_CI...  6.800000e-03   \n",
       " GO:0006081_GOBP_CELLULAR_ALDEHYDE_METABOLIC_PRO...  7.600000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0005839_GOCC_PROTEASOME_CORE_COMPLEX             6.600000e-16         46   \n",
       " GO:0010499_GOBP_PROTEASOMAL_UBIQUITIN_INDEPENDE...  1.100000e-15         46   \n",
       " GO:0006521_GOBP_REGULATION_OF_CELLULAR_AMINO_AC...  2.300000e-13         46   \n",
       " R-HSA-162909_REACTOME_HOST_INTERACTIONS_OF_HIV_...  2.300000e-13         46   \n",
       " hsa03050_KEGG_PROTEASOME                            9.100000e-13         46   \n",
       " ...                                                          ...        ...   \n",
       " GO:1901505_GOMF_CARBOHYDRATE_DERIVATIVE_TRANSME...  2.200000e-01         46   \n",
       " GO:0032387_GOBP_NEGATIVE_REGULATION_OF_INTRACEL...  2.300000e-01         46   \n",
       " GO:0097345_GOBP_MITOCHONDRIAL_OUTER_MEMBRANE_PE...  2.300000e-01         46   \n",
       " R-HSA-71406_REACTOME_PYRUVATE_METABOLISM_AND_CI...  2.300000e-01         46   \n",
       " GO:0006081_GOBP_CELLULAR_ALDEHYDE_METABOLIC_PRO...  2.500000e-01         46   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0005839_GOCC_PROTEASOME_CORE_COMPLEX                  18        9   \n",
       " GO:0010499_GOBP_PROTEASOMAL_UBIQUITIN_INDEPENDE...       20        9   \n",
       " GO:0006521_GOBP_REGULATION_OF_CELLULAR_AMINO_AC...       56       10   \n",
       " R-HSA-162909_REACTOME_HOST_INTERACTIONS_OF_HIV_...      121       12   \n",
       " hsa03050_KEGG_PROTEASOME                                 41        9   \n",
       " ...                                                     ...      ...   \n",
       " GO:1901505_GOMF_CARBOHYDRATE_DERIVATIVE_TRANSME...       50        2   \n",
       " GO:0032387_GOBP_NEGATIVE_REGULATION_OF_INTRACEL...       51        2   \n",
       " GO:0097345_GOBP_MITOCHONDRIAL_OUTER_MEMBRANE_PE...       51        2   \n",
       " R-HSA-71406_REACTOME_PYRUVATE_METABOLISM_AND_CI...       51        2   \n",
       " GO:0006081_GOBP_CELLULAR_ALDEHYDE_METABOLIC_PRO...       54        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0005839_GOCC_PROTEASOME_CORE_COMPLEX                  18965   \n",
       " GO:0010499_GOBP_PROTEASOMAL_UBIQUITIN_INDEPENDE...       18965   \n",
       " GO:0006521_GOBP_REGULATION_OF_CELLULAR_AMINO_AC...       18965   \n",
       " R-HSA-162909_REACTOME_HOST_INTERACTIONS_OF_HIV_...       18965   \n",
       " hsa03050_KEGG_PROTEASOME                                 18965   \n",
       " ...                                                        ...   \n",
       " GO:1901505_GOMF_CARBOHYDRATE_DERIVATIVE_TRANSME...       18965   \n",
       " GO:0032387_GOBP_NEGATIVE_REGULATION_OF_INTRACEL...       18965   \n",
       " GO:0097345_GOBP_MITOCHONDRIAL_OUTER_MEMBRANE_PE...       18965   \n",
       " R-HSA-71406_REACTOME_PYRUVATE_METABOLISM_AND_CI...       18965   \n",
       " GO:0006081_GOBP_CELLULAR_ALDEHYDE_METABOLIC_PRO...       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0005839_GOCC_PROTEASOME_CORE_COMPLEX             Psma2,Psma3,Psma4,Psma6,Psma7,Psmb1,Psmb2,Psmb...   \n",
       " GO:0010499_GOBP_PROTEASOMAL_UBIQUITIN_INDEPENDE...  Psma2,Psma3,Psma4,Psma6,Psma7,Psmb1,Psmb2,Psmb...   \n",
       " GO:0006521_GOBP_REGULATION_OF_CELLULAR_AMINO_AC...  Park7,Psma2,Psma3,Psma4,Psma6,Psma7,Psmb1,Psmb...   \n",
       " R-HSA-162909_REACTOME_HOST_INTERACTIONS_OF_HIV_...  Psma2,Psma3,Psma4,Psma6,Psma7,Psmb1,Psmb2,Psmb...   \n",
       " hsa03050_KEGG_PROTEASOME                            Psma2,Psma3,Psma4,Psma6,Psma7,Psmb1,Psmb2,Psmb...   \n",
       " ...                                                                                               ...   \n",
       " GO:1901505_GOMF_CARBOHYDRATE_DERIVATIVE_TRANSME...                                    Slc25a4,Slc25a5   \n",
       " GO:0032387_GOBP_NEGATIVE_REGULATION_OF_INTRACEL...                                         Mrln,Park7   \n",
       " GO:0097345_GOBP_MITOCHONDRIAL_OUTER_MEMBRANE_PE...                                      Slc25a5,Ywhaq   \n",
       " R-HSA-71406_REACTOME_PYRUVATE_METABOLISM_AND_CI...                                          Mdh2,Sdhb   \n",
       " GO:0006081_GOBP_CELLULAR_ALDEHYDE_METABOLIC_PRO...                                       Akr1a1,Park7   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0005839_GOCC_PROTEASOME_CORE_COMPLEX             0.500000          46  \n",
       " GO:0010499_GOBP_PROTEASOMAL_UBIQUITIN_INDEPENDE...  0.450000          46  \n",
       " GO:0006521_GOBP_REGULATION_OF_CELLULAR_AMINO_AC...  0.178571          46  \n",
       " R-HSA-162909_REACTOME_HOST_INTERACTIONS_OF_HIV_...  0.099174          46  \n",
       " hsa03050_KEGG_PROTEASOME                            0.219512          46  \n",
       " ...                                                      ...         ...  \n",
       " GO:1901505_GOMF_CARBOHYDRATE_DERIVATIVE_TRANSME...  0.040000          46  \n",
       " GO:0032387_GOBP_NEGATIVE_REGULATION_OF_INTRACEL...  0.039216          46  \n",
       " GO:0097345_GOBP_MITOCHONDRIAL_OUTER_MEMBRANE_PE...  0.039216          46  \n",
       " R-HSA-71406_REACTOME_PYRUVATE_METABOLISM_AND_CI...  0.039216          46  \n",
       " GO:0006081_GOBP_CELLULAR_ALDEHYDE_METABOLIC_PRO...  0.037037          46  \n",
       " \n",
       " [325 rows x 9 columns],\n",
       " 'GP12_enrichment':                                                             pval     fdr  \\\n",
       " gene_set                                                                   \n",
       " hsa00380_KEGG_TRYPTOPHAN_METABOLISM                 1.800000e-08  0.0002   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...  2.100000e-06  0.0120   \n",
       " GO:0005249_GOMF_VOLTAGE_GATED_POTASSIUM_CHANNEL...  2.800000e-05  0.1000   \n",
       " GO:0036454_GOCC_GROWTH_FACTOR_COMPLEX               5.500000e-05  0.1500   \n",
       " GO:0006600_GOBP_CREATINE_METABOLIC_PROCESS          8.200000e-05  0.1700   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY          9.500000e-05  0.1700   \n",
       " hsa00280_KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEG...  1.300000e-04  0.1800   \n",
       " hsa00310_KEGG_LYSINE_DEGRADATION                    1.400000e-04  0.1800   \n",
       " R-HSA-71288_REACTOME_CREATINE_METABOLISM            1.500000e-04  0.1800   \n",
       " GO:0003012_GOBP_MUSCLE_SYSTEM_PROCESS               1.800000e-04  0.1800   \n",
       " GO:0022843_GOMF_VOLTAGE_GATED_CATION_CHANNEL_AC...  1.800000e-04  0.1800   \n",
       " GO:0042428_GOBP_SEROTONIN_METABOLIC_PROCESS         2.000000e-04  0.1800   \n",
       " GO:0015079_GOMF_POTASSIUM_ION_TRANSMEMBRANE_TRA...  2.200000e-04  0.1900   \n",
       " hsa00903_KEGG_LIMONENE_AND_PINENE_DEGRADATION       2.400000e-04  0.1900   \n",
       " GO:1901160_GOBP_PRIMARY_AMINO_COMPOUND_METABOLI...  3.600000e-04  0.2400   \n",
       " R-HSA-209776_REACTOME_METABOLISM_OF_AMINE_DERIV...  3.600000e-04  0.2400   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " hsa00380_KEGG_TRYPTOPHAN_METABOLISM                        45       35   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...         45      255   \n",
       " GO:0005249_GOMF_VOLTAGE_GATED_POTASSIUM_CHANNEL...         45       74   \n",
       " GO:0036454_GOCC_GROWTH_FACTOR_COMPLEX                      45        5   \n",
       " GO:0006600_GOBP_CREATINE_METABOLIC_PROCESS                 45        6   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY                 45      101   \n",
       " hsa00280_KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEG...         45       42   \n",
       " hsa00310_KEGG_LYSINE_DEGRADATION                           45       43   \n",
       " R-HSA-71288_REACTOME_CREATINE_METABOLISM                   45        8   \n",
       " GO:0003012_GOBP_MUSCLE_SYSTEM_PROCESS                      45      353   \n",
       " GO:0022843_GOMF_VOLTAGE_GATED_CATION_CHANNEL_AC...         45      119   \n",
       " GO:0042428_GOBP_SEROTONIN_METABOLIC_PROCESS                45        9   \n",
       " GO:0015079_GOMF_POTASSIUM_ION_TRANSMEMBRANE_TRA...         45      126   \n",
       " hsa00903_KEGG_LIMONENE_AND_PINENE_DEGRADATION              45       10   \n",
       " GO:1901160_GOBP_PRIMARY_AMINO_COMPOUND_METABOLI...         45       12   \n",
       " R-HSA-209776_REACTOME_METABOLISM_OF_AMINE_DERIV...         45       12   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " hsa00380_KEGG_TRYPTOPHAN_METABOLISM                       5       18965   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...        7       18965   \n",
       " GO:0005249_GOMF_VOLTAGE_GATED_POTASSIUM_CHANNEL...        4       18965   \n",
       " GO:0036454_GOCC_GROWTH_FACTOR_COMPLEX                     2       18965   \n",
       " GO:0006600_GOBP_CREATINE_METABOLIC_PROCESS                2       18965   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY                4       18965   \n",
       " hsa00280_KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEG...        3       18965   \n",
       " hsa00310_KEGG_LYSINE_DEGRADATION                          3       18965   \n",
       " R-HSA-71288_REACTOME_CREATINE_METABOLISM                  2       18965   \n",
       " GO:0003012_GOBP_MUSCLE_SYSTEM_PROCESS                     6       18965   \n",
       " GO:0022843_GOMF_VOLTAGE_GATED_CATION_CHANNEL_AC...        4       18965   \n",
       " GO:0042428_GOBP_SEROTONIN_METABOLIC_PROCESS               2       18965   \n",
       " GO:0015079_GOMF_POTASSIUM_ION_TRANSMEMBRANE_TRA...        4       18965   \n",
       " hsa00903_KEGG_LIMONENE_AND_PINENE_DEGRADATION             2       18965   \n",
       " GO:1901160_GOBP_PRIMARY_AMINO_COMPOUND_METABOLI...        2       18965   \n",
       " R-HSA-209776_REACTOME_METABOLISM_OF_AMINE_DERIV...        2       18965   \n",
       " \n",
       "                                                                                         hits  \\\n",
       " gene_set                                                                                       \n",
       " hsa00380_KEGG_TRYPTOPHAN_METABOLISM                            Aldh9a1,Ddc,Ehhadh,Ogdhl,Tph1   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...  Aldh9a1,Ddc,Gatm,Ivd,Serinc5,Slc6a8,Tph1   \n",
       " GO:0005249_GOMF_VOLTAGE_GATED_POTASSIUM_CHANNEL...                   Hcn1,Kcnj5,Kcnk9,Lrrc55   \n",
       " GO:0036454_GOCC_GROWTH_FACTOR_COMPLEX                                          Igfals,Igfbp5   \n",
       " GO:0006600_GOBP_CREATINE_METABOLIC_PROCESS                                       Gatm,Slc6a8   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY                           Hcn1,Kcnj5,Kcnk9,Lrrc55   \n",
       " hsa00280_KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEG...                        Aldh9a1,Ehhadh,Ivd   \n",
       " hsa00310_KEGG_LYSINE_DEGRADATION                                        Aldh9a1,Ehhadh,Ogdhl   \n",
       " R-HSA-71288_REACTOME_CREATINE_METABOLISM                                         Gatm,Slc6a8   \n",
       " GO:0003012_GOBP_MUSCLE_SYSTEM_PROCESS                     Gatm,Hsbp1,Igfbp5,Kcnj5,Neb,Slc6a8   \n",
       " GO:0022843_GOMF_VOLTAGE_GATED_CATION_CHANNEL_AC...                   Hcn1,Kcnj5,Kcnk9,Lrrc55   \n",
       " GO:0042428_GOBP_SEROTONIN_METABOLIC_PROCESS                                         Ddc,Tph1   \n",
       " GO:0015079_GOMF_POTASSIUM_ION_TRANSMEMBRANE_TRA...                   Hcn1,Kcnj5,Kcnk9,Lrrc55   \n",
       " hsa00903_KEGG_LIMONENE_AND_PINENE_DEGRADATION                                 Aldh9a1,Ehhadh   \n",
       " GO:1901160_GOBP_PRIMARY_AMINO_COMPOUND_METABOLI...                                  Ddc,Tph1   \n",
       " R-HSA-209776_REACTOME_METABOLISM_OF_AMINE_DERIV...                                  Ddc,Tph1   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " hsa00380_KEGG_TRYPTOPHAN_METABOLISM                 0.142857          45  \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...  0.027451          45  \n",
       " GO:0005249_GOMF_VOLTAGE_GATED_POTASSIUM_CHANNEL...  0.054054          45  \n",
       " GO:0036454_GOCC_GROWTH_FACTOR_COMPLEX               0.400000          45  \n",
       " GO:0006600_GOBP_CREATINE_METABOLIC_PROCESS          0.333333          45  \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY          0.039604          45  \n",
       " hsa00280_KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEG...  0.071429          45  \n",
       " hsa00310_KEGG_LYSINE_DEGRADATION                    0.069767          45  \n",
       " R-HSA-71288_REACTOME_CREATINE_METABOLISM            0.250000          45  \n",
       " GO:0003012_GOBP_MUSCLE_SYSTEM_PROCESS               0.016997          45  \n",
       " GO:0022843_GOMF_VOLTAGE_GATED_CATION_CHANNEL_AC...  0.033613          45  \n",
       " GO:0042428_GOBP_SEROTONIN_METABOLIC_PROCESS         0.222222          45  \n",
       " GO:0015079_GOMF_POTASSIUM_ION_TRANSMEMBRANE_TRA...  0.031746          45  \n",
       " hsa00903_KEGG_LIMONENE_AND_PINENE_DEGRADATION       0.200000          45  \n",
       " GO:1901160_GOBP_PRIMARY_AMINO_COMPOUND_METABOLI...  0.166667          45  \n",
       " R-HSA-209776_REACTOME_METABOLISM_OF_AMINE_DERIV...  0.166667          45  ,\n",
       " 'GP13_enrichment':                                                             pval     fdr  \\\n",
       " gene_set                                                                   \n",
       " GO:0098803_GOCC_RESPIRATORY_CHAIN_COMPLEX           3.300000e-07  0.0012   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    3.700000e-07  0.0012   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             4.200000e-07  0.0012   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  4.500000e-07  0.0012   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...  8.200000e-07  0.0015   \n",
       " ...                                                          ...     ...   \n",
       " GO:0051146_GOBP_STRIATED_MUSCLE_CELL_DIFFERENTI...  1.600000e-03  0.2400   \n",
       " GO:0021681_GOBP_CEREBELLAR_GRANULAR_LAYER_DEVEL...  1.700000e-03  0.2500   \n",
       " GO:0061025_GOBP_MEMBRANE_FUSION                     1.700000e-03  0.2500   \n",
       " GO:0045275_GOCC_RESPIRATORY_CHAIN_COMPLEX_III       1.700000e-03  0.2500   \n",
       " GO:0005078_GOMF_MAP_KINASE_SCAFFOLD_ACTIVITY        1.700000e-03  0.2500   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0098803_GOCC_RESPIRATORY_CHAIN_COMPLEX                 108       79   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                          108      118   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                   108      120   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...        108      121   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...        108       31   \n",
       " ...                                                       ...      ...   \n",
       " GO:0051146_GOBP_STRIATED_MUSCLE_CELL_DIFFERENTI...        108      218   \n",
       " GO:0021681_GOBP_CEREBELLAR_GRANULAR_LAYER_DEVEL...        108       11   \n",
       " GO:0061025_GOBP_MEMBRANE_FUSION                           108      150   \n",
       " GO:0045275_GOCC_RESPIRATORY_CHAIN_COMPLEX_III             108       11   \n",
       " GO:0005078_GOMF_MAP_KINASE_SCAFFOLD_ACTIVITY              108       11   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0098803_GOCC_RESPIRATORY_CHAIN_COMPLEX                 7       18965   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                          8       18965   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                   8       18965   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...        8       18965   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...        5       18965   \n",
       " ...                                                     ...         ...   \n",
       " GO:0051146_GOBP_STRIATED_MUSCLE_CELL_DIFFERENTI...        6       18965   \n",
       " GO:0021681_GOBP_CEREBELLAR_GRANULAR_LAYER_DEVEL...        2       18965   \n",
       " GO:0061025_GOBP_MEMBRANE_FUSION                           5       18965   \n",
       " GO:0045275_GOCC_RESPIRATORY_CHAIN_COMPLEX_III             2       18965   \n",
       " GO:0005078_GOMF_MAP_KINASE_SCAFFOLD_ACTIVITY              2       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0098803_GOCC_RESPIRATORY_CHAIN_COMPLEX           mt-Co1,mt-Co2,mt-Co3,mt-Cytb,mt-Nd2,mt-Nd3,mt-Nd4   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    mt-Atp6,mt-Co1,mt-Co2,mt-Co3,mt-Cytb,mt-Nd2,mt...   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             mt-Atp6,mt-Co1,mt-Co2,mt-Co3,mt-Cytb,mt-Nd2,mt...   \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  mt-Atp6,mt-Co1,mt-Co2,mt-Co3,mt-Cytb,mt-Nd2,mt...   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...                    Dusp16,Dusp5,Dusp8,Iqgap1,Tnip1   \n",
       " ...                                                                                               ...   \n",
       " GO:0051146_GOBP_STRIATED_MUSCLE_CELL_DIFFERENTI...                Actn2,Akap13,Kdm6b,Prox1,Rgs4,Vegfa   \n",
       " GO:0021681_GOBP_CEREBELLAR_GRANULAR_LAYER_DEVEL...                                        Prox1,Ttbk2   \n",
       " GO:0061025_GOBP_MEMBRANE_FUSION                                         Huwe1,Snap25,Stx11,Syt13,Syt4   \n",
       " GO:0045275_GOCC_RESPIRATORY_CHAIN_COMPLEX_III                                          mt-Co1,mt-Cytb   \n",
       " GO:0005078_GOMF_MAP_KINASE_SCAFFOLD_ACTIVITY                                            Akap13,Iqgap1   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0098803_GOCC_RESPIRATORY_CHAIN_COMPLEX           0.088608         108  \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    0.067797         108  \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             0.066667         108  \n",
       " R-HSA-163200_REACTOME_RESPIRATORY_ELECTRON_TRAN...  0.066116         108  \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...  0.161290         108  \n",
       " ...                                                      ...         ...  \n",
       " GO:0051146_GOBP_STRIATED_MUSCLE_CELL_DIFFERENTI...  0.027523         108  \n",
       " GO:0021681_GOBP_CEREBELLAR_GRANULAR_LAYER_DEVEL...  0.181818         108  \n",
       " GO:0061025_GOBP_MEMBRANE_FUSION                     0.033333         108  \n",
       " GO:0045275_GOCC_RESPIRATORY_CHAIN_COMPLEX_III       0.181818         108  \n",
       " GO:0005078_GOMF_MAP_KINASE_SCAFFOLD_ACTIVITY        0.181818         108  \n",
       " \n",
       " [74 rows x 9 columns],\n",
       " 'GP14_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " R-HSA-9648895_REACTOME_RESPONSE_OF_EIF2AK1_HRI_...  3.700000e-14   \n",
       " R-HSA-9633012_REACTOME_RESPONSE_OF_EIF2AK4_GCN2...  1.800000e-07   \n",
       " R-HSA-380994_REACTOME_ATF4_ACTIVATES_GENES_IN_R...  4.500000e-07   \n",
       " R-HSA-381042_REACTOME_PERK_REGULATES_GENE_EXPRE...  8.000000e-07   \n",
       " GO:0006520_GOBP_CELLULAR_AMINO_ACID_METABOLIC_P...  2.500000e-06   \n",
       " GO:0009069_GOBP_SERINE_FAMILY_AMINO_ACID_METABO...  3.100000e-06   \n",
       " GO:1901605_GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS  7.300000e-06   \n",
       " GO:0009064_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_MET...  1.800000e-05   \n",
       " R-HSA-9711097_REACTOME_CELLULAR_RESPONSE_TO_STA...  1.900000e-05   \n",
       " R-HSA-381119_REACTOME_UNFOLDED_PROTEIN_RESPONSE...  3.400000e-05   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...  3.900000e-05   \n",
       " GO:0019184_GOBP_NONRIBOSOMAL_PEPTIDE_BIOSYNTHET...  7.900000e-05   \n",
       " GO:0006541_GOBP_GLUTAMINE_METABOLIC_PROCESS         1.100000e-04   \n",
       " GO:0140467_GOBP_INTEGRATED_STRESS_RESPONSE_SIGN...  1.100000e-04   \n",
       " GO:0034248_GOBP_REGULATION_OF_CELLULAR_AMIDE_ME...  1.600000e-04   \n",
       " R-HSA-379716_REACTOME_CYTOSOLIC_TRNA_AMINOACYLA...  1.600000e-04   \n",
       " GO:0016874_GOMF_LIGASE_ACTIVITY                     2.100000e-04   \n",
       " GO:0140354_GOBP_LIPID_IMPORT_INTO_CELL              3.000000e-04   \n",
       " R-HSA-2408522_REACTOME_SELENOAMINO_ACID_METABOLISM  3.400000e-04   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  3.500000e-04   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE         4.000000e-04   \n",
       " R-HSA-72766_REACTOME_TRANSLATION                    4.600000e-04   \n",
       " GO:0042398_GOBP_CELLULAR_MODIFIED_AMINO_ACID_BI...  4.700000e-04   \n",
       " GO:0016875_GOMF_LIGASE_ACTIVITY_FORMING_CARBON_...  4.700000e-04   \n",
       " hsa00970_KEGG_AMINOACYL_TRNA_BIOSYNTHESIS           4.700000e-04   \n",
       " R-HSA-379724_REACTOME_TRNA_AMINOACYLATION           4.900000e-04   \n",
       " GO:0006984_GOBP_ER_NUCLEUS_SIGNALING_PATHWAY        5.900000e-04   \n",
       " GO:0009066_GOBP_ASPARTATE_FAMILY_AMINO_ACID_MET...  5.900000e-04   \n",
       " GO:0043038_GOBP_AMINO_ACID_ACTIVATION               6.700000e-04   \n",
       " GO:0006749_GOBP_GLUTATHIONE_METABOLIC_PROCESS       7.000000e-04   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " R-HSA-9648895_REACTOME_RESPONSE_OF_EIF2AK1_HRI_...  4.000000e-10         15   \n",
       " R-HSA-9633012_REACTOME_RESPONSE_OF_EIF2AK4_GCN2...  9.900000e-04         15   \n",
       " R-HSA-380994_REACTOME_ATF4_ACTIVATES_GENES_IN_R...  1.600000e-03         15   \n",
       " R-HSA-381042_REACTOME_PERK_REGULATES_GENE_EXPRE...  2.200000e-03         15   \n",
       " GO:0006520_GOBP_CELLULAR_AMINO_ACID_METABOLIC_P...  5.400000e-03         15   \n",
       " GO:0009069_GOBP_SERINE_FAMILY_AMINO_ACID_METABO...  5.600000e-03         15   \n",
       " GO:1901605_GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS  1.100000e-02         15   \n",
       " GO:0009064_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_MET...  2.200000e-02         15   \n",
       " R-HSA-9711097_REACTOME_CELLULAR_RESPONSE_TO_STA...  2.200000e-02         15   \n",
       " R-HSA-381119_REACTOME_UNFOLDED_PROTEIN_RESPONSE...  3.700000e-02         15   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...  3.800000e-02         15   \n",
       " GO:0019184_GOBP_NONRIBOSOMAL_PEPTIDE_BIOSYNTHET...  7.200000e-02         15   \n",
       " GO:0006541_GOBP_GLUTAMINE_METABOLIC_PROCESS         8.600000e-02         15   \n",
       " GO:0140467_GOBP_INTEGRATED_STRESS_RESPONSE_SIGN...  8.600000e-02         15   \n",
       " GO:0034248_GOBP_REGULATION_OF_CELLULAR_AMIDE_ME...  1.100000e-01         15   \n",
       " R-HSA-379716_REACTOME_CYTOSOLIC_TRNA_AMINOACYLA...  1.100000e-01         15   \n",
       " GO:0016874_GOMF_LIGASE_ACTIVITY                     1.400000e-01         15   \n",
       " GO:0140354_GOBP_LIPID_IMPORT_INTO_CELL              1.800000e-01         15   \n",
       " R-HSA-2408522_REACTOME_SELENOAMINO_ACID_METABOLISM  1.900000e-01         15   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  1.900000e-01         15   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE         2.000000e-01         15   \n",
       " R-HSA-72766_REACTOME_TRANSLATION                    2.000000e-01         15   \n",
       " GO:0042398_GOBP_CELLULAR_MODIFIED_AMINO_ACID_BI...  2.000000e-01         15   \n",
       " GO:0016875_GOMF_LIGASE_ACTIVITY_FORMING_CARBON_...  2.000000e-01         15   \n",
       " hsa00970_KEGG_AMINOACYL_TRNA_BIOSYNTHESIS           2.000000e-01         15   \n",
       " R-HSA-379724_REACTOME_TRNA_AMINOACYLATION           2.100000e-01         15   \n",
       " GO:0006984_GOBP_ER_NUCLEUS_SIGNALING_PATHWAY        2.300000e-01         15   \n",
       " GO:0009066_GOBP_ASPARTATE_FAMILY_AMINO_ACID_MET...  2.300000e-01         15   \n",
       " GO:0043038_GOBP_AMINO_ACID_ACTIVATION               2.500000e-01         15   \n",
       " GO:0006749_GOBP_GLUTATHIONE_METABOLIC_PROCESS       2.500000e-01         15   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " R-HSA-9648895_REACTOME_RESPONSE_OF_EIF2AK1_HRI_...       10        5   \n",
       " R-HSA-9633012_REACTOME_RESPONSE_OF_EIF2AK4_GCN2...       15        3   \n",
       " R-HSA-380994_REACTOME_ATF4_ACTIVATES_GENES_IN_R...       20        3   \n",
       " R-HSA-381042_REACTOME_PERK_REGULATES_GENE_EXPRE...       24        3   \n",
       " GO:0006520_GOBP_CELLULAR_AMINO_ACID_METABOLIC_P...      299        5   \n",
       " GO:0009069_GOBP_SERINE_FAMILY_AMINO_ACID_METABO...       37        3   \n",
       " GO:1901605_GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS      167        4   \n",
       " GO:0009064_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_MET...       66        3   \n",
       " R-HSA-9711097_REACTOME_CELLULAR_RESPONSE_TO_STA...       67        3   \n",
       " R-HSA-381119_REACTOME_UNFOLDED_PROTEIN_RESPONSE...       82        3   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...      255        4   \n",
       " GO:0019184_GOBP_NONRIBOSOMAL_PEPTIDE_BIOSYNTHET...       17        2   \n",
       " GO:0006541_GOBP_GLUTAMINE_METABOLIC_PROCESS              20        2   \n",
       " GO:0140467_GOBP_INTEGRATED_STRESS_RESPONSE_SIGN...       20        2   \n",
       " GO:0034248_GOBP_REGULATION_OF_CELLULAR_AMIDE_ME...      365        4   \n",
       " R-HSA-379716_REACTOME_CYTOSOLIC_TRNA_AMINOACYLA...       24        2   \n",
       " GO:0016874_GOMF_LIGASE_ACTIVITY                         152        3   \n",
       " GO:0140354_GOBP_LIPID_IMPORT_INTO_CELL                   33        2   \n",
       " R-HSA-2408522_REACTOME_SELENOAMINO_ACID_METABOLISM       35        2   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...      179        3   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE             188        3   \n",
       " R-HSA-72766_REACTOME_TRANSLATION                        198        3   \n",
       " GO:0042398_GOBP_CELLULAR_MODIFIED_AMINO_ACID_BI...       41        2   \n",
       " GO:0016875_GOMF_LIGASE_ACTIVITY_FORMING_CARBON_...       41        2   \n",
       " hsa00970_KEGG_AMINOACYL_TRNA_BIOSYNTHESIS                41        2   \n",
       " R-HSA-379724_REACTOME_TRNA_AMINOACYLATION                42        2   \n",
       " GO:0006984_GOBP_ER_NUCLEUS_SIGNALING_PATHWAY             46        2   \n",
       " GO:0009066_GOBP_ASPARTATE_FAMILY_AMINO_ACID_MET...       46        2   \n",
       " GO:0043038_GOBP_AMINO_ACID_ACTIVATION                    49        2   \n",
       " GO:0006749_GOBP_GLUTATHIONE_METABOLIC_PROCESS            50        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " R-HSA-9648895_REACTOME_RESPONSE_OF_EIF2AK1_HRI_...       18965   \n",
       " R-HSA-9633012_REACTOME_RESPONSE_OF_EIF2AK4_GCN2...       18965   \n",
       " R-HSA-380994_REACTOME_ATF4_ACTIVATES_GENES_IN_R...       18965   \n",
       " R-HSA-381042_REACTOME_PERK_REGULATES_GENE_EXPRE...       18965   \n",
       " GO:0006520_GOBP_CELLULAR_AMINO_ACID_METABOLIC_P...       18965   \n",
       " GO:0009069_GOBP_SERINE_FAMILY_AMINO_ACID_METABO...       18965   \n",
       " GO:1901605_GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS       18965   \n",
       " GO:0009064_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_MET...       18965   \n",
       " R-HSA-9711097_REACTOME_CELLULAR_RESPONSE_TO_STA...       18965   \n",
       " R-HSA-381119_REACTOME_UNFOLDED_PROTEIN_RESPONSE...       18965   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...       18965   \n",
       " GO:0019184_GOBP_NONRIBOSOMAL_PEPTIDE_BIOSYNTHET...       18965   \n",
       " GO:0006541_GOBP_GLUTAMINE_METABOLIC_PROCESS              18965   \n",
       " GO:0140467_GOBP_INTEGRATED_STRESS_RESPONSE_SIGN...       18965   \n",
       " GO:0034248_GOBP_REGULATION_OF_CELLULAR_AMIDE_ME...       18965   \n",
       " R-HSA-379716_REACTOME_CYTOSOLIC_TRNA_AMINOACYLA...       18965   \n",
       " GO:0016874_GOMF_LIGASE_ACTIVITY                          18965   \n",
       " GO:0140354_GOBP_LIPID_IMPORT_INTO_CELL                   18965   \n",
       " R-HSA-2408522_REACTOME_SELENOAMINO_ACID_METABOLISM       18965   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...       18965   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE              18965   \n",
       " R-HSA-72766_REACTOME_TRANSLATION                         18965   \n",
       " GO:0042398_GOBP_CELLULAR_MODIFIED_AMINO_ACID_BI...       18965   \n",
       " GO:0016875_GOMF_LIGASE_ACTIVITY_FORMING_CARBON_...       18965   \n",
       " hsa00970_KEGG_AMINOACYL_TRNA_BIOSYNTHESIS                18965   \n",
       " R-HSA-379724_REACTOME_TRNA_AMINOACYLATION                18965   \n",
       " GO:0006984_GOBP_ER_NUCLEUS_SIGNALING_PATHWAY             18965   \n",
       " GO:0009066_GOBP_ASPARTATE_FAMILY_AMINO_ACID_MET...       18965   \n",
       " GO:0043038_GOBP_AMINO_ACID_ACTIVATION                    18965   \n",
       " GO:0006749_GOBP_GLUTATHIONE_METABOLIC_PROCESS            18965   \n",
       " \n",
       "                                                                             hits  \\\n",
       " gene_set                                                                           \n",
       " R-HSA-9648895_REACTOME_RESPONSE_OF_EIF2AK1_HRI_...   Asns,Atf5,Cebpg,Chac1,Ddit3   \n",
       " R-HSA-9633012_REACTOME_RESPONSE_OF_EIF2AK4_GCN2...              Asns,Cebpg,Ddit3   \n",
       " R-HSA-380994_REACTOME_ATF4_ACTIVATES_GENES_IN_R...              Asns,Cebpg,Ddit3   \n",
       " R-HSA-381042_REACTOME_PERK_REGULATES_GENE_EXPRE...              Asns,Cebpg,Ddit3   \n",
       " GO:0006520_GOBP_CELLULAR_AMINO_ACID_METABOLIC_P...  Asns,Eprs,Phgdh,Sars,Slc7a11   \n",
       " GO:0009069_GOBP_SERINE_FAMILY_AMINO_ACID_METABO...            Phgdh,Sars,Slc7a11   \n",
       " GO:1901605_GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS       Asns,Phgdh,Sars,Slc7a11   \n",
       " GO:0009064_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_MET...            Asns,Phgdh,Slc7a11   \n",
       " R-HSA-9711097_REACTOME_CELLULAR_RESPONSE_TO_STA...              Asns,Cebpg,Ddit3   \n",
       " R-HSA-381119_REACTOME_UNFOLDED_PROTEIN_RESPONSE...              Asns,Cebpg,Ddit3   \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...          Asns,Eprs,Phgdh,Sars   \n",
       " GO:0019184_GOBP_NONRIBOSOMAL_PEPTIDE_BIOSYNTHET...                 Chac1,Slc7a11   \n",
       " GO:0006541_GOBP_GLUTAMINE_METABOLIC_PROCESS                           Asns,Phgdh   \n",
       " GO:0140467_GOBP_INTEGRATED_STRESS_RESPONSE_SIGN...                    Asns,Ddit3   \n",
       " GO:0034248_GOBP_REGULATION_OF_CELLULAR_AMIDE_ME...        Ang,Eif3c,Eprs,Slc7a11   \n",
       " R-HSA-379716_REACTOME_CYTOSOLIC_TRNA_AMINOACYLA...                     Eprs,Sars   \n",
       " GO:0016874_GOMF_LIGASE_ACTIVITY                                   Asns,Eprs,Sars   \n",
       " GO:0140354_GOBP_LIPID_IMPORT_INTO_CELL                              Eprs,Slc7a11   \n",
       " R-HSA-2408522_REACTOME_SELENOAMINO_ACID_METABOLISM                     Eprs,Sars   \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...              Asns,Chac1,Ddit3   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE                    Asns,Chka,Slc7a11   \n",
       " R-HSA-72766_REACTOME_TRANSLATION                                 Eif3c,Eprs,Sars   \n",
       " GO:0042398_GOBP_CELLULAR_MODIFIED_AMINO_ACID_BI...                 Chac1,Slc7a11   \n",
       " GO:0016875_GOMF_LIGASE_ACTIVITY_FORMING_CARBON_...                     Eprs,Sars   \n",
       " hsa00970_KEGG_AMINOACYL_TRNA_BIOSYNTHESIS                              Eprs,Sars   \n",
       " R-HSA-379724_REACTOME_TRNA_AMINOACYLATION                              Eprs,Sars   \n",
       " GO:0006984_GOBP_ER_NUCLEUS_SIGNALING_PATHWAY                          Asns,Ddit3   \n",
       " GO:0009066_GOBP_ASPARTATE_FAMILY_AMINO_ACID_MET...                    Asns,Phgdh   \n",
       " GO:0043038_GOBP_AMINO_ACID_ACTIVATION                                  Eprs,Sars   \n",
       " GO:0006749_GOBP_GLUTATHIONE_METABOLIC_PROCESS                      Chac1,Slc7a11   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " R-HSA-9648895_REACTOME_RESPONSE_OF_EIF2AK1_HRI_...  0.500000          15  \n",
       " R-HSA-9633012_REACTOME_RESPONSE_OF_EIF2AK4_GCN2...  0.200000          15  \n",
       " R-HSA-380994_REACTOME_ATF4_ACTIVATES_GENES_IN_R...  0.150000          15  \n",
       " R-HSA-381042_REACTOME_PERK_REGULATES_GENE_EXPRE...  0.125000          15  \n",
       " GO:0006520_GOBP_CELLULAR_AMINO_ACID_METABOLIC_P...  0.016722          15  \n",
       " GO:0009069_GOBP_SERINE_FAMILY_AMINO_ACID_METABO...  0.081081          15  \n",
       " GO:1901605_GOBP_ALPHA_AMINO_ACID_METABOLIC_PROCESS  0.023952          15  \n",
       " GO:0009064_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_MET...  0.045455          15  \n",
       " R-HSA-9711097_REACTOME_CELLULAR_RESPONSE_TO_STA...  0.044776          15  \n",
       " R-HSA-381119_REACTOME_UNFOLDED_PROTEIN_RESPONSE...  0.036585          15  \n",
       " R-HSA-71291_REACTOME_METABOLISM_OF_AMINO_ACIDS_...  0.015686          15  \n",
       " GO:0019184_GOBP_NONRIBOSOMAL_PEPTIDE_BIOSYNTHET...  0.117647          15  \n",
       " GO:0006541_GOBP_GLUTAMINE_METABOLIC_PROCESS         0.100000          15  \n",
       " GO:0140467_GOBP_INTEGRATED_STRESS_RESPONSE_SIGN...  0.100000          15  \n",
       " GO:0034248_GOBP_REGULATION_OF_CELLULAR_AMIDE_ME...  0.010959          15  \n",
       " R-HSA-379716_REACTOME_CYTOSOLIC_TRNA_AMINOACYLA...  0.083333          15  \n",
       " GO:0016874_GOMF_LIGASE_ACTIVITY                     0.019737          15  \n",
       " GO:0140354_GOBP_LIPID_IMPORT_INTO_CELL              0.060606          15  \n",
       " R-HSA-2408522_REACTOME_SELENOAMINO_ACID_METABOLISM  0.057143          15  \n",
       " GO:0035966_GOBP_RESPONSE_TO_TOPOLOGICALLY_INCOR...  0.016760          15  \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE         0.015957          15  \n",
       " R-HSA-72766_REACTOME_TRANSLATION                    0.015152          15  \n",
       " GO:0042398_GOBP_CELLULAR_MODIFIED_AMINO_ACID_BI...  0.048780          15  \n",
       " GO:0016875_GOMF_LIGASE_ACTIVITY_FORMING_CARBON_...  0.048780          15  \n",
       " hsa00970_KEGG_AMINOACYL_TRNA_BIOSYNTHESIS           0.048780          15  \n",
       " R-HSA-379724_REACTOME_TRNA_AMINOACYLATION           0.047619          15  \n",
       " GO:0006984_GOBP_ER_NUCLEUS_SIGNALING_PATHWAY        0.043478          15  \n",
       " GO:0009066_GOBP_ASPARTATE_FAMILY_AMINO_ACID_MET...  0.043478          15  \n",
       " GO:0043038_GOBP_AMINO_ACID_ACTIVATION               0.040816          15  \n",
       " GO:0006749_GOBP_GLUTATHIONE_METABOLIC_PROCESS       0.040000          15  ,\n",
       " 'GP15_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0030522_GOBP_INTRACELLULAR_RECEPTOR_SIGNALIN...  6.700000e-09  0.000058   \n",
       " R-HSA-400253_REACTOME_CIRCADIAN_CLOCK               1.400000e-08  0.000058   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                    1.600000e-08  0.000058   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM      2.100000e-07  0.000460   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                    2.100000e-07  0.000460   \n",
       " ...                                                          ...       ...   \n",
       " GO:0010469_GOBP_REGULATION_OF_SIGNALING_RECEPTO...  6.300000e-03  0.250000   \n",
       " GO:0021536_GOBP_DIENCEPHALON_DEVELOPMENT            6.300000e-03  0.250000   \n",
       " GO:0034762_GOBP_REGULATION_OF_TRANSMEMBRANE_TRA...  6.400000e-03  0.250000   \n",
       " GO:0043409_GOBP_NEGATIVE_REGULATION_OF_MAPK_CAS...  6.400000e-03  0.250000   \n",
       " GO:0044057_GOBP_REGULATION_OF_SYSTEM_PROCESS        6.400000e-03  0.250000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0030522_GOBP_INTRACELLULAR_RECEPTOR_SIGNALIN...         47      241   \n",
       " R-HSA-400253_REACTOME_CIRCADIAN_CLOCK                      47       66   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                           47      185   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM             47      104   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                           47      258   \n",
       " ...                                                       ...      ...   \n",
       " GO:0010469_GOBP_REGULATION_OF_SIGNALING_RECEPTO...         47      152   \n",
       " GO:0021536_GOBP_DIENCEPHALON_DEVELOPMENT                   47       48   \n",
       " GO:0034762_GOBP_REGULATION_OF_TRANSMEMBRANE_TRA...         47      478   \n",
       " GO:0043409_GOBP_NEGATIVE_REGULATION_OF_MAPK_CAS...         47      153   \n",
       " GO:0044057_GOBP_REGULATION_OF_SYSTEM_PROCESS               47      478   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0030522_GOBP_INTRACELLULAR_RECEPTOR_SIGNALIN...        9       18965   \n",
       " R-HSA-400253_REACTOME_CIRCADIAN_CLOCK                     6       18965   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                          8       18965   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM            6       18965   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                          8       18965   \n",
       " ...                                                     ...         ...   \n",
       " GO:0010469_GOBP_REGULATION_OF_SIGNALING_RECEPTO...        3       18965   \n",
       " GO:0021536_GOBP_DIENCEPHALON_DEVELOPMENT                  2       18965   \n",
       " GO:0034762_GOBP_REGULATION_OF_TRANSMEMBRANE_TRA...        5       18965   \n",
       " GO:0043409_GOBP_NEGATIVE_REGULATION_OF_MAPK_CAS...        3       18965   \n",
       " GO:0044057_GOBP_REGULATION_OF_SYSTEM_PROCESS              5       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0030522_GOBP_INTRACELLULAR_RECEPTOR_SIGNALIN...  Cry2,Hspa1b,Isl1,Klf2,Nr1d1,Nr4a1,Nr4a2,Per1,T...   \n",
       " R-HSA-400253_REACTOME_CIRCADIAN_CLOCK                             Cry2,Nfil3,Nr1d1,Per1,Ppargc1a,Sik1   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                        Cry2,Klf9,Nfil3,Nr1d1,Per1,Ppargc1a,Sik1,Usp2   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM                     Cry2,Nr1d1,Per1,Ppargc1a,Sik1,Usp2   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                        Cry2,Klf9,Nfil3,Nr1d1,Per1,Ppargc1a,Sik1,Usp2   \n",
       " ...                                                                                               ...   \n",
       " GO:0010469_GOBP_REGULATION_OF_SIGNALING_RECEPTO...                                Adra2a,Arc,Ppargc1a   \n",
       " GO:0021536_GOBP_DIENCEPHALON_DEVELOPMENT                                                   Isl1,Nr4a2   \n",
       " GO:0034762_GOBP_REGULATION_OF_TRANSMEMBRANE_TRA...                       Adra2a,Arc,Gem,Irs2,Ppargc1a   \n",
       " GO:0043409_GOBP_NEGATIVE_REGULATION_OF_MAPK_CAS...                                   Dusp18,Klf4,Per1   \n",
       " GO:0044057_GOBP_REGULATION_OF_SYSTEM_PROCESS                            Adra2a,Isl1,Klf4,Ppargc1a,Tnr   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0030522_GOBP_INTRACELLULAR_RECEPTOR_SIGNALIN...  0.037344          47  \n",
       " R-HSA-400253_REACTOME_CIRCADIAN_CLOCK               0.090909          47  \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                    0.043243          47  \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM      0.057692          47  \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                    0.031008          47  \n",
       " ...                                                      ...         ...  \n",
       " GO:0010469_GOBP_REGULATION_OF_SIGNALING_RECEPTO...  0.019737          47  \n",
       " GO:0021536_GOBP_DIENCEPHALON_DEVELOPMENT            0.041667          47  \n",
       " GO:0034762_GOBP_REGULATION_OF_TRANSMEMBRANE_TRA...  0.010460          47  \n",
       " GO:0043409_GOBP_NEGATIVE_REGULATION_OF_MAPK_CAS...  0.019608          47  \n",
       " GO:0044057_GOBP_REGULATION_OF_SYSTEM_PROCESS        0.010460          47  \n",
       " \n",
       " [275 rows x 9 columns],\n",
       " 'GP16_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS         5.000000e-09  0.000054   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS      1.900000e-08  0.000100   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...  3.900000e-08  0.000110   \n",
       " GO:0016125_GOBP_STEROL_METABOLIC_PROCESS            4.100000e-08  0.000110   \n",
       " GO:0061061_GOBP_MUSCLE_STRUCTURE_DEVELOPMENT        1.200000e-07  0.000270   \n",
       " ...                                                          ...       ...   \n",
       " GO:0033233_GOBP_REGULATION_OF_PROTEIN_SUMOYLATION   2.500000e-03  0.250000   \n",
       " GO:0099515_GOBP_ACTIN_FILAMENT_BASED_TRANSPORT      2.500000e-03  0.250000   \n",
       " GO:0030139_GOCC_ENDOCYTIC_VESICLE                   2.500000e-03  0.250000   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION         2.600000e-03  0.250000   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  2.600000e-03  0.250000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS                67       70   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS             67       24   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...         67       55   \n",
       " GO:0016125_GOBP_STEROL_METABOLIC_PROCESS                   67      144   \n",
       " GO:0061061_GOBP_MUSCLE_STRUCTURE_DEVELOPMENT               67      489   \n",
       " ...                                                       ...      ...   \n",
       " GO:0033233_GOBP_REGULATION_OF_PROTEIN_SUMOYLATION          67       21   \n",
       " GO:0099515_GOBP_ACTIN_FILAMENT_BASED_TRANSPORT             67       21   \n",
       " GO:0030139_GOCC_ENDOCYTIC_VESICLE                          67      266   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION                67      393   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...         67       77   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS               7       18965   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS            5       18965   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...        6       18965   \n",
       " GO:0016125_GOBP_STEROL_METABOLIC_PROCESS                  8       18965   \n",
       " GO:0061061_GOBP_MUSCLE_STRUCTURE_DEVELOPMENT             12       18965   \n",
       " ...                                                     ...         ...   \n",
       " GO:0033233_GOBP_REGULATION_OF_PROTEIN_SUMOYLATION         2       18965   \n",
       " GO:0099515_GOBP_ACTIN_FILAMENT_BASED_TRANSPORT            2       18965   \n",
       " GO:0030139_GOCC_ENDOCYTIC_VESICLE                         5       18965   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION               6       18965   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...        3       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS               Cyp51,Hmgcr,Hmgcs1,Idi1,Insig1,Msmo1,Srebf2   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS                          Cyp51,Hmgcr,Hmgcs1,Idi1,Msmo1   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...              Cyp51,Hmgcr,Hmgcs1,Idi1,Insig1,Srebf2   \n",
       " GO:0016125_GOBP_STEROL_METABOLIC_PROCESS             Cyp51,Hmgcr,Hmgcs1,Idi1,Insig1,Ldlr,Msmo1,Srebf2   \n",
       " GO:0061061_GOBP_MUSCLE_STRUCTURE_DEVELOPMENT        Cfl2,Ctnnb1,Emd,Gpx1,Hbegf,Ifrd1,Maff,Mapk14,N...   \n",
       " ...                                                                                               ...   \n",
       " GO:0033233_GOBP_REGULATION_OF_PROTEIN_SUMOYLATION                                         Ctnnb1,Rela   \n",
       " GO:0099515_GOBP_ACTIN_FILAMENT_BASED_TRANSPORT                                            Actn4,Myo1e   \n",
       " GO:0030139_GOCC_ENDOCYTIC_VESICLE                                        Hbegf,Ldlr,Myo1e,Ocln,Stxbp1   \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION                       Cfl2,Coro1c,Myo1e,Shroom3,Tpm4,Wdr1   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...                                   Ctnnb1,Prr7,Rela   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS         0.100000          67  \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS      0.208333          67  \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...  0.109091          67  \n",
       " GO:0016125_GOBP_STEROL_METABOLIC_PROCESS            0.055556          67  \n",
       " GO:0061061_GOBP_MUSCLE_STRUCTURE_DEVELOPMENT        0.024540          67  \n",
       " ...                                                      ...         ...  \n",
       " GO:0033233_GOBP_REGULATION_OF_PROTEIN_SUMOYLATION   0.095238          67  \n",
       " GO:0099515_GOBP_ACTIN_FILAMENT_BASED_TRANSPORT      0.095238          67  \n",
       " GO:0030139_GOCC_ENDOCYTIC_VESICLE                   0.018797          67  \n",
       " GO:0007015_GOBP_ACTIN_FILAMENT_ORGANIZATION         0.015267          67  \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  0.038961          67  \n",
       " \n",
       " [112 rows x 9 columns],\n",
       " 'GP17_enrichment':                                                         pval    fdr  \\\n",
       " gene_set                                                              \n",
       " GO:0038066_GOBP_P38MAPK_CASCADE                     0.000003  0.032   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...  0.000052  0.250   \n",
       " GO:1903753_GOBP_NEGATIVE_REGULATION_OF_P38MAPK_...  0.000068  0.250   \n",
       " GO:0045333_GOBP_CELLULAR_RESPIRATION                0.000093  0.250   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0038066_GOBP_P38MAPK_CASCADE                            50       38   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...         50      253   \n",
       " GO:1903753_GOBP_NEGATIVE_REGULATION_OF_P38MAPK_...         50        5   \n",
       " GO:0045333_GOBP_CELLULAR_RESPIRATION                       50      174   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0038066_GOBP_P38MAPK_CASCADE                           4       18965   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...        6       18965   \n",
       " GO:1903753_GOBP_NEGATIVE_REGULATION_OF_P38MAPK_...        2       18965   \n",
       " GO:0045333_GOBP_CELLULAR_RESPIRATION                      5       18965   \n",
       " \n",
       "                                                                                         hits  \\\n",
       " gene_set                                                                                       \n",
       " GO:0038066_GOBP_P38MAPK_CASCADE                                   Dusp10,Ezr,Gadd45b,Gadd45g   \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...  Coq10b,Cox6a2,Ndufs2,Nr4a3,Per2,Slc25a25   \n",
       " GO:1903753_GOBP_NEGATIVE_REGULATION_OF_P38MAPK_...                                Dusp10,Ezr   \n",
       " GO:0045333_GOBP_CELLULAR_RESPIRATION                     Coq10b,Cox6a2,Ndufs2,Nr4a3,Slc25a25   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0038066_GOBP_P38MAPK_CASCADE                     0.105263          50  \n",
       " GO:0015980_GOBP_ENERGY_DERIVATION_BY_OXIDATION_...  0.023715          50  \n",
       " GO:1903753_GOBP_NEGATIVE_REGULATION_OF_P38MAPK_...  0.400000          50  \n",
       " GO:0045333_GOBP_CELLULAR_RESPIRATION                0.028736          50  ,\n",
       " 'GP18_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'GP19_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   5.000000e-12   \n",
       " GO:0030072_GOBP_PEPTIDE_HORMONE_SECRETION           3.000000e-11   \n",
       " GO:0090276_GOBP_REGULATION_OF_PEPTIDE_HORMONE_S...  1.200000e-10   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   3.100000e-10   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     7.800000e-10   \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      2.100000e-09   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                   3.500000e-09   \n",
       " GO:0002791_GOBP_REGULATION_OF_PEPTIDE_SECRETION     3.700000e-09   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        2.900000e-08   \n",
       " GO:0034284_GOBP_RESPONSE_TO_MONOSACCHARIDE          6.200000e-07   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE            1.400000e-06   \n",
       " GO:0002793_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...  2.300000e-06   \n",
       " GO:0090277_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...  4.500000e-06   \n",
       " GO:0046887_GOBP_POSITIVE_REGULATION_OF_HORMONE_...  1.600000e-05   \n",
       " R-HSA-210745_REACTOME_REGULATION_OF_GENE_EXPRES...  2.800000e-05   \n",
       " hsa04950_KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG  3.700000e-05   \n",
       " GO:0032024_GOBP_POSITIVE_REGULATION_OF_INSULIN_...  3.900000e-05   \n",
       " GO:0051047_GOBP_POSITIVE_REGULATION_OF_SECRETION    1.000000e-04   \n",
       " R-HSA-186712_REACTOME_REGULATION_OF_BETA_CELL_D...  2.000000e-04   \n",
       " GO:0003857_GOMF_3_HYDROXYACYL_COA_DEHYDROGENASE...  2.200000e-04   \n",
       " GO:0033500_GOBP_CARBOHYDRATE_HOMEOSTASIS            2.500000e-04   \n",
       " GO:0016614_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  2.700000e-04   \n",
       " GO:0071322_GOBP_CELLULAR_RESPONSE_TO_CARBOHYDRA...  3.900000e-04   \n",
       " GO:0001678_GOBP_CELLULAR_GLUCOSE_HOMEOSTASIS        5.000000e-04   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   5.500000e-08         54   \n",
       " GO:0030072_GOBP_PEPTIDE_HORMONE_SECRETION           1.600000e-07         54   \n",
       " GO:0090276_GOBP_REGULATION_OF_PEPTIDE_HORMONE_S...  4.300000e-07         54   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   8.500000e-07         54   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     1.700000e-06         54   \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      3.800000e-06         54   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                   5.100000e-06         54   \n",
       " GO:0002791_GOBP_REGULATION_OF_PEPTIDE_SECRETION     5.100000e-06         54   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        3.500000e-05         54   \n",
       " GO:0034284_GOBP_RESPONSE_TO_MONOSACCHARIDE          6.800000e-04         54   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE            1.400000e-03         54   \n",
       " GO:0002793_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...  2.100000e-03         54   \n",
       " GO:0090277_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...  3.800000e-03         54   \n",
       " GO:0046887_GOBP_POSITIVE_REGULATION_OF_HORMONE_...  1.200000e-02         54   \n",
       " R-HSA-210745_REACTOME_REGULATION_OF_GENE_EXPRES...  2.000000e-02         54   \n",
       " hsa04950_KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG  2.500000e-02         54   \n",
       " GO:0032024_GOBP_POSITIVE_REGULATION_OF_INSULIN_...  2.500000e-02         54   \n",
       " GO:0051047_GOBP_POSITIVE_REGULATION_OF_SECRETION    6.100000e-02         54   \n",
       " R-HSA-186712_REACTOME_REGULATION_OF_BETA_CELL_D...  1.100000e-01         54   \n",
       " GO:0003857_GOMF_3_HYDROXYACYL_COA_DEHYDROGENASE...  1.200000e-01         54   \n",
       " GO:0033500_GOBP_CARBOHYDRATE_HOMEOSTASIS            1.300000e-01         54   \n",
       " GO:0016614_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  1.300000e-01         54   \n",
       " GO:0071322_GOBP_CELLULAR_RESPONSE_TO_CARBOHYDRA...  1.800000e-01         54   \n",
       " GO:0001678_GOBP_CELLULAR_GLUCOSE_HOMEOSTASIS        2.300000e-01         54   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                       190       11   \n",
       " GO:0030072_GOBP_PEPTIDE_HORMONE_SECRETION               224       11   \n",
       " GO:0090276_GOBP_REGULATION_OF_PEPTIDE_HORMONE_S...      188       10   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                       279       11   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION         228       10   \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                          426       12   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                       351       11   \n",
       " GO:0002791_GOBP_REGULATION_OF_PEPTIDE_SECRETION         268       10   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS            430       11   \n",
       " GO:0034284_GOBP_RESPONSE_TO_MONOSACCHARIDE              175        7   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE                198        7   \n",
       " GO:0002793_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...      135        6   \n",
       " GO:0090277_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...       86        5   \n",
       " GO:0046887_GOBP_POSITIVE_REGULATION_OF_HORMONE_...      111        5   \n",
       " R-HSA-210745_REACTOME_REGULATION_OF_GENE_EXPRES...       21        3   \n",
       " hsa04950_KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG       23        3   \n",
       " GO:0032024_GOBP_POSITIVE_REGULATION_OF_INSULIN_...       67        4   \n",
       " GO:0051047_GOBP_POSITIVE_REGULATION_OF_SECRETION        264        6   \n",
       " R-HSA-186712_REACTOME_REGULATION_OF_BETA_CELL_D...       40        3   \n",
       " GO:0003857_GOMF_3_HYDROXYACYL_COA_DEHYDROGENASE...        8        2   \n",
       " GO:0033500_GOBP_CARBOHYDRATE_HOMEOSTASIS                198        5   \n",
       " GO:0016614_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...      110        4   \n",
       " GO:0071322_GOBP_CELLULAR_RESPONSE_TO_CARBOHYDRA...      121        4   \n",
       " GO:0001678_GOBP_CELLULAR_GLUCOSE_HOMEOSTASIS            129        4   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                        18965   \n",
       " GO:0030072_GOBP_PEPTIDE_HORMONE_SECRETION                18965   \n",
       " GO:0090276_GOBP_REGULATION_OF_PEPTIDE_HORMONE_S...       18965   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                        18965   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION          18965   \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                           18965   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                        18965   \n",
       " GO:0002791_GOBP_REGULATION_OF_PEPTIDE_SECRETION          18965   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS             18965   \n",
       " GO:0034284_GOBP_RESPONSE_TO_MONOSACCHARIDE               18965   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE                 18965   \n",
       " GO:0002793_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...       18965   \n",
       " GO:0090277_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...       18965   \n",
       " GO:0046887_GOBP_POSITIVE_REGULATION_OF_HORMONE_...       18965   \n",
       " R-HSA-210745_REACTOME_REGULATION_OF_GENE_EXPRES...       18965   \n",
       " hsa04950_KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG       18965   \n",
       " GO:0032024_GOBP_POSITIVE_REGULATION_OF_INSULIN_...       18965   \n",
       " GO:0051047_GOBP_POSITIVE_REGULATION_OF_SECRETION         18965   \n",
       " R-HSA-186712_REACTOME_REGULATION_OF_BETA_CELL_D...       18965   \n",
       " GO:0003857_GOMF_3_HYDROXYACYL_COA_DEHYDROGENASE...       18965   \n",
       " GO:0033500_GOBP_CARBOHYDRATE_HOMEOSTASIS                 18965   \n",
       " GO:0016614_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...       18965   \n",
       " GO:0071322_GOBP_CELLULAR_RESPONSE_TO_CARBOHYDRA...       18965   \n",
       " GO:0001678_GOBP_CELLULAR_GLUCOSE_HOMEOSTASIS             18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   Ffar1,Fkbp1b,Hadh,Mafa,Neurod1,Oxct1,Selenot,S...   \n",
       " GO:0030072_GOBP_PEPTIDE_HORMONE_SECRETION           Ffar1,Fkbp1b,Hadh,Mafa,Neurod1,Oxct1,Selenot,S...   \n",
       " GO:0090276_GOBP_REGULATION_OF_PEPTIDE_HORMONE_S...  Ffar1,Fkbp1b,Hadh,Neurod1,Oxct1,Selenot,Slc2a2...   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   Ffar1,Fkbp1b,Hadh,Mafa,Neurod1,Oxct1,Selenot,S...   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     Ffar1,Fkbp1b,Hadh,Neurod1,Oxct1,Selenot,Slc2a2...   \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      Ffar1,Fkbp1b,Hadh,Mafa,Maob,Neurod1,Oxct1,Sele...   \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                   Ffar1,Fkbp1b,Hadh,Mafa,Neurod1,Oxct1,Selenot,S...   \n",
       " GO:0002791_GOBP_REGULATION_OF_PEPTIDE_SECRETION     Ffar1,Fkbp1b,Hadh,Neurod1,Oxct1,Selenot,Slc2a2...   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        Ffar1,Fkbp1b,Hadh,Mafa,Neurod1,Oxct1,Selenot,S...   \n",
       " GO:0034284_GOBP_RESPONSE_TO_MONOSACCHARIDE             Fkbp1b,Mafa,Neurod1,Oxct1,Selenot,Slc30a8,Ucn3   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE               Fkbp1b,Mafa,Neurod1,Oxct1,Selenot,Slc30a8,Ucn3   \n",
       " GO:0002793_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...             Ffar1,Oxct1,Selenot,Slc30a8,Sytl4,Ucn3   \n",
       " GO:0090277_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...                   Ffar1,Oxct1,Selenot,Slc30a8,Ucn3   \n",
       " GO:0046887_GOBP_POSITIVE_REGULATION_OF_HORMONE_...                   Ffar1,Oxct1,Selenot,Slc30a8,Ucn3   \n",
       " R-HSA-210745_REACTOME_REGULATION_OF_GENE_EXPRES...                                Mafa,Neurod1,Slc2a2   \n",
       " hsa04950_KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG                                Mafa,Neurod1,Slc2a2   \n",
       " GO:0032024_GOBP_POSITIVE_REGULATION_OF_INSULIN_...                           Ffar1,Oxct1,Slc30a8,Ucn3   \n",
       " GO:0051047_GOBP_POSITIVE_REGULATION_OF_SECRETION               Ffar1,Oxct1,Selenot,Slc30a8,Sytl4,Ucn3   \n",
       " R-HSA-186712_REACTOME_REGULATION_OF_BETA_CELL_D...                                Mafa,Neurod1,Slc2a2   \n",
       " GO:0003857_GOMF_3_HYDROXYACYL_COA_DEHYDROGENASE...                                         Cryl1,Hadh   \n",
       " GO:0033500_GOBP_CARBOHYDRATE_HOMEOSTASIS                           Ffar1,Fkbp1b,Neurod1,Oxct1,Selenot   \n",
       " GO:0016614_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...                              Cryl1,Dhrs4,Gpd2,Hadh   \n",
       " GO:0071322_GOBP_CELLULAR_RESPONSE_TO_CARBOHYDRA...                       Fkbp1b,Neurod1,Oxct1,Selenot   \n",
       " GO:0001678_GOBP_CELLULAR_GLUCOSE_HOMEOSTASIS                             Fkbp1b,Neurod1,Oxct1,Selenot   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   0.057895          54  \n",
       " GO:0030072_GOBP_PEPTIDE_HORMONE_SECRETION           0.049107          54  \n",
       " GO:0090276_GOBP_REGULATION_OF_PEPTIDE_HORMONE_S...  0.053191          54  \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   0.039427          54  \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     0.043860          54  \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      0.028169          54  \n",
       " GO:0002790_GOBP_PEPTIDE_SECRETION                   0.031339          54  \n",
       " GO:0002791_GOBP_REGULATION_OF_PEPTIDE_SECRETION     0.037313          54  \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        0.025581          54  \n",
       " GO:0034284_GOBP_RESPONSE_TO_MONOSACCHARIDE          0.040000          54  \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE            0.035354          54  \n",
       " GO:0002793_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...  0.044444          54  \n",
       " GO:0090277_GOBP_POSITIVE_REGULATION_OF_PEPTIDE_...  0.058140          54  \n",
       " GO:0046887_GOBP_POSITIVE_REGULATION_OF_HORMONE_...  0.045045          54  \n",
       " R-HSA-210745_REACTOME_REGULATION_OF_GENE_EXPRES...  0.142857          54  \n",
       " hsa04950_KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG  0.130435          54  \n",
       " GO:0032024_GOBP_POSITIVE_REGULATION_OF_INSULIN_...  0.059701          54  \n",
       " GO:0051047_GOBP_POSITIVE_REGULATION_OF_SECRETION    0.022727          54  \n",
       " R-HSA-186712_REACTOME_REGULATION_OF_BETA_CELL_D...  0.075000          54  \n",
       " GO:0003857_GOMF_3_HYDROXYACYL_COA_DEHYDROGENASE...  0.250000          54  \n",
       " GO:0033500_GOBP_CARBOHYDRATE_HOMEOSTASIS            0.025253          54  \n",
       " GO:0016614_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  0.036364          54  \n",
       " GO:0071322_GOBP_CELLULAR_RESPONSE_TO_CARBOHYDRA...  0.033058          54  \n",
       " GO:0001678_GOBP_CELLULAR_GLUCOSE_HOMEOSTASIS        0.031008          54  ,\n",
       " 'GP20_enrichment':                                                         pval   fdr  signature  \\\n",
       " gene_set                                                                        \n",
       " R-HSA-982772_REACTOME_GROWTH_HORMONE_RECEPTOR_S...  0.000040  0.18         61   \n",
       " GO:0098901_GOBP_REGULATION_OF_CARDIAC_MUSCLE_CE...  0.000047  0.18         61   \n",
       " GO:0038111_GOBP_INTERLEUKIN_7_MEDIATED_SIGNALIN...  0.000088  0.18         61   \n",
       " GO:1901652_GOBP_RESPONSE_TO_PEPTIDE                 0.000098  0.18         61   \n",
       " GO:1901653_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE        0.000110  0.18         61   \n",
       " GO:1903115_GOBP_REGULATION_OF_ACTIN_FILAMENT_BA...  0.000130  0.18         61   \n",
       " R-HSA-1266695_REACTOME_INTERLEUKIN_7_SIGNALING      0.000130  0.18         61   \n",
       " GO:0016198_GOBP_AXON_CHOICE_POINT_RECOGNITION       0.000150  0.18         61   \n",
       " GO:0033292_GOBP_T_TUBULE_ORGANIZATION               0.000150  0.18         61   \n",
       " GO:0032483_GOBP_REGULATION_OF_RAB_PROTEIN_SIGNA...  0.000210  0.20         61   \n",
       " GO:0098760_GOBP_RESPONSE_TO_INTERLEUKIN_7           0.000210  0.20         61   \n",
       " GO:0043434_GOBP_RESPONSE_TO_PEPTIDE_HORMONE         0.000220  0.20         61   \n",
       " GO:0005126_GOMF_CYTOKINE_RECEPTOR_BINDING           0.000300  0.25         61   \n",
       " GO:0071375_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE_HO...  0.000320  0.25         61   \n",
       " GO:0098900_GOBP_REGULATION_OF_ACTION_POTENTIAL      0.000380  0.25         61   \n",
       " GO:0070757_GOBP_INTERLEUKIN_35_MEDIATED_SIGNALI...  0.000450  0.25         61   \n",
       " GO:2000574_GOBP_REGULATION_OF_MICROTUBULE_MOTOR...  0.000450  0.25         61   \n",
       " R-HSA-1059683_REACTOME_INTERLEUKIN_6_SIGNALING      0.000450  0.25         61   \n",
       " GO:0060048_GOBP_CARDIAC_MUSCLE_CONTRACTION          0.000460  0.25         61   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY            0.000480  0.25         61   \n",
       " GO:0045747_GOBP_POSITIVE_REGULATION_OF_NOTCH_SI...  0.000490  0.25         61   \n",
       " GO:0070106_GOBP_INTERLEUKIN_27_MEDIATED_SIGNALI...  0.000550  0.25         61   \n",
       " R-HSA-8984722_REACTOME_INTERLEUKIN_35_SIGNALLING    0.000550  0.25         61   \n",
       " R-HSA-9020956_REACTOME_INTERLEUKIN_27_SIGNALING     0.000550  0.25         61   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " R-HSA-982772_REACTOME_GROWTH_HORMONE_RECEPTOR_S...       21        3   \n",
       " GO:0098901_GOBP_REGULATION_OF_CARDIAC_MUSCLE_CE...       22        3   \n",
       " GO:0038111_GOBP_INTERLEUKIN_7_MEDIATED_SIGNALIN...       27        3   \n",
       " GO:1901652_GOBP_RESPONSE_TO_PEPTIDE                     454        8   \n",
       " GO:1901653_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE            344        7   \n",
       " GO:1903115_GOBP_REGULATION_OF_ACTIN_FILAMENT_BA...       31        3   \n",
       " R-HSA-1266695_REACTOME_INTERLEUKIN_7_SIGNALING           31        3   \n",
       " GO:0016198_GOBP_AXON_CHOICE_POINT_RECOGNITION             6        2   \n",
       " GO:0033292_GOBP_T_TUBULE_ORGANIZATION                     6        2   \n",
       " GO:0032483_GOBP_REGULATION_OF_RAB_PROTEIN_SIGNA...        7        2   \n",
       " GO:0098760_GOBP_RESPONSE_TO_INTERLEUKIN_7                36        3   \n",
       " GO:0043434_GOBP_RESPONSE_TO_PEPTIDE_HORMONE             382        7   \n",
       " GO:0005126_GOMF_CYTOKINE_RECEPTOR_BINDING               183        5   \n",
       " GO:0071375_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE_HO...      288        6   \n",
       " GO:0098900_GOBP_REGULATION_OF_ACTION_POTENTIAL           44        3   \n",
       " GO:0070757_GOBP_INTERLEUKIN_35_MEDIATED_SIGNALI...       10        2   \n",
       " GO:2000574_GOBP_REGULATION_OF_MICROTUBULE_MOTOR...       10        2   \n",
       " R-HSA-1059683_REACTOME_INTERLEUKIN_6_SIGNALING           10        2   \n",
       " GO:0060048_GOBP_CARDIAC_MUSCLE_CONTRACTION              112        4   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY                113        4   \n",
       " GO:0045747_GOBP_POSITIVE_REGULATION_OF_NOTCH_SI...       48        3   \n",
       " GO:0070106_GOBP_INTERLEUKIN_27_MEDIATED_SIGNALI...       11        2   \n",
       " R-HSA-8984722_REACTOME_INTERLEUKIN_35_SIGNALLING         11        2   \n",
       " R-HSA-9020956_REACTOME_INTERLEUKIN_27_SIGNALING          11        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " R-HSA-982772_REACTOME_GROWTH_HORMONE_RECEPTOR_S...       18965   \n",
       " GO:0098901_GOBP_REGULATION_OF_CARDIAC_MUSCLE_CE...       18965   \n",
       " GO:0038111_GOBP_INTERLEUKIN_7_MEDIATED_SIGNALIN...       18965   \n",
       " GO:1901652_GOBP_RESPONSE_TO_PEPTIDE                      18965   \n",
       " GO:1901653_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE             18965   \n",
       " GO:1903115_GOBP_REGULATION_OF_ACTIN_FILAMENT_BA...       18965   \n",
       " R-HSA-1266695_REACTOME_INTERLEUKIN_7_SIGNALING           18965   \n",
       " GO:0016198_GOBP_AXON_CHOICE_POINT_RECOGNITION            18965   \n",
       " GO:0033292_GOBP_T_TUBULE_ORGANIZATION                    18965   \n",
       " GO:0032483_GOBP_REGULATION_OF_RAB_PROTEIN_SIGNA...       18965   \n",
       " GO:0098760_GOBP_RESPONSE_TO_INTERLEUKIN_7                18965   \n",
       " GO:0043434_GOBP_RESPONSE_TO_PEPTIDE_HORMONE              18965   \n",
       " GO:0005126_GOMF_CYTOKINE_RECEPTOR_BINDING                18965   \n",
       " GO:0071375_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE_HO...       18965   \n",
       " GO:0098900_GOBP_REGULATION_OF_ACTION_POTENTIAL           18965   \n",
       " GO:0070757_GOBP_INTERLEUKIN_35_MEDIATED_SIGNALI...       18965   \n",
       " GO:2000574_GOBP_REGULATION_OF_MICROTUBULE_MOTOR...       18965   \n",
       " R-HSA-1059683_REACTOME_INTERLEUKIN_6_SIGNALING           18965   \n",
       " GO:0060048_GOBP_CARDIAC_MUSCLE_CONTRACTION               18965   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY                 18965   \n",
       " GO:0045747_GOBP_POSITIVE_REGULATION_OF_NOTCH_SI...       18965   \n",
       " GO:0070106_GOBP_INTERLEUKIN_27_MEDIATED_SIGNALI...       18965   \n",
       " R-HSA-8984722_REACTOME_INTERLEUKIN_35_SIGNALLING         18965   \n",
       " R-HSA-9020956_REACTOME_INTERLEUKIN_27_SIGNALING          18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " R-HSA-982772_REACTOME_GROWTH_HORMONE_RECEPTOR_S...                                   Cish,Socs2,Stat3   \n",
       " GO:0098901_GOBP_REGULATION_OF_CARDIAC_MUSCLE_CE...                                    Ank2,Atp2a2,Dsp   \n",
       " GO:0038111_GOBP_INTERLEUKIN_7_MEDIATED_SIGNALIN...                                   Cish,Socs2,Stat3   \n",
       " GO:1901652_GOBP_RESPONSE_TO_PEPTIDE                 Acvr1c,Apobec1,App,Cpeb1,Dennd4c,Insrr,Socs2,S...   \n",
       " GO:1901653_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE              Apobec1,App,Cpeb1,Dennd4c,Insrr,Socs2,Stat3   \n",
       " GO:1903115_GOBP_REGULATION_OF_ACTIN_FILAMENT_BA...                                    Ank2,Atp2a2,Dsp   \n",
       " R-HSA-1266695_REACTOME_INTERLEUKIN_7_SIGNALING                                       Cish,Socs2,Stat3   \n",
       " GO:0016198_GOBP_AXON_CHOICE_POINT_RECOGNITION                                               App,Robo2   \n",
       " GO:0033292_GOBP_T_TUBULE_ORGANIZATION                                                     Ank2,Atp2a2   \n",
       " GO:0032483_GOBP_REGULATION_OF_RAB_PROTEIN_SIGNA...                                       Dennd4c,Madd   \n",
       " GO:0098760_GOBP_RESPONSE_TO_INTERLEUKIN_7                                            Cish,Socs2,Stat3   \n",
       " GO:0043434_GOBP_RESPONSE_TO_PEPTIDE_HORMONE            Acvr1c,Apobec1,Cpeb1,Dennd4c,Insrr,Socs2,Stat3   \n",
       " GO:0005126_GOMF_CYTOKINE_RECEPTOR_BINDING                               Il6st,Madd,Socs2,Trim37,Vegfb   \n",
       " GO:0071375_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE_HO...            Apobec1,Cpeb1,Dennd4c,Insrr,Socs2,Stat3   \n",
       " GO:0098900_GOBP_REGULATION_OF_ACTION_POTENTIAL                                        Ank2,Atp2a2,Dsp   \n",
       " GO:0070757_GOBP_INTERLEUKIN_35_MEDIATED_SIGNALI...                                        Il6st,Stat3   \n",
       " GO:2000574_GOBP_REGULATION_OF_MICROTUBULE_MOTOR...                                        Dynll2,Map2   \n",
       " R-HSA-1059683_REACTOME_INTERLEUKIN_6_SIGNALING                                            Il6st,Stat3   \n",
       " GO:0060048_GOBP_CARDIAC_MUSCLE_CONTRACTION                                      Ank2,Atp2a2,Dsp,Vegfb   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY                                       Cish,Il6st,Socs2,Stat3   \n",
       " GO:0045747_GOBP_POSITIVE_REGULATION_OF_NOTCH_SI...                                  Il6st,Robo2,Stat3   \n",
       " GO:0070106_GOBP_INTERLEUKIN_27_MEDIATED_SIGNALI...                                        Il6st,Stat3   \n",
       " R-HSA-8984722_REACTOME_INTERLEUKIN_35_SIGNALLING                                          Il6st,Stat3   \n",
       " R-HSA-9020956_REACTOME_INTERLEUKIN_27_SIGNALING                                           Il6st,Stat3   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " R-HSA-982772_REACTOME_GROWTH_HORMONE_RECEPTOR_S...  0.142857          61  \n",
       " GO:0098901_GOBP_REGULATION_OF_CARDIAC_MUSCLE_CE...  0.136364          61  \n",
       " GO:0038111_GOBP_INTERLEUKIN_7_MEDIATED_SIGNALIN...  0.111111          61  \n",
       " GO:1901652_GOBP_RESPONSE_TO_PEPTIDE                 0.017621          61  \n",
       " GO:1901653_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE        0.020349          61  \n",
       " GO:1903115_GOBP_REGULATION_OF_ACTIN_FILAMENT_BA...  0.096774          61  \n",
       " R-HSA-1266695_REACTOME_INTERLEUKIN_7_SIGNALING      0.096774          61  \n",
       " GO:0016198_GOBP_AXON_CHOICE_POINT_RECOGNITION       0.333333          61  \n",
       " GO:0033292_GOBP_T_TUBULE_ORGANIZATION               0.333333          61  \n",
       " GO:0032483_GOBP_REGULATION_OF_RAB_PROTEIN_SIGNA...  0.285714          61  \n",
       " GO:0098760_GOBP_RESPONSE_TO_INTERLEUKIN_7           0.083333          61  \n",
       " GO:0043434_GOBP_RESPONSE_TO_PEPTIDE_HORMONE         0.018325          61  \n",
       " GO:0005126_GOMF_CYTOKINE_RECEPTOR_BINDING           0.027322          61  \n",
       " GO:0071375_GOBP_CELLULAR_RESPONSE_TO_PEPTIDE_HO...  0.020833          61  \n",
       " GO:0098900_GOBP_REGULATION_OF_ACTION_POTENTIAL      0.068182          61  \n",
       " GO:0070757_GOBP_INTERLEUKIN_35_MEDIATED_SIGNALI...  0.200000          61  \n",
       " GO:2000574_GOBP_REGULATION_OF_MICROTUBULE_MOTOR...  0.200000          61  \n",
       " R-HSA-1059683_REACTOME_INTERLEUKIN_6_SIGNALING      0.200000          61  \n",
       " GO:0060048_GOBP_CARDIAC_MUSCLE_CONTRACTION          0.035714          61  \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY            0.035398          61  \n",
       " GO:0045747_GOBP_POSITIVE_REGULATION_OF_NOTCH_SI...  0.062500          61  \n",
       " GO:0070106_GOBP_INTERLEUKIN_27_MEDIATED_SIGNALI...  0.181818          61  \n",
       " R-HSA-8984722_REACTOME_INTERLEUKIN_35_SIGNALLING    0.181818          61  \n",
       " R-HSA-9020956_REACTOME_INTERLEUKIN_27_SIGNALING     0.181818          61  ,\n",
       " 'GP21_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'GP22_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'GP23_enrichment':                                                         pval    fdr  \\\n",
       " gene_set                                                              \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION       0.000010  0.054   \n",
       " GO:0045540_GOBP_REGULATION_OF_CHOLESTEROL_BIOSY...  0.000015  0.054   \n",
       " R-HSA-2426168_REACTOME_ACTIVATION_OF_GENE_EXPRE...  0.000018  0.054   \n",
       " GO:0071280_GOBP_CELLULAR_RESPONSE_TO_COPPER_ION     0.000025  0.054   \n",
       " GO:0005911_GOCC_CELL_CELL_JUNCTION                  0.000025  0.054   \n",
       " GO:0046890_GOBP_REGULATION_OF_LIPID_BIOSYNTHETI...  0.000033  0.059   \n",
       " GO:0090181_GOBP_REGULATION_OF_CHOLESTEROL_METAB...  0.000038  0.059   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...  0.000048  0.065   \n",
       " GO:0045296_GOMF_CADHERIN_BINDING                    0.000084  0.096   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS      0.000088  0.096   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...  0.000098  0.097   \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS         0.000120  0.100   \n",
       " GO:0036017_GOBP_RESPONSE_TO_ERYTHROPOIETIN          0.000130  0.100   \n",
       " GO:0045338_GOBP_FARNESYL_DIPHOSPHATE_METABOLIC_...  0.000130  0.100   \n",
       " GO:0046688_GOBP_RESPONSE_TO_COPPER_ION              0.000170  0.120   \n",
       " GO:0050810_GOBP_REGULATION_OF_STEROID_BIOSYNTHE...  0.000170  0.120   \n",
       " R-HSA-446728_REACTOME_CELL_JUNCTION_ORGANIZATION    0.000190  0.120   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS       0.000190  0.120   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...  0.000270  0.160   \n",
       " GO:0050839_GOMF_CELL_ADHESION_MOLECULE_BINDING      0.000310  0.170   \n",
       " GO:0010043_GOBP_RESPONSE_TO_ZINC_ION                0.000330  0.170   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION        0.000360  0.180   \n",
       " GO:0098632_GOMF_CELL_CELL_ADHESION_MEDIATOR_ACT...  0.000480  0.230   \n",
       " GO:0019218_GOBP_REGULATION_OF_STEROID_METABOLIC...  0.000500  0.230   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                 0.000580  0.240   \n",
       " R-HSA-73621_REACTOME_PYRIMIDINE_CATABOLISM          0.000580  0.240   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION              69       12   \n",
       " GO:0045540_GOBP_REGULATION_OF_CHOLESTEROL_BIOSY...         69       41   \n",
       " R-HSA-2426168_REACTOME_ACTIVATION_OF_GENE_EXPRE...         69       43   \n",
       " GO:0071280_GOBP_CELLULAR_RESPONSE_TO_COPPER_ION            69       16   \n",
       " GO:0005911_GOCC_CELL_CELL_JUNCTION                         69      430   \n",
       " GO:0046890_GOBP_REGULATION_OF_LIPID_BIOSYNTHETI...         69      168   \n",
       " GO:0090181_GOBP_REGULATION_OF_CHOLESTEROL_METAB...         69       52   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...         69       55   \n",
       " GO:0045296_GOMF_CADHERIN_BINDING                           69      288   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS             69       24   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...         69       66   \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS                69       70   \n",
       " GO:0036017_GOBP_RESPONSE_TO_ERYTHROPOIETIN                 69        5   \n",
       " GO:0045338_GOBP_FARNESYL_DIPHOSPHATE_METABOLIC_...         69        5   \n",
       " GO:0046688_GOBP_RESPONSE_TO_COPPER_ION                     69       30   \n",
       " GO:0050810_GOBP_REGULATION_OF_STEROID_BIOSYNTHE...         69       76   \n",
       " R-HSA-446728_REACTOME_CELL_JUNCTION_ORGANIZATION           69       78   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS              69        6   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...         69        7   \n",
       " GO:0050839_GOMF_CELL_ADHESION_MOLECULE_BINDING             69      472   \n",
       " GO:0010043_GOBP_RESPONSE_TO_ZINC_ION                       69       37   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION               69        8   \n",
       " GO:0098632_GOMF_CELL_CELL_ADHESION_MEDIATOR_ACT...         69       42   \n",
       " GO:0019218_GOBP_REGULATION_OF_STEROID_METABOLIC...         69      101   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                        69       10   \n",
       " R-HSA-73621_REACTOME_PYRIMIDINE_CATABOLISM                 69       10   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION             3       18965   \n",
       " GO:0045540_GOBP_REGULATION_OF_CHOLESTEROL_BIOSY...        4       18965   \n",
       " R-HSA-2426168_REACTOME_ACTIVATION_OF_GENE_EXPRE...        4       18965   \n",
       " GO:0071280_GOBP_CELLULAR_RESPONSE_TO_COPPER_ION           3       18965   \n",
       " GO:0005911_GOCC_CELL_CELL_JUNCTION                        9       18965   \n",
       " GO:0046890_GOBP_REGULATION_OF_LIPID_BIOSYNTHETI...        6       18965   \n",
       " GO:0090181_GOBP_REGULATION_OF_CHOLESTEROL_METAB...        4       18965   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...        4       18965   \n",
       " GO:0045296_GOMF_CADHERIN_BINDING                          7       18965   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS            3       18965   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...        4       18965   \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS               4       18965   \n",
       " GO:0036017_GOBP_RESPONSE_TO_ERYTHROPOIETIN                2       18965   \n",
       " GO:0045338_GOBP_FARNESYL_DIPHOSPHATE_METABOLIC_...        2       18965   \n",
       " GO:0046688_GOBP_RESPONSE_TO_COPPER_ION                    3       18965   \n",
       " GO:0050810_GOBP_REGULATION_OF_STEROID_BIOSYNTHE...        4       18965   \n",
       " R-HSA-446728_REACTOME_CELL_JUNCTION_ORGANIZATION          4       18965   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS             2       18965   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...        2       18965   \n",
       " GO:0050839_GOMF_CELL_ADHESION_MOLECULE_BINDING            8       18965   \n",
       " GO:0010043_GOBP_RESPONSE_TO_ZINC_ION                      3       18965   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION              2       18965   \n",
       " GO:0098632_GOMF_CELL_CELL_ADHESION_MEDIATOR_ACT...        3       18965   \n",
       " GO:0019218_GOBP_REGULATION_OF_STEROID_METABOLIC...        4       18965   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                       2       18965   \n",
       " R-HSA-73621_REACTOME_PYRIMIDINE_CATABOLISM                2       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION                                           Kcnk3,Mt1,Mt2   \n",
       " GO:0045540_GOBP_REGULATION_OF_CHOLESTEROL_BIOSY...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " R-HSA-2426168_REACTOME_ACTIVATION_OF_GENE_EXPRE...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " GO:0071280_GOBP_CELLULAR_RESPONSE_TO_COPPER_ION                                      Map1lc3a,Mt1,Mt2   \n",
       " GO:0005911_GOCC_CELL_CELL_JUNCTION                  Baiap2,Cldn7,Cnn3,F11r,Gjd2,Jup,Pdlim7,Vasp,Ywhah   \n",
       " GO:0046890_GOBP_REGULATION_OF_LIPID_BIOSYNTHETI...                Fdft1,Fdps,Pdgfa,Pmvk,Scd2,Tnfrsf1a   \n",
       " GO:0090181_GOBP_REGULATION_OF_CHOLESTEROL_METAB...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " GO:0045296_GOMF_CADHERIN_BINDING                                Baiap2,Cnn3,F11r,Jup,Rtn4,Slc3a2,Vasp   \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS                                        Fdft1,Fdps,Pmvk   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS                                      Fdft1,Fdps,Pmvk,Scd2   \n",
       " GO:0036017_GOBP_RESPONSE_TO_ERYTHROPOIETIN                                                    Mt1,Mt2   \n",
       " GO:0045338_GOBP_FARNESYL_DIPHOSPHATE_METABOLIC_...                                         Fdft1,Fdps   \n",
       " GO:0046688_GOBP_RESPONSE_TO_COPPER_ION                                               Map1lc3a,Mt1,Mt2   \n",
       " GO:0050810_GOBP_REGULATION_OF_STEROID_BIOSYNTHE...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " R-HSA-446728_REACTOME_CELL_JUNCTION_ORGANIZATION                                  Cldn7,F11r,Jup,Vasp   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS                                                 Mt1,Mt2   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...                                            Mt1,Mt2   \n",
       " GO:0050839_GOMF_CELL_ADHESION_MOLECULE_BINDING            Baiap2,Cldn7,Cnn3,F11r,Jup,Rtn4,Slc3a2,Vasp   \n",
       " GO:0010043_GOBP_RESPONSE_TO_ZINC_ION                                                    Kcnk3,Mt1,Mt2   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION                                                  Mt1,Mt2   \n",
       " GO:0098632_GOMF_CELL_CELL_ADHESION_MEDIATOR_ACT...                                    Baiap2,Cnn3,Jup   \n",
       " GO:0019218_GOBP_REGULATION_OF_STEROID_METABOLIC...                               Fdft1,Fdps,Pmvk,Scd2   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                                                           Mt1,Mt2   \n",
       " R-HSA-73621_REACTOME_PYRIMIDINE_CATABOLISM                                                 Nt5c3,Upp1   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION       0.250000          69  \n",
       " GO:0045540_GOBP_REGULATION_OF_CHOLESTEROL_BIOSY...  0.097561          69  \n",
       " R-HSA-2426168_REACTOME_ACTIVATION_OF_GENE_EXPRE...  0.093023          69  \n",
       " GO:0071280_GOBP_CELLULAR_RESPONSE_TO_COPPER_ION     0.187500          69  \n",
       " GO:0005911_GOCC_CELL_CELL_JUNCTION                  0.020930          69  \n",
       " GO:0046890_GOBP_REGULATION_OF_LIPID_BIOSYNTHETI...  0.035714          69  \n",
       " GO:0090181_GOBP_REGULATION_OF_CHOLESTEROL_METAB...  0.076923          69  \n",
       " R-HSA-1655829_REACTOME_REGULATION_OF_CHOLESTERO...  0.072727          69  \n",
       " GO:0045296_GOMF_CADHERIN_BINDING                    0.024306          69  \n",
       " R-HSA-191273_REACTOME_CHOLESTEROL_BIOSYNTHESIS      0.125000          69  \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...  0.060606          69  \n",
       " GO:0016126_GOBP_STEROL_BIOSYNTHETIC_PROCESS         0.057143          69  \n",
       " GO:0036017_GOBP_RESPONSE_TO_ERYTHROPOIETIN          0.400000          69  \n",
       " GO:0045338_GOBP_FARNESYL_DIPHOSPHATE_METABOLIC_...  0.400000          69  \n",
       " GO:0046688_GOBP_RESPONSE_TO_COPPER_ION              0.100000          69  \n",
       " GO:0050810_GOBP_REGULATION_OF_STEROID_BIOSYNTHE...  0.052632          69  \n",
       " R-HSA-446728_REACTOME_CELL_JUNCTION_ORGANIZATION    0.051282          69  \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS       0.333333          69  \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...  0.285714          69  \n",
       " GO:0050839_GOMF_CELL_ADHESION_MOLECULE_BINDING      0.016949          69  \n",
       " GO:0010043_GOBP_RESPONSE_TO_ZINC_ION                0.081081          69  \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION        0.250000          69  \n",
       " GO:0098632_GOMF_CELL_CELL_ADHESION_MEDIATOR_ACT...  0.071429          69  \n",
       " GO:0019218_GOBP_REGULATION_OF_STEROID_METABOLIC...  0.039604          69  \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                 0.200000          69  \n",
       " R-HSA-73621_REACTOME_PYRIMIDINE_CATABOLISM          0.200000          69  ,\n",
       " 'GP24_enrichment':                                                         pval   fdr  signature  \\\n",
       " gene_set                                                                        \n",
       " GO:0061615_GOBP_GLYCOLYTIC_PROCESS_THROUGH_FRUC...  0.000054  0.15         52   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                    0.000065  0.15         52   \n",
       " R-HSA-70263_REACTOME_GLUCONEOGENESIS                0.000075  0.15         52   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                       0.000089  0.15         52   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...  0.000090  0.15         52   \n",
       " GO:0006007_GOBP_GLUCOSE_CATABOLIC_PROCESS           0.000100  0.15         52   \n",
       " GO:0042989_GOBP_SEQUESTERING_OF_ACTIN_MONOMERS      0.000110  0.15         52   \n",
       " hsa00051_KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM       0.000110  0.15         52   \n",
       " GO:0006734_GOBP_NADH_METABOLIC_PROCESS              0.000190  0.23         52   \n",
       " GO:1904813_GOCC_FICOLIN_1_RICH_GRANULE_LUMEN        0.000210  0.23         52   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0061615_GOBP_GLYCOLYTIC_PROCESS_THROUGH_FRUC...       27        3   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                         79        4   \n",
       " R-HSA-70263_REACTOME_GLUCONEOGENESIS                     30        3   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                           268        6   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...       86        4   \n",
       " GO:0006007_GOBP_GLUCOSE_CATABOLIC_PROCESS                33        3   \n",
       " GO:0042989_GOBP_SEQUESTERING_OF_ACTIN_MONOMERS            6        2   \n",
       " hsa00051_KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM            34        3   \n",
       " GO:0006734_GOBP_NADH_METABOLIC_PROCESS                   41        3   \n",
       " GO:1904813_GOCC_FICOLIN_1_RICH_GRANULE_LUMEN            107        4   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0061615_GOBP_GLYCOLYTIC_PROCESS_THROUGH_FRUC...       18965   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                         18965   \n",
       " R-HSA-70263_REACTOME_GLUCONEOGENESIS                     18965   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                            18965   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...       18965   \n",
       " GO:0006007_GOBP_GLUCOSE_CATABOLIC_PROCESS                18965   \n",
       " GO:0042989_GOBP_SEQUESTERING_OF_ACTIN_MONOMERS           18965   \n",
       " hsa00051_KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM            18965   \n",
       " GO:0006734_GOBP_NADH_METABOLIC_PROCESS                   18965   \n",
       " GO:1904813_GOCC_FICOLIN_1_RICH_GRANULE_LUMEN             18965   \n",
       " \n",
       "                                                                                  hits  \\\n",
       " gene_set                                                                                \n",
       " GO:0061615_GOBP_GLYCOLYTIC_PROCESS_THROUGH_FRUC...                    Aldoa,Eno1,Tpi1   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                                      Gcg,Ppy,Pyy,Sst   \n",
       " R-HSA-70263_REACTOME_GLUCONEOGENESIS                                  Aldoa,Eno1,Tpi1   \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                       Acaa1a,Aldoa,Cotl1,Gcg,Gstp1,Idh1   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...                Aldoa,Eno1,Gcg,Tpi1   \n",
       " GO:0006007_GOBP_GLUCOSE_CATABOLIC_PROCESS                             Aldoa,Eno1,Tpi1   \n",
       " GO:0042989_GOBP_SEQUESTERING_OF_ACTIN_MONOMERS                      Tmsb15b1,Tmsb15b2   \n",
       " hsa00051_KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM                        Aldoa,Phpt1,Tpi1   \n",
       " GO:0006734_GOBP_NADH_METABOLIC_PROCESS                                Aldoa,Eno1,Tpi1   \n",
       " GO:1904813_GOCC_FICOLIN_1_RICH_GRANULE_LUMEN                   Aldoa,Cotl1,Gstp1,Idh1   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0061615_GOBP_GLYCOLYTIC_PROCESS_THROUGH_FRUC...  0.111111          52  \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                    0.050633          52  \n",
       " R-HSA-70263_REACTOME_GLUCONEOGENESIS                0.100000          52  \n",
       " GO:0031983_GOCC_VESICLE_LUMEN                       0.022388          52  \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...  0.046512          52  \n",
       " GO:0006007_GOBP_GLUCOSE_CATABOLIC_PROCESS           0.090909          52  \n",
       " GO:0042989_GOBP_SEQUESTERING_OF_ACTIN_MONOMERS      0.333333          52  \n",
       " hsa00051_KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM       0.088235          52  \n",
       " GO:0006734_GOBP_NADH_METABOLIC_PROCESS              0.073171          52  \n",
       " GO:1904813_GOCC_FICOLIN_1_RICH_GRANULE_LUMEN        0.037383          52  ,\n",
       " 'GP25_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   1.500000e-36   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           2.000000e-36   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          5.400000e-35   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  7.300000e-30   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       9.700000e-25   \n",
       " ...                                                          ...   \n",
       " GO:2001032_GOBP_REGULATION_OF_DOUBLE_STRAND_BRE...  2.300000e-03   \n",
       " R-HSA-196807_REACTOME_NICOTINATE_METABOLISM         2.300000e-03   \n",
       " GO:0140098_GOMF_CATALYTIC_ACTIVITY_ACTING_ON_RNA    2.500000e-03   \n",
       " R-HSA-5656169_REACTOME_TERMINATION_OF_TRANSLESI...  2.600000e-03   \n",
       " GO:0070566_GOMF_ADENYLYLTRANSFERASE_ACTIVITY        2.800000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   1.100000e-32         50   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           1.100000e-32         50   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          2.000000e-31         50   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  2.000000e-26         50   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       2.100000e-21         50   \n",
       " ...                                                          ...        ...   \n",
       " GO:2001032_GOBP_REGULATION_OF_DOUBLE_STRAND_BRE...  2.100000e-01         50   \n",
       " R-HSA-196807_REACTOME_NICOTINATE_METABOLISM         2.100000e-01         50   \n",
       " GO:0140098_GOMF_CATALYTIC_ACTIVITY_ACTING_ON_RNA    2.200000e-01         50   \n",
       " R-HSA-5656169_REACTOME_TERMINATION_OF_TRANSLESI...  2.400000e-01         50   \n",
       " GO:0070566_GOMF_ADENYLYLTRANSFERASE_ACTIVITY        2.500000e-01         50   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                       286       27   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS               207       25   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING              163       23   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...       54       16   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON            79       15   \n",
       " ...                                                     ...      ...   \n",
       " GO:2001032_GOBP_REGULATION_OF_DOUBLE_STRAND_BRE...       27        2   \n",
       " R-HSA-196807_REACTOME_NICOTINATE_METABOLISM              27        2   \n",
       " GO:0140098_GOMF_CATALYTIC_ACTIVITY_ACTING_ON_RNA        359        5   \n",
       " R-HSA-5656169_REACTOME_TERMINATION_OF_TRANSLESI...       29        2   \n",
       " GO:0070566_GOMF_ADENYLYLTRANSFERASE_ACTIVITY             30        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                        18965   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS                18965   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING               18965   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...       18965   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON            18965   \n",
       " ...                                                        ...   \n",
       " GO:2001032_GOBP_REGULATION_OF_DOUBLE_STRAND_BRE...       18965   \n",
       " R-HSA-196807_REACTOME_NICOTINATE_METABOLISM              18965   \n",
       " GO:0140098_GOMF_CATALYTIC_ACTIVITY_ACTING_ON_RNA         18965   \n",
       " R-HSA-5656169_REACTOME_TERMINATION_OF_TRANSLESI...       18965   \n",
       " GO:0070566_GOMF_ADENYLYLTRANSFERASE_ACTIVITY             18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   Apobec3,Cxcl10,Ddx58,Ddx60,Dhx58,Dtx3l,Eif2ak2...   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           Apobec3,Cxcl10,Ddx58,Ddx60,Dhx58,Dtx3l,Eif2ak2...   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          Ddx58,Eif2ak2,Gbp7,Herc6,H2-Q4,Ifi35,Ifit2,Ifi...   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  H2-Q4,Ifi35,Ifit2,Ifit3,Irf7,Irf9,Isg15,Oas1a,...   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       H2-Q4,Ifi35,Ifit2,Ifit3,Irf7,Irf9,Isg15,Oas1a,...   \n",
       " ...                                                                                               ...   \n",
       " GO:2001032_GOBP_REGULATION_OF_DOUBLE_STRAND_BRE...                                        Dtx3l,Parp9   \n",
       " R-HSA-196807_REACTOME_NICOTINATE_METABOLISM                                              Parp14,Parp9   \n",
       " GO:0140098_GOMF_CATALYTIC_ACTIVITY_ACTING_ON_RNA                        Ddx58,Ddx60,Dhx58,Helz2,Ifih1   \n",
       " R-HSA-5656169_REACTOME_TERMINATION_OF_TRANSLESI...                                       Isg15,Ube2l6   \n",
       " GO:0070566_GOMF_ADENYLYLTRANSFERASE_ACTIVITY                                               Oas1a,Oas2   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   0.094406          50  \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           0.120773          50  \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          0.141104          50  \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  0.296296          50  \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       0.189873          50  \n",
       " ...                                                      ...         ...  \n",
       " GO:2001032_GOBP_REGULATION_OF_DOUBLE_STRAND_BRE...  0.074074          50  \n",
       " R-HSA-196807_REACTOME_NICOTINATE_METABOLISM         0.074074          50  \n",
       " GO:0140098_GOMF_CATALYTIC_ACTIVITY_ACTING_ON_RNA    0.013928          50  \n",
       " R-HSA-5656169_REACTOME_TERMINATION_OF_TRANSLESI...  0.068966          50  \n",
       " GO:0070566_GOMF_ADENYLYLTRANSFERASE_ACTIVITY        0.066667          50  \n",
       " \n",
       " [123 rows x 9 columns],\n",
       " 'GP26_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               1.100000e-13   \n",
       " GO:0002237_GOBP_RESPONSE_TO_MOLECULE_OF_BACTERI...  1.100000e-12   \n",
       " GO:0071219_GOBP_CELLULAR_RESPONSE_TO_MOLECULE_O...  3.600000e-12   \n",
       " GO:0071216_GOBP_CELLULAR_RESPONSE_TO_BIOTIC_STI...  1.400000e-11   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        4.100000e-09   \n",
       " ...                                                          ...   \n",
       " GO:0032728_GOBP_POSITIVE_REGULATION_OF_INTERFER...  2.200000e-03   \n",
       " GO:0042110_GOBP_T_CELL_ACTIVATION                   2.500000e-03   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          2.500000e-03   \n",
       " GO:0045622_GOBP_REGULATION_OF_T_HELPER_CELL_DIF...  2.600000e-03   \n",
       " GO:0045776_GOBP_NEGATIVE_REGULATION_OF_BLOOD_PR...  2.600000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               1.200000e-09         46   \n",
       " GO:0002237_GOBP_RESPONSE_TO_MOLECULE_OF_BACTERI...  5.700000e-09         46   \n",
       " GO:0071219_GOBP_CELLULAR_RESPONSE_TO_MOLECULE_O...  1.300000e-08         46   \n",
       " GO:0071216_GOBP_CELLULAR_RESPONSE_TO_BIOTIC_STI...  3.800000e-08         46   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        8.900000e-06         46   \n",
       " ...                                                          ...        ...   \n",
       " GO:0032728_GOBP_POSITIVE_REGULATION_OF_INTERFER...  2.300000e-01         46   \n",
       " GO:0042110_GOBP_T_CELL_ACTIVATION                   2.500000e-01         46   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          2.500000e-01         46   \n",
       " GO:0045622_GOBP_REGULATION_OF_T_HELPER_CELL_DIF...  2.500000e-01         46   \n",
       " GO:0045776_GOBP_NEGATIVE_REGULATION_OF_BLOOD_PR...  2.500000e-01         46   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM                   456       15   \n",
       " GO:0002237_GOBP_RESPONSE_TO_MOLECULE_OF_BACTERI...      264       12   \n",
       " GO:0071219_GOBP_CELLULAR_RESPONSE_TO_MOLECULE_O...      157       10   \n",
       " GO:0071216_GOBP_CELLULAR_RESPONSE_TO_BIOTIC_STI...      180       10   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA            159        8   \n",
       " ...                                                     ...      ...   \n",
       " GO:0032728_GOBP_POSITIVE_REGULATION_OF_INTERFER...       29        2   \n",
       " GO:0042110_GOBP_T_CELL_ACTIVATION                       391        5   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS              112        3   \n",
       " GO:0045622_GOBP_REGULATION_OF_T_HELPER_CELL_DIF...       31        2   \n",
       " GO:0045776_GOBP_NEGATIVE_REGULATION_OF_BLOOD_PR...       31        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM                    18965   \n",
       " GO:0002237_GOBP_RESPONSE_TO_MOLECULE_OF_BACTERI...       18965   \n",
       " GO:0071219_GOBP_CELLULAR_RESPONSE_TO_MOLECULE_O...       18965   \n",
       " GO:0071216_GOBP_CELLULAR_RESPONSE_TO_BIOTIC_STI...       18965   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA             18965   \n",
       " ...                                                        ...   \n",
       " GO:0032728_GOBP_POSITIVE_REGULATION_OF_INTERFER...       18965   \n",
       " GO:0042110_GOBP_T_CELL_ACTIVATION                        18965   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS               18965   \n",
       " GO:0045622_GOBP_REGULATION_OF_T_HELPER_CELL_DIF...       18965   \n",
       " GO:0045776_GOBP_NEGATIVE_REGULATION_OF_BLOOD_PR...       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               Acod1,Bcl3,Ccl20,Cd40,Csf3,Icam1,Litaf,Lrg1,Nf...   \n",
       " GO:0002237_GOBP_RESPONSE_TO_MOLECULE_OF_BACTERI...  Acod1,Cd40,Csf3,Icam1,Litaf,Nfkb2,Nos2,Ripk2,S...   \n",
       " GO:0071219_GOBP_CELLULAR_RESPONSE_TO_MOLECULE_O...  Acod1,Cd40,Csf3,Icam1,Litaf,Nos2,Ripk2,Sbno2,T...   \n",
       " GO:0071216_GOBP_CELLULAR_RESPONSE_TO_BIOTIC_STI...  Acod1,Cd40,Csf3,Icam1,Litaf,Nos2,Ripk2,Sbno2,T...   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA             Acod1,Ccl20,Cd40,Gbp5,Icam1,Ifngr2,Nos2,Tlr2   \n",
       " ...                                                                                               ...   \n",
       " GO:0032728_GOBP_POSITIVE_REGULATION_OF_INTERFER...                                         Ripk2,Tlr2   \n",
       " GO:0042110_GOBP_T_CELL_ACTIVATION                                       Bcl3,Icam1,Ripk2,Sdc4,Zc3h12a   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS                                            Cd40,Icam1,Sdc4   \n",
       " GO:0045622_GOBP_REGULATION_OF_T_HELPER_CELL_DIF...                                      Ripk2,Zc3h12a   \n",
       " GO:0045776_GOBP_NEGATIVE_REGULATION_OF_BLOOD_PR...                                          Nos2,Sod2   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               0.032895          46  \n",
       " GO:0002237_GOBP_RESPONSE_TO_MOLECULE_OF_BACTERI...  0.045455          46  \n",
       " GO:0071219_GOBP_CELLULAR_RESPONSE_TO_MOLECULE_O...  0.063694          46  \n",
       " GO:0071216_GOBP_CELLULAR_RESPONSE_TO_BIOTIC_STI...  0.055556          46  \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        0.050314          46  \n",
       " ...                                                      ...         ...  \n",
       " GO:0032728_GOBP_POSITIVE_REGULATION_OF_INTERFER...  0.068966          46  \n",
       " GO:0042110_GOBP_T_CELL_ACTIVATION                   0.012788          46  \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          0.026786          46  \n",
       " GO:0045622_GOBP_REGULATION_OF_T_HELPER_CELL_DIF...  0.064516          46  \n",
       " GO:0045776_GOBP_NEGATIVE_REGULATION_OF_BLOOD_PR...  0.064516          46  \n",
       " \n",
       " [112 rows x 9 columns],\n",
       " 'GP27_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        1.100000e-14   \n",
       " GO:0042742_GOBP_DEFENSE_RESPONSE_TO_BACTERIUM       6.200000e-13   \n",
       " GO:0001562_GOBP_RESPONSE_TO_PROTOZOAN               1.200000e-12   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  2.400000e-12   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    1.500000e-11   \n",
       " ...                                                          ...   \n",
       " GO:0050865_GOBP_REGULATION_OF_CELL_ACTIVATION       6.600000e-03   \n",
       " R-HSA-8878166_REACTOME_TRANSCRIPTIONAL_REGULATI...  6.600000e-03   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...  6.800000e-03   \n",
       " GO:0009595_GOBP_DETECTION_OF_BIOTIC_STIMULUS        6.800000e-03   \n",
       " hsa04672_KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA...  6.800000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        1.100000e-10         67   \n",
       " GO:0042742_GOBP_DEFENSE_RESPONSE_TO_BACTERIUM       3.400000e-09         67   \n",
       " GO:0001562_GOBP_RESPONSE_TO_PROTOZOAN               4.200000e-09         67   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  6.400000e-09         67   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    3.200000e-08         67   \n",
       " ...                                                          ...        ...   \n",
       " GO:0050865_GOBP_REGULATION_OF_CELL_ACTIVATION       2.500000e-01         67   \n",
       " R-HSA-8878166_REACTOME_TRANSCRIPTIONAL_REGULATI...  2.500000e-01         67   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...  2.500000e-01         67   \n",
       " GO:0009595_GOBP_DETECTION_OF_BIOTIC_STIMULUS        2.500000e-01         67   \n",
       " hsa04672_KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA...  2.500000e-01         67   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA            159       13   \n",
       " GO:0042742_GOBP_DEFENSE_RESPONSE_TO_BACTERIUM           168       12   \n",
       " GO:0001562_GOBP_RESPONSE_TO_PROTOZOAN                    23        7   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...       69        9   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING         84        9   \n",
       " ...                                                     ...      ...   \n",
       " GO:0050865_GOBP_REGULATION_OF_CELL_ACTIVATION           475        6   \n",
       " R-HSA-8878166_REACTOME_TRANSCRIPTIONAL_REGULATI...      108        3   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...      337        5   \n",
       " GO:0009595_GOBP_DETECTION_OF_BIOTIC_STIMULUS             35        2   \n",
       " hsa04672_KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA...       35        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA             18965   \n",
       " GO:0042742_GOBP_DEFENSE_RESPONSE_TO_BACTERIUM            18965   \n",
       " GO:0001562_GOBP_RESPONSE_TO_PROTOZOAN                    18965   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...       18965   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING         18965   \n",
       " ...                                                        ...   \n",
       " GO:0050865_GOBP_REGULATION_OF_CELL_ACTIVATION            18965   \n",
       " R-HSA-8878166_REACTOME_TRANSCRIPTIONAL_REGULATI...       18965   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...       18965   \n",
       " GO:0009595_GOBP_DETECTION_OF_BIOTIC_STIMULUS             18965   \n",
       " hsa04672_KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA...       18965   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        Bst2,Ccl21a,Gbp2,Gbp3,Gbp10,Gbp4,Gbp8,Gbp9,H2-...   \n",
       " GO:0042742_GOBP_DEFENSE_RESPONSE_TO_BACTERIUM       Gbp2,Gbp3,Gbp10,Gbp4,Gbp8,Gbp9,H2-Q7,H2-T23,Ig...   \n",
       " GO:0001562_GOBP_RESPONSE_TO_PROTOZOAN                             Gbp2,Gbp3,Gbp10,Gbp4,Gbp8,Gbp9,Irf8   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  Gbp2,Gbp3,Gbp10,Gbp4,Gbp8,Gbp9,H2-Q7,H2-T23,Ra...   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING     Gbp2,Gbp3,Gbp10,Gbp4,Gbp8,Gbp9,H2-Q7,H2-T23,Irf8   \n",
       " ...                                                                                               ...   \n",
       " GO:0050865_GOBP_REGULATION_OF_CELL_ACTIVATION                  Ccl21a,Cd274,H2-Q7,H2-DMb1,H2-T23,Igha   \n",
       " R-HSA-8878166_REACTOME_TRANSCRIPTIONAL_REGULATI...                                 Psmb10,Psmb9,Psme1   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...                    Bst2,H2-Q7,H2-T23,H2-DMb1,Htra1   \n",
       " GO:0009595_GOBP_DETECTION_OF_BIOTIC_STIMULUS                                             H2-Q7,H2-T23   \n",
       " hsa04672_KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA...                                     Cxcl12,H2-DMb1   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        0.081761          67  \n",
       " GO:0042742_GOBP_DEFENSE_RESPONSE_TO_BACTERIUM       0.071429          67  \n",
       " GO:0001562_GOBP_RESPONSE_TO_PROTOZOAN               0.304348          67  \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  0.130435          67  \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    0.107143          67  \n",
       " ...                                                      ...         ...  \n",
       " GO:0050865_GOBP_REGULATION_OF_CELL_ACTIVATION       0.012632          67  \n",
       " R-HSA-8878166_REACTOME_TRANSCRIPTIONAL_REGULATI...  0.027778          67  \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...  0.014837          67  \n",
       " GO:0009595_GOBP_DETECTION_OF_BIOTIC_STIMULUS        0.057143          67  \n",
       " hsa04672_KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA...  0.057143          67  \n",
       " \n",
       " [293 rows x 9 columns],\n",
       " 'sample_explained_var':                           GP19  GP20   GP3   GP4 species  n_cells  \\\n",
       " sample                                                              \n",
       " Ngn3Cre_Control_1         0.00  0.00  0.00  0.00      mm      210   \n",
       " Ngn3Cre_Control_2         0.00  0.00  0.00  0.00      mm      138   \n",
       " 4m_head_Fltp-_mouse1      0.00  0.00  0.00  0.00      mm     3245   \n",
       " 4m_head_Fltp+_mouse2      0.00  0.00  0.00  0.00      mm     4252   \n",
       " 4m_tail_Fltp-_mouse3      0.00  0.00  0.00  0.00      mm     1916   \n",
       " 4m_tail_Fltp+_mouse4      0.00  0.00  0.00  0.00      mm     4325   \n",
       " 8-16wNOD_8w_SRR7610295    0.00  0.00  0.01  0.00      mm      189   \n",
       " 8-16wNOD_8w_SRR7610296    0.00  0.00  0.00  0.00      mm      128   \n",
       " 8-16wNOD_8w_SRR7610297    0.00  0.00  0.00  0.00      mm      231   \n",
       " mSTZ_control_G1           0.00  0.00  0.00  0.00      mm     5795   \n",
       " db/db_chow_WT_MUC13633    0.00  0.00  0.00  0.00      mm     3656   \n",
       " db/db_chow_WT_MUC13634    0.00  0.00  0.00  0.00      mm     4050   \n",
       " chem_DMSO_r1_SRR10751504  0.00  0.00  0.00  0.00      mm     1883   \n",
       " chem_DMSO_r2_SRR10751509  0.00  0.00  0.00  0.00      mm      426   \n",
       " chem_DMSO_r3_SRR10751514  0.00  0.01  0.11  0.00      mm     1010   \n",
       " H1                        0.13  0.22  0.67  0.07      hs      322   \n",
       " H2                        0.59  0.31  0.31  0.14      hs      247   \n",
       " H3                        0.18  0.13  0.17  0.05      hs     2302   \n",
       " H4                        0.19  0.06  0.26  0.17      hs     2357   \n",
       " H5                        0.52  0.04  0.30  0.10      hs     1202   \n",
       " H6                        0.19  0.01  0.21  0.14      hs     1662   \n",
       " R242                      0.64  0.18  0.00  0.00      hs      246   \n",
       " HPAP022                   0.00  0.01  0.00  0.01      hs     1134   \n",
       " HPAP035                   0.04  0.00  0.00  0.00      hs      408   \n",
       " HPAP036                   0.00  0.00  0.04  0.00      hs      274   \n",
       " HPAP037                   0.00  0.45  0.00  0.00      hs      444   \n",
       " HPAP040                   0.16  0.17  0.08  0.01      hs      332   \n",
       " R229                      0.00  0.00  0.00  0.00      hs     1975   \n",
       " R237                      0.01  0.13  0.03  0.02      hs     2196   \n",
       " R239                      0.00  0.00  0.00  0.00      hs     1485   \n",
       " R245                      0.00  0.00  0.00  0.00      hs     2440   \n",
       " \n",
       "                                            dataset  \n",
       " sample                                              \n",
       " Ngn3Cre_Control_1                        GSE137909  \n",
       " Ngn3Cre_Control_2                        GSE137909  \n",
       " 4m_head_Fltp-_mouse1                    mouseAtlas  \n",
       " 4m_head_Fltp+_mouse2                    mouseAtlas  \n",
       " 4m_tail_Fltp-_mouse3                    mouseAtlas  \n",
       " 4m_tail_Fltp+_mouse4                    mouseAtlas  \n",
       " 8-16wNOD_8w_SRR7610295                  mouseAtlas  \n",
       " 8-16wNOD_8w_SRR7610296                  mouseAtlas  \n",
       " 8-16wNOD_8w_SRR7610297                  mouseAtlas  \n",
       " mSTZ_control_G1                         mouseAtlas  \n",
       " db/db_chow_WT_MUC13633                  mouseAtlas  \n",
       " db/db_chow_WT_MUC13634                  mouseAtlas  \n",
       " chem_DMSO_r1_SRR10751504                mouseAtlas  \n",
       " chem_DMSO_r2_SRR10751509                mouseAtlas  \n",
       " chem_DMSO_r3_SRR10751514                mouseAtlas  \n",
       " H1                                       GSE101207  \n",
       " H2                                       GSE101207  \n",
       " H3                                       GSE101207  \n",
       " H4                                       GSE101207  \n",
       " H5                                       GSE101207  \n",
       " H6                                       GSE101207  \n",
       " R242                      GSE124742_GSE164875_FACS  \n",
       " HPAP022                                  GSE148073  \n",
       " HPAP035                                  GSE148073  \n",
       " HPAP036                                  GSE148073  \n",
       " HPAP037                                  GSE148073  \n",
       " HPAP040                                  GSE148073  \n",
       " R229                                     GSE198623  \n",
       " R237                                     GSE198623  \n",
       " R239                                     GSE198623  \n",
       " R245                                     GSE198623  }"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "db6bfc77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'GPs.xlsx',engine='xlsxwriter') \n",
    "for name,res_sub1 in res.items():\n",
    "    res_sub1.to_excel(writer, sheet_name=name,index=True)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "e3fe42ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/GPs.xlsx'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'GPs.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8a95027",
   "metadata": {},
   "source": [
    "## Healthy GP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "5d33545f",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "# Load table parts\n",
    "res={}\n",
    "res['GPs']=pd.read_table(path_data+'combined/moransi_healthy/sfintegrated/gene_hc_t3.tsv',\n",
    "                         index_col=0)\n",
    "res['GPs'].index.name='EID'\n",
    "for gp,res_sub in pd.read_excel(\n",
    "    path_data+'combined/moransi_healthy/sfintegrated/gene_hc_t3_enrichment.xlsx', \n",
    "    index_col=0,sheet_name=None).items():\n",
    "    res_sub.index.name='gene_set'\n",
    "    res_sub.drop('label',axis=1,inplace=True)\n",
    "    res[f'GP{gp}_enrichment']=res_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "ed801229",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'GPs':                     hc gene_symbol  rel_beta_expr  mean_expr_in_expr_cells  \\\n",
       " EID                                                                          \n",
       " ENSMUSG00000073530   3      Pappa2       0.649646                 1.198574   \n",
       " ENSMUSG00000036856   2        Wnt4       1.000000                 1.376984   \n",
       " ENSMUSG00000053560   4        Ier2       0.546635                 1.050378   \n",
       " ENSMUSG00000044988   1        Ucn3       1.000000                 1.515771   \n",
       " ENSMUSG00000028757   3       Ddost       1.000000                 1.624765   \n",
       " ...                 ..         ...            ...                      ...   \n",
       " ENSMUSG00000064351   1      mt-Co1       1.000000                 4.754871   \n",
       " ENSMUSG00000029780   3       Nt5c3       1.000000                 0.965134   \n",
       " ENSMUSG00000073411   3       H2-D1       0.603752                 1.738229   \n",
       " ENSMUSG00000018339   2        Gpx3       0.562181                 1.332709   \n",
       " ENSMUSG00000060802   1         B2m       0.463461                 1.116816   \n",
       " \n",
       "                     N_PMID  N_PMID_pancreas_notCancerNonendo  \n",
       " EID                                                           \n",
       " ENSMUSG00000073530    47.0                               0.0  \n",
       " ENSMUSG00000036856   111.0                               0.0  \n",
       " ENSMUSG00000053560    17.0                               0.0  \n",
       " ENSMUSG00000044988    34.0                               3.0  \n",
       " ENSMUSG00000028757   117.0                               6.0  \n",
       " ...                    ...                               ...  \n",
       " ENSMUSG00000064351    78.0                               6.0  \n",
       " ENSMUSG00000029780    52.0                               0.0  \n",
       " ENSMUSG00000073411     0.0                               0.0  \n",
       " ENSMUSG00000018339   112.0                               6.0  \n",
       " ENSMUSG00000060802   416.0                              13.0  \n",
       " \n",
       " [265 rows x 6 columns],\n",
       " 'GP1_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0046034_GOBP_ATP_METABOLIC_PROCESS               8.500000e-17   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           1.300000e-16   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    5.600000e-16   \n",
       " GO:0006091_GOBP_GENERATION_OF_PRECURSOR_METABOL...  1.500000e-14   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             2.500000e-14   \n",
       " ...                                                          ...   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE            4.700000e-03   \n",
       " GO:0015986_GOBP_ATP_SYNTHESIS_COUPLED_PROTON_TR...  4.700000e-03   \n",
       " GO:0048662_GOBP_NEGATIVE_REGULATION_OF_SMOOTH_M...  4.700000e-03   \n",
       " GO:1904036_GOBP_NEGATIVE_REGULATION_OF_EPITHELI...  4.700000e-03   \n",
       " GO:0062013_GOBP_POSITIVE_REGULATION_OF_SMALL_MO...  4.900000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0046034_GOBP_ATP_METABOLIC_PROCESS               6.100000e-13         50   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           6.100000e-13         50   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    1.700000e-12         50   \n",
       " GO:0006091_GOBP_GENERATION_OF_PRECURSOR_METABOL...  3.500000e-11         50   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             4.700000e-11         50   \n",
       " ...                                                          ...        ...   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE            2.400000e-01         50   \n",
       " GO:0015986_GOBP_ATP_SYNTHESIS_COUPLED_PROTON_TR...  2.400000e-01         50   \n",
       " GO:0048662_GOBP_NEGATIVE_REGULATION_OF_SMOOTH_M...  2.400000e-01         50   \n",
       " GO:1904036_GOBP_NEGATIVE_REGULATION_OF_EPITHELI...  2.400000e-01         50   \n",
       " GO:0062013_GOBP_POSITIVE_REGULATION_OF_SMALL_MO...  2.500000e-01         50   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0046034_GOBP_ATP_METABOLIC_PROCESS                   252       17   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION               133       14   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                        113       13   \n",
       " GO:0006091_GOBP_GENERATION_OF_PRECURSOR_METABOL...      407       18   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                 114       12   \n",
       " ...                                                     ...      ...   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE                164        4   \n",
       " GO:0015986_GOBP_ATP_SYNTHESIS_COUPLED_PROTON_TR...       25        2   \n",
       " GO:0048662_GOBP_NEGATIVE_REGULATION_OF_SMOOTH_M...       25        2   \n",
       " GO:1904036_GOBP_NEGATIVE_REGULATION_OF_EPITHELI...       25        2   \n",
       " GO:0062013_GOBP_POSITIVE_REGULATION_OF_SMALL_MO...       83        3   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0046034_GOBP_ATP_METABOLIC_PROCESS                    12069   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION                12069   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                         12069   \n",
       " GO:0006091_GOBP_GENERATION_OF_PRECURSOR_METABOL...       12069   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION                  12069   \n",
       " ...                                                        ...   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE                 12069   \n",
       " GO:0015986_GOBP_ATP_SYNTHESIS_COUPLED_PROTON_TR...       12069   \n",
       " GO:0048662_GOBP_NEGATIVE_REGULATION_OF_SMOOTH_M...       12069   \n",
       " GO:1904036_GOBP_NEGATIVE_REGULATION_OF_EPITHELI...       12069   \n",
       " GO:0062013_GOBP_POSITIVE_REGULATION_OF_SMALL_MO...       12069   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0046034_GOBP_ATP_METABOLIC_PROCESS               App,Chchd10,Gapdh,mt-Atp6,mt-Atp8,mt-Co1,mt-Co...   \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           Chchd10,mt-Atp6,mt-Atp8,mt-Co1,mt-Co2,mt-Co3,m...   \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    mt-Atp6,mt-Atp8,mt-Co1,mt-Co2,mt-Co3,mt-Cytb,m...   \n",
       " GO:0006091_GOBP_GENERATION_OF_PRECURSOR_METABOL...  App,Chchd10,Gapdh,mt-Atp6,mt-Atp8,mt-Co1,mt-Co...   \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             mt-Atp6,mt-Atp8,mt-Co1,mt-Co2,mt-Co3,mt-Cytb,m...   \n",
       " ...                                                                                               ...   \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE                                     Camk2n1,Ppargc1a,Th,Ucn3   \n",
       " GO:0015986_GOBP_ATP_SYNTHESIS_COUPLED_PROTON_TR...                                    mt-Atp6,mt-Atp8   \n",
       " GO:0048662_GOBP_NEGATIVE_REGULATION_OF_SMOOTH_M...                                       Ang,Ppargc1a   \n",
       " GO:1904036_GOBP_NEGATIVE_REGULATION_OF_EPITHELI...                                         Nupr1,Scg2   \n",
       " GO:0062013_GOBP_POSITIVE_REGULATION_OF_SMALL_MO...                                App,mt-Co2,Ppargc1a   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0046034_GOBP_ATP_METABOLIC_PROCESS               0.067460          50  \n",
       " GO:0006119_GOBP_OXIDATIVE_PHOSPHORYLATION           0.105263          50  \n",
       " hsa05012_KEGG_PARKINSONS_DISEASE                    0.115044          50  \n",
       " GO:0006091_GOBP_GENERATION_OF_PRECURSOR_METABOL...  0.044226          50  \n",
       " hsa00190_KEGG_OXIDATIVE_PHOSPHORYLATION             0.105263          50  \n",
       " ...                                                      ...         ...  \n",
       " GO:0009743_GOBP_RESPONSE_TO_CARBOHYDRATE            0.024390          50  \n",
       " GO:0015986_GOBP_ATP_SYNTHESIS_COUPLED_PROTON_TR...  0.080000          50  \n",
       " GO:0048662_GOBP_NEGATIVE_REGULATION_OF_SMOOTH_M...  0.080000          50  \n",
       " GO:1904036_GOBP_NEGATIVE_REGULATION_OF_EPITHELI...  0.080000          50  \n",
       " GO:0062013_GOBP_POSITIVE_REGULATION_OF_SMALL_MO...  0.036145          50  \n",
       " \n",
       " [184 rows x 9 columns],\n",
       " 'GP2_enrichment':                                                         pval   fdr  signature  \\\n",
       " gene_set                                                                        \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING         0.000016  0.11         27   \n",
       " GO:1903034_GOBP_REGULATION_OF_RESPONSE_TO_WOUNDING  0.000043  0.11         27   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING           0.000043  0.11         27   \n",
       " GO:1905050_GOBP_POSITIVE_REGULATION_OF_METALLOP...  0.000048  0.11         27   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...  0.000057  0.11         27   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY       0.000072  0.11         27   \n",
       " GO:1903035_GOBP_NEGATIVE_REGULATION_OF_RESPONSE...  0.000120  0.16         27   \n",
       " GO:0016338_GOBP_CALCIUM_INDEPENDENT_CELL_CELL_A...  0.000130  0.16         27   \n",
       " GO:0070661_GOBP_LEUKOCYTE_PROLIFERATION             0.000180  0.18         27   \n",
       " GO:1901890_GOBP_POSITIVE_REGULATION_OF_CELL_JUN...  0.000210  0.18         27   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING       0.000210  0.18         27   \n",
       " GO:0098609_GOBP_CELL_CELL_ADHESION                  0.000220  0.18         27   \n",
       " GO:1905048_GOBP_REGULATION_OF_METALLOPEPTIDASE_...  0.000260  0.18         27   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          0.000260  0.18         27   \n",
       " GO:0042060_GOBP_WOUND_HEALING                       0.000370  0.23         27   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING              69        4   \n",
       " GO:1903034_GOBP_REGULATION_OF_RESPONSE_TO_WOUNDING       89        4   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING                89        4   \n",
       " GO:1905050_GOBP_POSITIVE_REGULATION_OF_METALLOP...        5        2   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...       34        3   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY             6        2   \n",
       " GO:1903035_GOBP_NEGATIVE_REGULATION_OF_RESPONSE...       44        3   \n",
       " GO:0016338_GOBP_CALCIUM_INDEPENDENT_CELL_CELL_A...        8        2   \n",
       " GO:0070661_GOBP_LEUKOCYTE_PROLIFERATION                 129        4   \n",
       " GO:1901890_GOBP_POSITIVE_REGULATION_OF_CELL_JUN...       52        3   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING            10        2   \n",
       " GO:0098609_GOBP_CELL_CELL_ADHESION                      405        6   \n",
       " GO:1905048_GOBP_REGULATION_OF_METALLOPEPTIDASE_...       11        2   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS               56        3   \n",
       " GO:0042060_GOBP_WOUND_HEALING                           285        5   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING              12069   \n",
       " GO:1903034_GOBP_REGULATION_OF_RESPONSE_TO_WOUNDING       12069   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING                12069   \n",
       " GO:1905050_GOBP_POSITIVE_REGULATION_OF_METALLOP...       12069   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...       12069   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY            12069   \n",
       " GO:1903035_GOBP_NEGATIVE_REGULATION_OF_RESPONSE...       12069   \n",
       " GO:0016338_GOBP_CALCIUM_INDEPENDENT_CELL_CELL_A...       12069   \n",
       " GO:0070661_GOBP_LEUKOCYTE_PROLIFERATION                  12069   \n",
       " GO:1901890_GOBP_POSITIVE_REGULATION_OF_CELL_JUN...       12069   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING            12069   \n",
       " GO:0098609_GOBP_CELL_CELL_ADHESION                       12069   \n",
       " GO:1905048_GOBP_REGULATION_OF_METALLOPEPTIDASE_...       12069   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS               12069   \n",
       " GO:0042060_GOBP_WOUND_HEALING                            12069   \n",
       " \n",
       "                                                                                  hits  \\\n",
       " gene_set                                                                                \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING                   Cldn3,Cldn4,Tspan8,Wnt4   \n",
       " GO:1903034_GOBP_REGULATION_OF_RESPONSE_TO_WOUNDING            Cldn3,Cldn4,Tspan8,Wnt4   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING                          Ppy,Pyy,Ramp1,Wnt4   \n",
       " GO:1905050_GOBP_POSITIVE_REGULATION_OF_METALLOP...                        Cldn3,Cldn4   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...                  Cldn3,Tspan8,Wnt4   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY                                 Ppy,Pyy   \n",
       " GO:1903035_GOBP_NEGATIVE_REGULATION_OF_RESPONSE...                  Cldn3,Tspan8,Wnt4   \n",
       " GO:0016338_GOBP_CALCIUM_INDEPENDENT_CELL_CELL_A...                        Cldn3,Cldn4   \n",
       " GO:0070661_GOBP_LEUKOCYTE_PROLIFERATION                          Cd79a,Cd81,Sdc4,Wnt4   \n",
       " GO:1901890_GOBP_POSITIVE_REGULATION_OF_CELL_JUN...                    Cldn3,Sdc4,Wnt4   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING                                 Ppy,Pyy   \n",
       " GO:0098609_GOBP_CELL_CELL_ADHESION                  Baiap2,Cd81,Cldn3,Cldn4,Sdc4,Wnt4   \n",
       " GO:1905048_GOBP_REGULATION_OF_METALLOPEPTIDASE_...                        Cldn3,Cldn4   \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS                           Cldn3,Cldn4,Sdc4   \n",
       " GO:0042060_GOBP_WOUND_HEALING                            Cldn3,Cldn4,Sdc4,Tspan8,Wnt4   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING         0.057971          27  \n",
       " GO:1903034_GOBP_REGULATION_OF_RESPONSE_TO_WOUNDING  0.044944          27  \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING           0.044944          27  \n",
       " GO:1905050_GOBP_POSITIVE_REGULATION_OF_METALLOP...  0.400000          27  \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...  0.088235          27  \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY       0.333333          27  \n",
       " GO:1903035_GOBP_NEGATIVE_REGULATION_OF_RESPONSE...  0.068182          27  \n",
       " GO:0016338_GOBP_CALCIUM_INDEPENDENT_CELL_CELL_A...  0.250000          27  \n",
       " GO:0070661_GOBP_LEUKOCYTE_PROLIFERATION             0.031008          27  \n",
       " GO:1901890_GOBP_POSITIVE_REGULATION_OF_CELL_JUN...  0.057692          27  \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING       0.200000          27  \n",
       " GO:0098609_GOBP_CELL_CELL_ADHESION                  0.014815          27  \n",
       " GO:1905048_GOBP_REGULATION_OF_METALLOPEPTIDASE_...  0.181818          27  \n",
       " hsa04514_KEGG_CELL_ADHESION_MOLECULES_CAMS          0.053571          27  \n",
       " GO:0042060_GOBP_WOUND_HEALING                       0.017544          27  ,\n",
       " 'GP3_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  1.200000e-21   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         3.500000e-12   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX   2.500000e-11   \n",
       " R-HSA-2980736_REACTOME_PEPTIDE_HORMONE_METABOLISM   7.400000e-10   \n",
       " R-HSA-422085_REACTOME_SYNTHESIS_SECRETION_AND_D...  1.300000e-09   \n",
       " ...                                                          ...   \n",
       " GO:0016528_GOCC_SARCOPLASM                          2.400000e-03   \n",
       " GO:1903039_GOBP_POSITIVE_REGULATION_OF_LEUKOCYT...  2.600000e-03   \n",
       " GO:0051591_GOBP_RESPONSE_TO_CAMP                    2.700000e-03   \n",
       " GO:0071675_GOBP_REGULATION_OF_MONONUCLEAR_CELL_...  2.700000e-03   \n",
       " GO:1900102_GOBP_NEGATIVE_REGULATION_OF_ENDOPLAS...  2.800000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  1.100000e-17         69   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         1.700000e-08         69   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX   7.900000e-08         69   \n",
       " R-HSA-2980736_REACTOME_PEPTIDE_HORMONE_METABOLISM   1.700000e-06         69   \n",
       " R-HSA-422085_REACTOME_SYNTHESIS_SECRETION_AND_D...  2.400000e-06         69   \n",
       " ...                                                          ...        ...   \n",
       " GO:0016528_GOCC_SARCOPLASM                          2.100000e-01         69   \n",
       " GO:1903039_GOBP_POSITIVE_REGULATION_OF_LEUKOCYT...  2.200000e-01         69   \n",
       " GO:0051591_GOBP_RESPONSE_TO_CAMP                    2.300000e-01         69   \n",
       " GO:0071675_GOBP_REGULATION_OF_MONONUCLEAR_CELL_...  2.300000e-01         69   \n",
       " GO:1900102_GOBP_NEGATIVE_REGULATION_OF_ENDOPLAS...  2.300000e-01         69   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...      115       18   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN             156       13   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX        12        6   \n",
       " R-HSA-2980736_REACTOME_PEPTIDE_HORMONE_METABOLISM        55        8   \n",
       " R-HSA-422085_REACTOME_SYNTHESIS_SECRETION_AND_D...       10        5   \n",
       " ...                                                     ...      ...   \n",
       " GO:0016528_GOCC_SARCOPLASM                               47        3   \n",
       " GO:1903039_GOBP_POSITIVE_REGULATION_OF_LEUKOCYT...      101        4   \n",
       " GO:0051591_GOBP_RESPONSE_TO_CAMP                         49        3   \n",
       " GO:0071675_GOBP_REGULATION_OF_MONONUCLEAR_CELL_...       49        3   \n",
       " GO:1900102_GOBP_NEGATIVE_REGULATION_OF_ENDOPLAS...       14        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...       12069   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN              12069   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX        12069   \n",
       " R-HSA-2980736_REACTOME_PEPTIDE_HORMONE_METABOLISM        12069   \n",
       " R-HSA-422085_REACTOME_SYNTHESIS_SECRETION_AND_D...       12069   \n",
       " ...                                                        ...   \n",
       " GO:0016528_GOCC_SARCOPLASM                               12069   \n",
       " GO:1903039_GOBP_POSITIVE_REGULATION_OF_LEUKOCYT...       12069   \n",
       " GO:0051591_GOBP_RESPONSE_TO_CAMP                         12069   \n",
       " GO:0071675_GOBP_REGULATION_OF_MONONUCLEAR_CELL_...       12069   \n",
       " GO:1900102_GOBP_NEGATIVE_REGULATION_OF_ENDOPLAS...       12069   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  Calr,Dad1,Ddost,Derl3,H2-D1,Hspa5,Krtcap2,Ostc...   \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         Calr,Chgb,Clu,Cst3,Dnajb9,Hsp90b1,Hspa5,Ins2,M...   \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX                   Dad1,Ddost,Krtcap2,Ostc,Rpn1,Rpn2   \n",
       " R-HSA-2980736_REACTOME_PEPTIDE_HORMONE_METABOLISM       Cltrn,Cpe,Ins2,Pcsk1,Pcsk2,Sec11c,Spcs1,Spcs2   \n",
       " R-HSA-422085_REACTOME_SYNTHESIS_SECRETION_AND_D...                      Ins2,Pcsk1,Sec11c,Spcs1,Spcs2   \n",
       " ...                                                                                               ...   \n",
       " GO:0016528_GOCC_SARCOPLASM                                                          Calr,Hsp90b1,Manf   \n",
       " GO:1903039_GOBP_POSITIVE_REGULATION_OF_LEUKOCYT...                         Cd24a,H2-D1,Nfkbiz,Selenok   \n",
       " GO:0051591_GOBP_RESPONSE_TO_CAMP                                                      Bsg,Hspa5,Ptprn   \n",
       " GO:0071675_GOBP_REGULATION_OF_MONONUCLEAR_CELL_...                                 Calr,Cd200,Selenok   \n",
       " GO:1900102_GOBP_NEGATIVE_REGULATION_OF_ENDOPLAS...                                       Dnajb9,Hspa5   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0140534_GOCC_ENDOPLASMIC_RETICULUM_PROTEIN_C...  0.156522          69  \n",
       " GO:0005788_GOCC_ENDOPLASMIC_RETICULUM_LUMEN         0.083333          69  \n",
       " GO:0008250_GOCC_OLIGOSACCHARYLTRANSFERASE_COMPLEX   0.500000          69  \n",
       " R-HSA-2980736_REACTOME_PEPTIDE_HORMONE_METABOLISM   0.145455          69  \n",
       " R-HSA-422085_REACTOME_SYNTHESIS_SECRETION_AND_D...  0.500000          69  \n",
       " ...                                                      ...         ...  \n",
       " GO:0016528_GOCC_SARCOPLASM                          0.063830          69  \n",
       " GO:1903039_GOBP_POSITIVE_REGULATION_OF_LEUKOCYT...  0.039604          69  \n",
       " GO:0051591_GOBP_RESPONSE_TO_CAMP                    0.061224          69  \n",
       " GO:0071675_GOBP_REGULATION_OF_MONONUCLEAR_CELL_...  0.061224          69  \n",
       " GO:1900102_GOBP_NEGATIVE_REGULATION_OF_ENDOPLAS...  0.142857          69  \n",
       " \n",
       " [115 rows x 9 columns],\n",
       " 'GP4_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  2.100000e-09  0.000010   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                2.600000e-09  0.000010   \n",
       " GO:0002221_GOBP_PATTERN_RECOGNITION_RECEPTOR_SI...  3.100000e-09  0.000010   \n",
       " GO:0002753_GOBP_CYTOPLASMIC_PATTERN_RECOGNITION...  5.400000e-09  0.000012   \n",
       " GO:0035872_GOBP_NUCLEOTIDE_BINDING_DOMAIN_LEUCI...  6.900000e-09  0.000012   \n",
       " ...                                                          ...       ...   \n",
       " GO:0032480_GOBP_NEGATIVE_REGULATION_OF_TYPE_I_I...  9.300000e-03  0.250000   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...  9.400000e-03  0.250000   \n",
       " GO:0031348_GOBP_NEGATIVE_REGULATION_OF_DEFENSE_...  9.500000e-03  0.250000   \n",
       " GO:0071901_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  9.500000e-03  0.250000   \n",
       " GO:0032147_GOBP_ACTIVATION_OF_PROTEIN_KINASE_AC...  9.600000e-03  0.250000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...         49      238   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                       49      184   \n",
       " GO:0002221_GOBP_PATTERN_RECOGNITION_RECEPTOR_SI...         49      136   \n",
       " GO:0002753_GOBP_CYTOPLASMIC_PATTERN_RECOGNITION...         49       63   \n",
       " GO:0035872_GOBP_NUCLEOTIDE_BINDING_DOMAIN_LEUCI...         49       37   \n",
       " ...                                                       ...      ...   \n",
       " GO:0032480_GOBP_NEGATIVE_REGULATION_OF_TYPE_I_I...         49       36   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...         49      205   \n",
       " GO:0031348_GOBP_NEGATIVE_REGULATION_OF_DEFENSE_...         49      108   \n",
       " GO:0071901_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...         49      108   \n",
       " GO:0032147_GOBP_ACTIVATION_OF_PROTEIN_KINASE_AC...         49      206   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...       11       12069   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                     10       12069   \n",
       " GO:0002221_GOBP_PATTERN_RECOGNITION_RECEPTOR_SI...        9       12069   \n",
       " GO:0002753_GOBP_CYTOPLASMIC_PATTERN_RECOGNITION...        7       12069   \n",
       " GO:0035872_GOBP_NUCLEOTIDE_BINDING_DOMAIN_LEUCI...        6       12069   \n",
       " ...                                                     ...         ...   \n",
       " GO:0032480_GOBP_NEGATIVE_REGULATION_OF_TYPE_I_I...        2       12069   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...        4       12069   \n",
       " GO:0031348_GOBP_NEGATIVE_REGULATION_OF_DEFENSE_...        3       12069   \n",
       " GO:0071901_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...        3       12069   \n",
       " GO:0032147_GOBP_ACTIVATION_OF_PROTEIN_KINASE_AC...        4       12069   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  Atf3,Cebpd,Ddit3,Egr1,Elf3,Fos,Fosb,Irf1,Jun,J...   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                Ddit3,Dusp1,Fos,Gadd45b,Gadd45g,Hspa1a,Hspa1b,...   \n",
       " GO:0002221_GOBP_PATTERN_RECOGNITION_RECEPTOR_SI...  Birc3,Hspa1a,Hspa1b,Irf1,Nfkbia,Nr1d1,Tifa,Tnf...   \n",
       " GO:0002753_GOBP_CYTOPLASMIC_PATTERN_RECOGNITION...        Birc3,Hspa1a,Hspa1b,Nfkbia,Tifa,Tnfaip3,Ubc   \n",
       " GO:0035872_GOBP_NUCLEOTIDE_BINDING_DOMAIN_LEUCI...             Birc3,Hspa1a,Hspa1b,Nfkbia,Tnfaip3,Ubc   \n",
       " ...                                                                                               ...   \n",
       " GO:0032480_GOBP_NEGATIVE_REGULATION_OF_TYPE_I_I...                                       Relb,Tnfaip3   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...                           Dusp1,Rgs4,Socs3,Tnfaip3   \n",
       " GO:0031348_GOBP_NEGATIVE_REGULATION_OF_DEFENSE_...                                Nr1d1,Socs3,Tnfaip3   \n",
       " GO:0071901_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...                                 Dusp1,Rgs4,Tnfaip3   \n",
       " GO:0032147_GOBP_ACTIVATION_OF_PROTEIN_KINASE_AC...                        Adora2b,Gadd45b,Gadd45g,Ubc   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  0.046218          49  \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                0.054348          49  \n",
       " GO:0002221_GOBP_PATTERN_RECOGNITION_RECEPTOR_SI...  0.066176          49  \n",
       " GO:0002753_GOBP_CYTOPLASMIC_PATTERN_RECOGNITION...  0.111111          49  \n",
       " GO:0035872_GOBP_NUCLEOTIDE_BINDING_DOMAIN_LEUCI...  0.162162          49  \n",
       " ...                                                      ...         ...  \n",
       " GO:0032480_GOBP_NEGATIVE_REGULATION_OF_TYPE_I_I...  0.055556          49  \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...  0.019512          49  \n",
       " GO:0031348_GOBP_NEGATIVE_REGULATION_OF_DEFENSE_...  0.027778          49  \n",
       " GO:0071901_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  0.027778          49  \n",
       " GO:0032147_GOBP_ACTIVATION_OF_PROTEIN_KINASE_AC...  0.019417          49  \n",
       " \n",
       " [356 rows x 9 columns],\n",
       " 'GP5_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0006413_GOBP_TRANSLATIONAL_INITIATION            3.900000e-10  0.000004   \n",
       " R-HSA-72613_REACTOME_EUKARYOTIC_TRANSLATION_INI...  9.400000e-10  0.000004   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...  2.900000e-09  0.000007   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS  2.900000e-09  0.000007   \n",
       " R-HSA-9675108_REACTOME_NERVOUS_SYSTEM_DEVELOPMENT   1.500000e-08  0.000028   \n",
       " ...                                                          ...       ...   \n",
       " R-HSA-2173788_REACTOME_DOWNREGULATION_OF_TGF_BE...  7.800000e-03  0.250000   \n",
       " R-HSA-5696397_REACTOME_GAP_FILLING_DNA_REPAIR_S...  7.800000e-03  0.250000   \n",
       " R-HSA-6802957_REACTOME_ONCOGENIC_MAPK_SIGNALING     7.800000e-03  0.250000   \n",
       " R-HSA-8982491_REACTOME_GLYCOGEN_METABOLISM          7.800000e-03  0.250000   \n",
       " GO:0030856_GOBP_REGULATION_OF_EPITHELIAL_CELL_D...  8.100000e-03  0.250000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0006413_GOBP_TRANSLATIONAL_INITIATION                   70      179   \n",
       " R-HSA-72613_REACTOME_EUKARYOTIC_TRANSLATION_INI...         70      114   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...         70       23   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS         70       23   \n",
       " R-HSA-9675108_REACTOME_NERVOUS_SYSTEM_DEVELOPMENT          70      424   \n",
       " ...                                                       ...      ...   \n",
       " R-HSA-2173788_REACTOME_DOWNREGULATION_OF_TGF_BE...         70       23   \n",
       " R-HSA-5696397_REACTOME_GAP_FILLING_DNA_REPAIR_S...         70       23   \n",
       " R-HSA-6802957_REACTOME_ONCOGENIC_MAPK_SIGNALING            70       70   \n",
       " R-HSA-8982491_REACTOME_GLYCOGEN_METABOLISM                 70       23   \n",
       " GO:0030856_GOBP_REGULATION_OF_EPITHELIAL_CELL_D...         70       71   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0006413_GOBP_TRANSLATIONAL_INITIATION                 12       12069   \n",
       " R-HSA-72613_REACTOME_EUKARYOTIC_TRANSLATION_INI...       10       12069   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...        6       12069   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS        6       12069   \n",
       " R-HSA-9675108_REACTOME_NERVOUS_SYSTEM_DEVELOPMENT        15       12069   \n",
       " ...                                                     ...         ...   \n",
       " R-HSA-2173788_REACTOME_DOWNREGULATION_OF_TGF_BE...        2       12069   \n",
       " R-HSA-5696397_REACTOME_GAP_FILLING_DNA_REPAIR_S...        2       12069   \n",
       " R-HSA-6802957_REACTOME_ONCOGENIC_MAPK_SIGNALING           3       12069   \n",
       " R-HSA-8982491_REACTOME_GLYCOGEN_METABOLISM                2       12069   \n",
       " GO:0030856_GOBP_REGULATION_OF_EPITHELIAL_CELL_D...        3       12069   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0006413_GOBP_TRANSLATIONAL_INITIATION            Atf4,Eif1,Eif3m,Eif4a2,Eif4ebp1,Rpl10,Rpl32,Rp...   \n",
       " R-HSA-72613_REACTOME_EUKARYOTIC_TRANSLATION_INI...  Eif3m,Eif4a2,Eif4ebp1,Rpl10,Rpl32,Rplp0,Rplp1,...   \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...              Actb,Actg1,Tuba1a,Tubb2a,Tubb4b,Tubb6   \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS              Actb,Actg1,Tuba1a,Tubb2a,Tubb4b,Tubb6   \n",
       " R-HSA-9675108_REACTOME_NERVOUS_SYSTEM_DEVELOPMENT   Actb,Actg1,Cntn1,Rpl10,Rpl32,Rplp0,Rplp1,Rps24...   \n",
       " ...                                                                                               ...   \n",
       " R-HSA-2173788_REACTOME_DOWNREGULATION_OF_TGF_BE...                                         Rps27a,Ubb   \n",
       " R-HSA-5696397_REACTOME_GAP_FILLING_DNA_REPAIR_S...                                         Rps27a,Ubb   \n",
       " R-HSA-6802957_REACTOME_ONCOGENIC_MAPK_SIGNALING                                       Actb,Actg1,Lmna   \n",
       " R-HSA-8982491_REACTOME_GLYCOGEN_METABOLISM                                                 Rps27a,Ubb   \n",
       " GO:0030856_GOBP_REGULATION_OF_EPITHELIAL_CELL_D...                                  Errfi1,Maff,Ovol2   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0006413_GOBP_TRANSLATIONAL_INITIATION            0.067039          70  \n",
       " R-HSA-72613_REACTOME_EUKARYOTIC_TRANSLATION_INI...  0.087719          70  \n",
       " R-HSA-157858_REACTOME_GAP_JUNCTION_TRAFFICKING_...  0.260870          70  \n",
       " R-HSA-5626467_REACTOME_RHO_GTPASES_ACTIVATE_IQGAPS  0.260870          70  \n",
       " R-HSA-9675108_REACTOME_NERVOUS_SYSTEM_DEVELOPMENT   0.035377          70  \n",
       " ...                                                      ...         ...  \n",
       " R-HSA-2173788_REACTOME_DOWNREGULATION_OF_TGF_BE...  0.086957          70  \n",
       " R-HSA-5696397_REACTOME_GAP_FILLING_DNA_REPAIR_S...  0.086957          70  \n",
       " R-HSA-6802957_REACTOME_ONCOGENIC_MAPK_SIGNALING     0.042857          70  \n",
       " R-HSA-8982491_REACTOME_GLYCOGEN_METABOLISM          0.086957          70  \n",
       " GO:0030856_GOBP_REGULATION_OF_EPITHELIAL_CELL_D...  0.042254          70  \n",
       " \n",
       " [299 rows x 9 columns]}"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "a95b72a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'GPs_healthy.xlsx',engine='xlsxwriter') \n",
    "for name,res_sub1 in res.items():\n",
    "    res_sub1.to_excel(writer, sheet_name=name,index=True)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "d1618e7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/GPs_healthy.xlsx'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'GPs_healthy.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c795417",
   "metadata": {},
   "source": [
    "## Diabetes DE in beta cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "6db35236",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load table parts\n",
    "res={}\n",
    "for dt,subdir in [('T1D','de_diseased_T1_NODelim_meld'),\n",
    "                  ('T2D','de_diseased_T2_VSGSTZ_meld_covarStudy')]:\n",
    "    res[dt+'_DE']=pd.read_table(path_data+f'combined/de/{subdir}/deDataClusters.tsv',\n",
    "              index_col=0)\n",
    "    for hc,res_sub in pd.read_excel(\n",
    "        path_data+f'combined/de/{subdir}/deDataClusters_enrichment.xlsx', \n",
    "        index_col=0,sheet_name=None).items():\n",
    "        res_sub.index.name='gene_set'\n",
    "        res_sub.drop('label',axis=1,inplace=True)\n",
    "        res[f'{dt}_DE_{hc}_enrichment']=res_sub\n",
    "for direction, res_sub in pd.read_excel(\n",
    "    path_data+'combined/de/compare_T1D-NODelim_T2D-VSGSTZ/sharedClUpDown_enrichment.xlsx', \n",
    "    index_col=0,sheet_name=None).items():\n",
    "    res_sub.index.name='gene_set'\n",
    "    res_sub.drop('label',axis=1,inplace=True)\n",
    "    res['sharedT1DT2D_DE_enrichment_'+direction]=res_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "afc7b9ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'T1D_DE':                             pval          qval    log2fc       mean  \\\n",
       " gene                                                                  \n",
       " ENSMUSG00000033845  3.214516e-02  8.088905e-02  0.174473   0.613649   \n",
       " ENSMUSG00000025903  9.685890e-02  1.930702e-01  0.169455   0.368679   \n",
       " ENSMUSG00000033813  4.071410e-12  5.839446e-11  0.449678   1.195759   \n",
       " ENSMUSG00000033793  3.169995e-01  4.666418e-01 -0.081697   0.536161   \n",
       " ENSMUSG00000025907  6.325680e-03  2.125627e-02  0.249674   0.507885   \n",
       " ...                          ...           ...       ...        ...   \n",
       " ENSMUSG00000064368  4.229510e-01  5.701349e-01 -0.084558   0.543774   \n",
       " ENSMUSG00000064370  0.000000e+00  0.000000e+00 -0.717106  63.479608   \n",
       " ENSMUSG00000051412  2.510859e-02  6.614282e-02  0.158873   0.957314   \n",
       " ENSMUSG00000095041  1.123247e-01  2.155093e-01 -0.145508   0.462208   \n",
       " ENSMUSG00000063897  1.666407e-02  4.746895e-02 -0.466192   0.084285   \n",
       " \n",
       "                     zero_mean          grad  coef_mle   coef_sd            ll  \\\n",
       " gene                                                                            \n",
       " ENSMUSG00000033845      False  2.536053e-09  0.174473  0.081430  -3709.992764   \n",
       " ENSMUSG00000025903      False  3.779941e-01  0.169455  0.102064   -574.293738   \n",
       " ENSMUSG00000033813      False  1.528924e-08  0.449678  0.064845  -5042.057500   \n",
       " ENSMUSG00000033793      False  9.556359e-10 -0.081697  0.081644  -3451.115039   \n",
       " ENSMUSG00000025907      False  1.835053e-09  0.249674  0.091442  -3367.414571   \n",
       " ...                       ...           ...       ...       ...           ...   \n",
       " ENSMUSG00000064368      False  9.444358e-08 -0.084558  0.105524  -3606.304483   \n",
       " ENSMUSG00000064370      False  2.776893e-07 -0.717106  0.031945 -17726.818750   \n",
       " ENSMUSG00000051412      False  1.511566e-07  0.158873  0.070934  -4736.382210   \n",
       " ENSMUSG00000095041      False  8.763851e-09 -0.145508  0.091640  -3195.617625   \n",
       " ENSMUSG00000063897      False  1.606422e-10 -0.466192  0.194730  -1066.495916   \n",
       " \n",
       "                        gene_symbol   hc  rel_beta_expr  N_PMID  \\\n",
       " gene                                                             \n",
       " ENSMUSG00000033845          Mrpl15  NaN       0.577643    55.0   \n",
       " ENSMUSG00000025903          Lypla1  NaN       0.755665    42.0   \n",
       " ENSMUSG00000033813           Tcea1  NaN       0.898695    66.0   \n",
       " ENSMUSG00000033793         Atp6v1h  NaN       1.000000    65.0   \n",
       " ENSMUSG00000025907          Rb1cc1  NaN       1.000000   124.0   \n",
       " ...                            ...  ...            ...     ...   \n",
       " ENSMUSG00000064368          mt-Nd6  NaN       1.000000    42.0   \n",
       " ENSMUSG00000064370         mt-Cytb  NaN       1.000000    39.0   \n",
       " ENSMUSG00000051412           Vamp7  NaN       1.000000     0.0   \n",
       " ENSMUSG00000095041      AC149090.1  NaN       1.000000     0.0   \n",
       " ENSMUSG00000063897  CAAA01118383.1  NaN       0.619154     0.0   \n",
       " \n",
       "                     N_PMID_pancreas_notCancerNonendo  mean_expr_in_expr_cells  \n",
       " gene                                                                           \n",
       " ENSMUSG00000033845                               0.0                 0.665728  \n",
       " ENSMUSG00000025903                               0.0                 0.617923  \n",
       " ENSMUSG00000033813                               0.0                 0.794732  \n",
       " ENSMUSG00000033793                               1.0                 0.706451  \n",
       " ENSMUSG00000025907                               1.0                 0.808519  \n",
       " ...                                              ...                      ...  \n",
       " ENSMUSG00000064368                               4.0                 0.684435  \n",
       " ENSMUSG00000064370                               5.0                 4.252210  \n",
       " ENSMUSG00000051412                               0.0                 0.746626  \n",
       " ENSMUSG00000095041                               0.0                 0.812761  \n",
       " ENSMUSG00000063897                               0.0                 0.673011  \n",
       " \n",
       " [9839 rows x 15 columns],\n",
       " 'T1D_DE_down_1_enrichment':                                                             pval      fdr  \\\n",
       " gene_set                                                                    \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY       5.100000e-08  0.00046   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING       2.900000e-07  0.00130   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY      3.400000e-06  0.01000   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                    1.200000e-05  0.02600   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                    2.500000e-05  0.04400   \n",
       " R-HSA-418594_REACTOME_G_ALPHA_I_SIGNALLING_EVENTS   3.500000e-05  0.04900   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...  3.800000e-05  0.04900   \n",
       " GO:0001664_GOMF_G_PROTEIN_COUPLED_RECEPTOR_BINDING  5.300000e-05  0.05900   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...  1.000000e-04  0.10000   \n",
       " GO:0060575_GOBP_INTESTINAL_EPITHELIAL_CELL_DIFF...  2.400000e-04  0.22000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY              18        5   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING              18        8   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY             18       17   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                           18       25   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                           18       32   \n",
       " R-HSA-418594_REACTOME_G_ALPHA_I_SIGNALLING_EVENTS          18      106   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...         18       37   \n",
       " GO:0001664_GOMF_G_PROTEIN_COUPLED_RECEPTOR_BINDING         18      118   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...         18       51   \n",
       " GO:0060575_GOBP_INTESTINAL_EPITHELIAL_CELL_DIFF...         18       13   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY             3        9839   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING             3        9839   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY            3        9839   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                          3        9839   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                          3        9839   \n",
       " R-HSA-418594_REACTOME_G_ALPHA_I_SIGNALLING_EVENTS         4        9839   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...        3        9839   \n",
       " GO:0001664_GOMF_G_PROTEIN_COUPLED_RECEPTOR_BINDING        4        9839   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...        3        9839   \n",
       " GO:0060575_GOBP_INTESTINAL_EPITHELIAL_CELL_DIFF...        2        9839   \n",
       " \n",
       "                                                                  hits  \\\n",
       " gene_set                                                                \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY             Npy,Ppy,Pyy   \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING             Npy,Ppy,Pyy   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY            Npy,Ppy,Pyy   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                          Npy,Ppy,Pyy   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                          Npy,Ppy,Pyy   \n",
       " R-HSA-418594_REACTOME_G_ALPHA_I_SIGNALLING_EVENTS   Gnai1,Npy,Ppy,Pyy   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...        Npy,Ppy,Pyy   \n",
       " GO:0001664_GOMF_G_PROTEIN_COUPLED_RECEPTOR_BINDING  Gnai1,Npy,Ppy,Pyy   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...        Npy,Ppy,Pyy   \n",
       " GO:0060575_GOBP_INTESTINAL_EPITHELIAL_CELL_DIFF...            Npy,Pyy   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY       0.600000          18  \n",
       " GO:0071855_GOMF_NEUROPEPTIDE_RECEPTOR_BINDING       0.375000          18  \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY      0.176471          18  \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                    0.120000          18  \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                    0.093750          18  \n",
       " R-HSA-418594_REACTOME_G_ALPHA_I_SIGNALLING_EVENTS   0.037736          18  \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...  0.081081          18  \n",
       " GO:0001664_GOMF_G_PROTEIN_COUPLED_RECEPTOR_BINDING  0.033898          18  \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...  0.058824          18  \n",
       " GO:0060575_GOBP_INTESTINAL_EPITHELIAL_CELL_DIFF...  0.153846          18  ,\n",
       " 'T1D_DE_down_2_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'T1D_DE_down_3_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'T1D_DE_down_4_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  7.900000e-15   \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  3.500000e-10   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                    9.500000e-10   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                    1.400000e-08   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM      2.800000e-08   \n",
       " ...                                                          ...   \n",
       " GO:0046320_GOBP_REGULATION_OF_FATTY_ACID_OXIDATION  6.800000e-03   \n",
       " GO:0060674_GOBP_PLACENTA_BLOOD_VESSEL_DEVELOPMENT   6.800000e-03   \n",
       " GO:0005976_GOBP_POLYSACCHARIDE_METABOLIC_PROCESS    6.900000e-03   \n",
       " GO:0046982_GOMF_PROTEIN_HETERODIMERIZATION_ACTI...  6.900000e-03   \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               7.100000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  7.000000e-11         65   \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  1.500000e-06         65   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                    2.800000e-06         65   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                    3.200000e-05         65   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM      4.900000e-05         65   \n",
       " ...                                                          ...        ...   \n",
       " GO:0046320_GOBP_REGULATION_OF_FATTY_ACID_OXIDATION  2.400000e-01         65   \n",
       " GO:0060674_GOBP_PLACENTA_BLOOD_VESSEL_DEVELOPMENT   2.400000e-01         65   \n",
       " GO:0005976_GOBP_POLYSACCHARIDE_METABOLIC_PROCESS    2.400000e-01         65   \n",
       " GO:0046982_GOMF_PROTEIN_HETERODIMERIZATION_ACTI...  2.400000e-01         65   \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               2.500000e-01         65   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...      468       23   \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...      196       13   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                        134       11   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                        173       11   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM           75        8   \n",
       " ...                                                     ...      ...   \n",
       " GO:0046320_GOBP_REGULATION_OF_FATTY_ACID_OXIDATION       19        2   \n",
       " GO:0060674_GOBP_PLACENTA_BLOOD_VESSEL_DEVELOPMENT        19        2   \n",
       " GO:0005976_GOBP_POLYSACCHARIDE_METABOLIC_PROCESS         59        3   \n",
       " GO:0046982_GOMF_PROTEIN_HETERODIMERIZATION_ACTI...      182        5   \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM                   259        6   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...        9839   \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...        9839   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                          9839   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                          9839   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM            9839   \n",
       " ...                                                        ...   \n",
       " GO:0046320_GOBP_REGULATION_OF_FATTY_ACID_OXIDATION        9839   \n",
       " GO:0060674_GOBP_PLACENTA_BLOOD_VESSEL_DEVELOPMENT         9839   \n",
       " GO:0005976_GOBP_POLYSACCHARIDE_METABOLIC_PROCESS          9839   \n",
       " GO:0046982_GOMF_PROTEIN_HETERODIMERIZATION_ACTI...        9839   \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM                     9839   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  Atf3,Chd7,Ddit3,Egr1,Egr4,Fos,Fosb,Jun,Junb,Kl...   \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  Atf3,Ddit3,Egr1,Egr4,Fos,Fosb,Jun,Junb,Klf4,Kl...   \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                    Cry2,Egr1,Jun,Nfil3,Nr1d1,Nr1d2,Per1,Per2,Ppar...   \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                    Cry2,Egr1,Jun,Nfil3,Nr1d1,Nr1d2,Per1,Per2,Ppar...   \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM         Cry2,Nr1d1,Nr1d2,Per1,Per2,Ppargc1a,Siah2,Usp2   \n",
       " ...                                                                                               ...   \n",
       " GO:0046320_GOBP_REGULATION_OF_FATTY_ACID_OXIDATION                                      Irs2,Ppargc1a   \n",
       " GO:0060674_GOBP_PLACENTA_BLOOD_VESSEL_DEVELOPMENT                                          Junb,Socs3   \n",
       " GO:0005976_GOBP_POLYSACCHARIDE_METABOLIC_PROCESS                                      Irs2,Nr1d1,Per2   \n",
       " GO:0046982_GOMF_PROTEIN_HETERODIMERIZATION_ACTI...                       Atf3,Ddit3,Npas4,Nr4a1,Nr4a2   \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM                             Chd7,Fos,Jun,Nr1d1,Ppargc1a,Tnfaip3   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  0.049145          65  \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  0.066327          65  \n",
       " GO:0007623_GOBP_CIRCADIAN_RHYTHM                    0.082090          65  \n",
       " GO:0048511_GOBP_RHYTHMIC_PROCESS                    0.063584          65  \n",
       " GO:0042752_GOBP_REGULATION_OF_CIRCADIAN_RHYTHM      0.106667          65  \n",
       " ...                                                      ...         ...  \n",
       " GO:0046320_GOBP_REGULATION_OF_FATTY_ACID_OXIDATION  0.105263          65  \n",
       " GO:0060674_GOBP_PLACENTA_BLOOD_VESSEL_DEVELOPMENT   0.105263          65  \n",
       " GO:0005976_GOBP_POLYSACCHARIDE_METABOLIC_PROCESS    0.050847          65  \n",
       " GO:0046982_GOMF_PROTEIN_HETERODIMERIZATION_ACTI...  0.027473          65  \n",
       " GO:0009617_GOBP_RESPONSE_TO_BACTERIUM               0.023166          65  \n",
       " \n",
       " [252 rows x 9 columns],\n",
       " 'T1D_DE_down_5_enrichment':                                                         pval    fdr  \\\n",
       " gene_set                                                              \n",
       " GO:0008330_GOMF_PROTEIN_TYROSINE_THREONINE_PHOS...  0.000002  0.020   \n",
       " GO:0033549_GOMF_MAP_KINASE_PHOSPHATASE_ACTIVITY     0.000007  0.030   \n",
       " GO:0042326_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...  0.000027  0.049   \n",
       " GO:0035970_GOBP_PEPTIDYL_THREONINE_DEPHOSPHORYL...  0.000032  0.049   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...  0.000034  0.049   \n",
       " GO:0000188_GOBP_INACTIVATION_OF_MAPK_ACTIVITY       0.000038  0.049   \n",
       " R-HSA-112409_REACTOME_RAF_INDEPENDENT_MAPK1_3_A...  0.000038  0.049   \n",
       " GO:0051348_GOBP_NEGATIVE_REGULATION_OF_TRANSFER...  0.000063  0.070   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...  0.000090  0.089   \n",
       " GO:0010563_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...  0.000120  0.100   \n",
       " GO:0008138_GOMF_PROTEIN_TYROSINE_SERINE_THREONI...  0.000130  0.100   \n",
       " R-HSA-9652817_REACTOME_SIGNALING_BY_MAPK_MUTANTS    0.000180  0.140   \n",
       " GO:0043589_GOBP_SKIN_MORPHOGENESIS                  0.000260  0.170   \n",
       " R-HSA-5675221_REACTOME_NEGATIVE_REGULATION_OF_M...  0.000270  0.170   \n",
       " GO:0046165_GOBP_ALCOHOL_BIOSYNTHETIC_PROCESS        0.000290  0.170   \n",
       " GO:0001704_GOBP_FORMATION_OF_PRIMARY_GERM_LAYER     0.000350  0.180   \n",
       " GO:0062012_GOBP_REGULATION_OF_SMALL_MOLECULE_ME...  0.000350  0.180   \n",
       " GO:0016791_GOMF_PHOSPHATASE_ACTIVITY                0.000380  0.190   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0008330_GOMF_PROTEIN_TYROSINE_THREONINE_PHOS...         35        8   \n",
       " GO:0033549_GOMF_MAP_KINASE_PHOSPHATASE_ACTIVITY            35       11   \n",
       " GO:0042326_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...         35      257   \n",
       " GO:0035970_GOBP_PEPTIDYL_THREONINE_DEPHOSPHORYL...         35       18   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...         35      178   \n",
       " GO:0000188_GOBP_INACTIVATION_OF_MAPK_ACTIVITY              35       19   \n",
       " R-HSA-112409_REACTOME_RAF_INDEPENDENT_MAPK1_3_A...         35       19   \n",
       " GO:0051348_GOBP_NEGATIVE_REGULATION_OF_TRANSFER...         35      199   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...         35       25   \n",
       " GO:0010563_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...         35      324   \n",
       " GO:0008138_GOMF_PROTEIN_TYROSINE_SERINE_THREONI...         35       28   \n",
       " R-HSA-9652817_REACTOME_SIGNALING_BY_MAPK_MUTANTS           35        6   \n",
       " GO:0043589_GOBP_SKIN_MORPHOGENESIS                         35        7   \n",
       " R-HSA-5675221_REACTOME_NEGATIVE_REGULATION_OF_M...         35       36   \n",
       " GO:0046165_GOBP_ALCOHOL_BIOSYNTHETIC_PROCESS               35       91   \n",
       " GO:0001704_GOBP_FORMATION_OF_PRIMARY_GERM_LAYER            35       39   \n",
       " GO:0062012_GOBP_REGULATION_OF_SMALL_MOLECULE_ME...         35      272   \n",
       " GO:0016791_GOMF_PHOSPHATASE_ACTIVITY                       35      178   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0008330_GOMF_PROTEIN_TYROSINE_THREONINE_PHOS...        3        9839   \n",
       " GO:0033549_GOMF_MAP_KINASE_PHOSPHATASE_ACTIVITY           3        9839   \n",
       " GO:0042326_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...        7        9839   \n",
       " GO:0035970_GOBP_PEPTIDYL_THREONINE_DEPHOSPHORYL...        3        9839   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...        6        9839   \n",
       " GO:0000188_GOBP_INACTIVATION_OF_MAPK_ACTIVITY             3        9839   \n",
       " R-HSA-112409_REACTOME_RAF_INDEPENDENT_MAPK1_3_A...        3        9839   \n",
       " GO:0051348_GOBP_NEGATIVE_REGULATION_OF_TRANSFER...        6        9839   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...        3        9839   \n",
       " GO:0010563_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...        7        9839   \n",
       " GO:0008138_GOMF_PROTEIN_TYROSINE_SERINE_THREONI...        3        9839   \n",
       " R-HSA-9652817_REACTOME_SIGNALING_BY_MAPK_MUTANTS          2        9839   \n",
       " GO:0043589_GOBP_SKIN_MORPHOGENESIS                        2        9839   \n",
       " R-HSA-5675221_REACTOME_NEGATIVE_REGULATION_OF_M...        3        9839   \n",
       " GO:0046165_GOBP_ALCOHOL_BIOSYNTHETIC_PROCESS              4        9839   \n",
       " GO:0001704_GOBP_FORMATION_OF_PRIMARY_GERM_LAYER           3        9839   \n",
       " GO:0062012_GOBP_REGULATION_OF_SMALL_MOLECULE_ME...        6        9839   \n",
       " GO:0016791_GOMF_PHOSPHATASE_ACTIVITY                      5        9839   \n",
       " \n",
       "                                                                                             hits  \\\n",
       " gene_set                                                                                           \n",
       " GO:0008330_GOMF_PROTEIN_TYROSINE_THREONINE_PHOS...                            Dusp10,Dusp5,Dusp8   \n",
       " GO:0033549_GOMF_MAP_KINASE_PHOSPHATASE_ACTIVITY                               Dusp10,Dusp5,Dusp8   \n",
       " GO:0042326_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...  Dusp10,Dusp5,Dusp8,Errfi1,Ldlrad4,Midn,Trib1   \n",
       " GO:0035970_GOBP_PEPTIDYL_THREONINE_DEPHOSPHORYL...                            Dusp10,Dusp5,Dusp8   \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...          Dusp10,Dusp5,Dusp8,Errfi1,Midn,Trib1   \n",
       " GO:0000188_GOBP_INACTIVATION_OF_MAPK_ACTIVITY                                 Dusp10,Dusp5,Dusp8   \n",
       " R-HSA-112409_REACTOME_RAF_INDEPENDENT_MAPK1_3_A...                            Dusp10,Dusp5,Dusp8   \n",
       " GO:0051348_GOBP_NEGATIVE_REGULATION_OF_TRANSFER...          Dusp10,Dusp5,Dusp8,Errfi1,Midn,Trib1   \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...                            Dusp10,Dusp5,Dusp8   \n",
       " GO:0010563_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...  Dusp10,Dusp5,Dusp8,Errfi1,Ldlrad4,Midn,Trib1   \n",
       " GO:0008138_GOMF_PROTEIN_TYROSINE_SERINE_THREONI...                            Dusp10,Dusp5,Dusp8   \n",
       " R-HSA-9652817_REACTOME_SIGNALING_BY_MAPK_MUTANTS                                    Dusp10,Dusp8   \n",
       " GO:0043589_GOBP_SKIN_MORPHOGENESIS                                                    Errfi1,Srf   \n",
       " R-HSA-5675221_REACTOME_NEGATIVE_REGULATION_OF_M...                            Dusp10,Dusp5,Dusp8   \n",
       " GO:0046165_GOBP_ALCOHOL_BIOSYNTHETIC_PROCESS                            Hmgcs1,Idi1,Insig1,Itpkc   \n",
       " GO:0001704_GOBP_FORMATION_OF_PRIMARY_GERM_LAYER                                  Dusp5,Kdm6b,Srf   \n",
       " GO:0062012_GOBP_REGULATION_OF_SMALL_MOLECULE_ME...           Hmgcs1,Idi1,Insig1,Midn,Odc1,Prkab2   \n",
       " GO:0016791_GOMF_PHOSPHATASE_ACTIVITY                               Dusp10,Dusp5,Dusp8,Lpin1,Noct   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0008330_GOMF_PROTEIN_TYROSINE_THREONINE_PHOS...  0.375000          35  \n",
       " GO:0033549_GOMF_MAP_KINASE_PHOSPHATASE_ACTIVITY     0.272727          35  \n",
       " GO:0042326_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...  0.027237          35  \n",
       " GO:0035970_GOBP_PEPTIDYL_THREONINE_DEPHOSPHORYL...  0.166667          35  \n",
       " GO:0033673_GOBP_NEGATIVE_REGULATION_OF_KINASE_A...  0.033708          35  \n",
       " GO:0000188_GOBP_INACTIVATION_OF_MAPK_ACTIVITY       0.157895          35  \n",
       " R-HSA-112409_REACTOME_RAF_INDEPENDENT_MAPK1_3_A...  0.157895          35  \n",
       " GO:0051348_GOBP_NEGATIVE_REGULATION_OF_TRANSFER...  0.030151          35  \n",
       " GO:0051019_GOMF_MITOGEN_ACTIVATED_PROTEIN_KINAS...  0.120000          35  \n",
       " GO:0010563_GOBP_NEGATIVE_REGULATION_OF_PHOSPHOR...  0.021605          35  \n",
       " GO:0008138_GOMF_PROTEIN_TYROSINE_SERINE_THREONI...  0.107143          35  \n",
       " R-HSA-9652817_REACTOME_SIGNALING_BY_MAPK_MUTANTS    0.333333          35  \n",
       " GO:0043589_GOBP_SKIN_MORPHOGENESIS                  0.285714          35  \n",
       " R-HSA-5675221_REACTOME_NEGATIVE_REGULATION_OF_M...  0.083333          35  \n",
       " GO:0046165_GOBP_ALCOHOL_BIOSYNTHETIC_PROCESS        0.043956          35  \n",
       " GO:0001704_GOBP_FORMATION_OF_PRIMARY_GERM_LAYER     0.076923          35  \n",
       " GO:0062012_GOBP_REGULATION_OF_SMALL_MOLECULE_ME...  0.022059          35  \n",
       " GO:0016791_GOMF_PHOSPHATASE_ACTIVITY                0.028090          35  ,\n",
       " 'T1D_DE_up_1_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'T1D_DE_up_2_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              5.200000e-29   \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   8.300000e-26   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           8.400000e-26   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          4.600000e-25   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  8.900000e-22   \n",
       " ...                                                          ...   \n",
       " GO:0072524_GOBP_PYRIDINE_CONTAINING_COMPOUND_ME...  1.100000e-02   \n",
       " GO:1903320_GOBP_REGULATION_OF_PROTEIN_MODIFICAT...  1.100000e-02   \n",
       " GO:0035577_GOCC_AZUROPHIL_GRANULE_MEMBRANE          1.100000e-02   \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                 1.100000e-02   \n",
       " R-HSA-2979096_REACTOME_NOTCH2_ACTIVATION_AND_TR...  1.100000e-02   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              4.600000e-25         97   \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   2.500000e-22         97   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           2.500000e-22         97   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          1.000000e-21         97   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  1.600000e-18         97   \n",
       " ...                                                          ...        ...   \n",
       " GO:0072524_GOBP_PYRIDINE_CONTAINING_COMPOUND_ME...  2.400000e-01         97   \n",
       " GO:1903320_GOBP_REGULATION_OF_PROTEIN_MODIFICAT...  2.400000e-01         97   \n",
       " GO:0035577_GOCC_AZUROPHIL_GRANULE_MEMBRANE          2.500000e-01         97   \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                 2.400000e-01         97   \n",
       " R-HSA-2979096_REACTOME_NOTCH2_ACTIVATION_AND_TR...  2.400000e-01         97   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE                  440       40   \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                       212       29   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS               151       26   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING              142       25   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...      415       33   \n",
       " ...                                                     ...      ...   \n",
       " GO:0072524_GOBP_PYRIDINE_CONTAINING_COMPOUND_ME...       16        2   \n",
       " GO:1903320_GOBP_REGULATION_OF_PROTEIN_MODIFICAT...      188        6   \n",
       " GO:0035577_GOCC_AZUROPHIL_GRANULE_MEMBRANE               47        3   \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                      16        2   \n",
       " R-HSA-2979096_REACTOME_NOTCH2_ACTIVATION_AND_TR...       16        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE                    9839   \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                         9839   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS                 9839   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING                9839   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...        9839   \n",
       " ...                                                        ...   \n",
       " GO:0072524_GOBP_PYRIDINE_CONTAINING_COMPOUND_ME...        9839   \n",
       " GO:1903320_GOBP_REGULATION_OF_PROTEIN_MODIFICAT...        9839   \n",
       " GO:0035577_GOCC_AZUROPHIL_GRANULE_MEMBRANE                9839   \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                       9839   \n",
       " R-HSA-2979096_REACTOME_NOTCH2_ACTIVATION_AND_TR...        9839   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              Adar,Apobec3,Apol9a,Bst2,Ddx58,Ddx60,Defb1,Dtx...   \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   Adar,Apobec3,Bst2,Cxcl10,Ddx58,Ddx60,Dtx3l,Eif...   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           Adar,Apobec3,Bst2,Cxcl10,Ddx58,Ddx60,Dtx3l,Eif...   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          Adar,Bst2,Ddx58,Eif2ak2,Gbp7,Herc6,H2-Q4,H2-T2...   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  Adar,Bst2,Cdkn1a,Cxcl10,Gstp1,H2-Q4,H2-T22,Ifi...   \n",
       " ...                                                                                               ...   \n",
       " GO:0072524_GOBP_PYRIDINE_CONTAINING_COMPOUND_ME...                                       Parp10,Parp9   \n",
       " GO:1903320_GOBP_REGULATION_OF_PROTEIN_MODIFICAT...                Dtx3l,Isg15,Nmi,Parp10,Ripk2,Trim21   \n",
       " GO:0035577_GOCC_AZUROPHIL_GRANULE_MEMBRANE                                          Bst2,H2-Q4,H2-T22   \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                                                      H2-Q4,H2-T22   \n",
       " R-HSA-2979096_REACTOME_NOTCH2_ACTIVATION_AND_TR...                                      Neurl1a,Uba52   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              0.090909          97  \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   0.136792          97  \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           0.172185          97  \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          0.176056          97  \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  0.079518          97  \n",
       " ...                                                      ...         ...  \n",
       " GO:0072524_GOBP_PYRIDINE_CONTAINING_COMPOUND_ME...  0.125000          97  \n",
       " GO:1903320_GOBP_REGULATION_OF_PROTEIN_MODIFICAT...  0.031915          97  \n",
       " GO:0035577_GOCC_AZUROPHIL_GRANULE_MEMBRANE          0.063830          97  \n",
       " GO:0042611_GOCC_MHC_PROTEIN_COMPLEX                 0.125000          97  \n",
       " R-HSA-2979096_REACTOME_NOTCH2_ACTIVATION_AND_TR...  0.125000          97  \n",
       " \n",
       " [385 rows x 9 columns],\n",
       " 'T1D_DE_up_3_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    9.400000e-11   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  1.900000e-10   \n",
       " GO:0002475_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  5.800000e-10   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  3.200000e-09   \n",
       " GO:0002483_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  4.500000e-09   \n",
       " ...                                                          ...   \n",
       " GO:0002228_GOBP_NATURAL_KILLER_CELL_MEDIATED_IM...  5.000000e-03   \n",
       " GO:0006471_GOBP_PROTEIN_ADP_RIBOSYLATION            5.000000e-03   \n",
       " GO:0042743_GOBP_HYDROGEN_PEROXIDE_METABOLIC_PRO...  5.000000e-03   \n",
       " GO:0016684_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  5.000000e-03   \n",
       " GO:0070820_GOCC_TERTIARY_GRANULE                    5.300000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    8.300000e-07         39   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  8.300000e-07         39   \n",
       " GO:0002475_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  1.700000e-06         39   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  7.100000e-06         39   \n",
       " GO:0002483_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  8.000000e-06         39   \n",
       " ...                                                          ...        ...   \n",
       " GO:0002228_GOBP_NATURAL_KILLER_CELL_MEDIATED_IM...  2.400000e-01         39   \n",
       " GO:0006471_GOBP_PROTEIN_ADP_RIBOSYLATION            2.400000e-01         39   \n",
       " GO:0042743_GOBP_HYDROGEN_PEROXIDE_METABOLIC_PRO...  2.400000e-01         39   \n",
       " GO:0016684_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  2.400000e-01         39   \n",
       " GO:0070820_GOCC_TERTIARY_GRANULE                    2.500000e-01         39   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING         63        8   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       10        5   \n",
       " GO:0002475_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       12        5   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...       34        6   \n",
       " GO:0002483_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       17        5   \n",
       " ...                                                     ...      ...   \n",
       " GO:0002228_GOBP_NATURAL_KILLER_CELL_MEDIATED_IM...       27        2   \n",
       " GO:0006471_GOBP_PROTEIN_ADP_RIBOSYLATION                 27        2   \n",
       " GO:0042743_GOBP_HYDROGEN_PEROXIDE_METABOLIC_PRO...       27        2   \n",
       " GO:0016684_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...       27        2   \n",
       " GO:0070820_GOCC_TERTIARY_GRANULE                         90        3   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING          9839   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...        9839   \n",
       " GO:0002475_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...        9839   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...        9839   \n",
       " GO:0002483_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...        9839   \n",
       " ...                                                        ...   \n",
       " GO:0002228_GOBP_NATURAL_KILLER_CELL_MEDIATED_IM...        9839   \n",
       " GO:0006471_GOBP_PROTEIN_ADP_RIBOSYLATION                  9839   \n",
       " GO:0042743_GOBP_HYDROGEN_PEROXIDE_METABOLIC_PRO...        9839   \n",
       " GO:0016684_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...        9839   \n",
       " GO:0070820_GOCC_TERTIARY_GRANULE                          9839   \n",
       " \n",
       "                                                                                            hits  \\\n",
       " gene_set                                                                                          \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    B2m,Gbp2,Gbp10,Gbp4,Gbp8,H2-Q7,H2-T23,H2-K1   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...                  B2m,H2-T23,H2-Q7,H2-K1,Tap2   \n",
       " GO:0002475_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...                  B2m,H2-T23,H2-Q7,H2-K1,Tap2   \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...            Gbp2,Gbp10,Gbp4,Gbp8,H2-Q7,H2-T23   \n",
       " GO:0002483_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...                  B2m,H2-Q7,H2-T23,H2-K1,Tap2   \n",
       " ...                                                                                         ...   \n",
       " GO:0002228_GOBP_NATURAL_KILLER_CELL_MEDIATED_IM...                                 H2-Q7,H2-T23   \n",
       " GO:0006471_GOBP_PROTEIN_ADP_RIBOSYLATION                                            Parp11,Wars   \n",
       " GO:0042743_GOBP_HYDROGEN_PEROXIDE_METABOLIC_PRO...                                Hba-a1,Hbb-bt   \n",
       " GO:0016684_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...                                Hba-a1,Hbb-bt   \n",
       " GO:0070820_GOCC_TERTIARY_GRANULE                                                B2m,Hbb-bt,Qpct   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    0.126984          39  \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.500000          39  \n",
       " GO:0002475_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.416667          39  \n",
       " GO:0050830_GOBP_DEFENSE_RESPONSE_TO_GRAM_POSITI...  0.176471          39  \n",
       " GO:0002483_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.294118          39  \n",
       " ...                                                      ...         ...  \n",
       " GO:0002228_GOBP_NATURAL_KILLER_CELL_MEDIATED_IM...  0.074074          39  \n",
       " GO:0006471_GOBP_PROTEIN_ADP_RIBOSYLATION            0.074074          39  \n",
       " GO:0042743_GOBP_HYDROGEN_PEROXIDE_METABOLIC_PRO...  0.074074          39  \n",
       " GO:0016684_GOMF_OXIDOREDUCTASE_ACTIVITY_ACTING_...  0.074074          39  \n",
       " GO:0070820_GOCC_TERTIARY_GRANULE                    0.033333          39  \n",
       " \n",
       " [186 rows x 9 columns],\n",
       " 'T1D_DE_up_4_enrichment':                                                         pval    fdr  \\\n",
       " gene_set                                                              \n",
       " GO:0060322_GOBP_HEAD_DEVELOPMENT                    0.000006  0.057   \n",
       " GO:0007417_GOBP_CENTRAL_NERVOUS_SYSTEM_DEVELOPMENT  0.000021  0.094   \n",
       " GO:0090399_GOBP_REPLICATIVE_SENESCENCE              0.000034  0.100   \n",
       " R-HSA-5693554_REACTOME_RESOLUTION_OF_D_LOOP_STR...  0.000058  0.130   \n",
       " GO:0072331_GOBP_SIGNAL_TRANSDUCTION_BY_P53_CLAS...  0.000086  0.130   \n",
       " GO:0150076_GOBP_NEUROINFLAMMATORY_RESPONSE          0.000110  0.130   \n",
       " R-HSA-5693537_REACTOME_RESOLUTION_OF_D_LOOP_STR...  0.000110  0.130   \n",
       " GO:1901798_GOBP_POSITIVE_REGULATION_OF_SIGNAL_T...  0.000150  0.160   \n",
       " GO:0010259_GOBP_MULTICELLULAR_ORGANISM_AGING        0.000190  0.190   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0060322_GOBP_HEAD_DEVELOPMENT                           13      406   \n",
       " GO:0007417_GOBP_CENTRAL_NERVOUS_SYSTEM_DEVELOPMENT         13      500   \n",
       " GO:0090399_GOBP_REPLICATIVE_SENESCENCE                     13        7   \n",
       " R-HSA-5693554_REACTOME_RESOLUTION_OF_D_LOOP_STR...         13        9   \n",
       " GO:0072331_GOBP_SIGNAL_TRANSDUCTION_BY_P53_CLAS...         13      191   \n",
       " GO:0150076_GOBP_NEUROINFLAMMATORY_RESPONSE                 13       12   \n",
       " R-HSA-5693537_REACTOME_RESOLUTION_OF_D_LOOP_STR...         13       12   \n",
       " GO:1901798_GOBP_POSITIVE_REGULATION_OF_SIGNAL_T...         13       14   \n",
       " GO:0010259_GOBP_MULTICELLULAR_ORGANISM_AGING               13       16   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0060322_GOBP_HEAD_DEVELOPMENT                          6        9839   \n",
       " GO:0007417_GOBP_CENTRAL_NERVOUS_SYSTEM_DEVELOPMENT        6        9839   \n",
       " GO:0090399_GOBP_REPLICATIVE_SENESCENCE                    2        9839   \n",
       " R-HSA-5693554_REACTOME_RESOLUTION_OF_D_LOOP_STR...        2        9839   \n",
       " GO:0072331_GOBP_SIGNAL_TRANSDUCTION_BY_P53_CLAS...        4        9839   \n",
       " GO:0150076_GOBP_NEUROINFLAMMATORY_RESPONSE                2        9839   \n",
       " R-HSA-5693537_REACTOME_RESOLUTION_OF_D_LOOP_STR...        2        9839   \n",
       " GO:1901798_GOBP_POSITIVE_REGULATION_OF_SIGNAL_T...        2        9839   \n",
       " GO:0010259_GOBP_MULTICELLULAR_ORGANISM_AGING              2        9839   \n",
       " \n",
       "                                                                                  hits  \\\n",
       " gene_set                                                                                \n",
       " GO:0060322_GOBP_HEAD_DEVELOPMENT                    Atm,Chd5,Cttnbp2,Phgdh,Wrn,Zfp458   \n",
       " GO:0007417_GOBP_CENTRAL_NERVOUS_SYSTEM_DEVELOPMENT  Atm,Chd5,Cttnbp2,Phgdh,Wrn,Zfp458   \n",
       " GO:0090399_GOBP_REPLICATIVE_SENESCENCE                                        Atm,Wrn   \n",
       " R-HSA-5693554_REACTOME_RESOLUTION_OF_D_LOOP_STR...                            Atm,Wrn   \n",
       " GO:0072331_GOBP_SIGNAL_TRANSDUCTION_BY_P53_CLAS...                 Atm,Chd5,Nupr1,Wrn   \n",
       " GO:0150076_GOBP_NEUROINFLAMMATORY_RESPONSE                                  Atm,Nupr1   \n",
       " R-HSA-5693537_REACTOME_RESOLUTION_OF_D_LOOP_STR...                            Atm,Wrn   \n",
       " GO:1901798_GOBP_POSITIVE_REGULATION_OF_SIGNAL_T...                           Atm,Chd5   \n",
       " GO:0010259_GOBP_MULTICELLULAR_ORGANISM_AGING                                  Atm,Wrn   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0060322_GOBP_HEAD_DEVELOPMENT                    0.014778          13  \n",
       " GO:0007417_GOBP_CENTRAL_NERVOUS_SYSTEM_DEVELOPMENT  0.012000          13  \n",
       " GO:0090399_GOBP_REPLICATIVE_SENESCENCE              0.285714          13  \n",
       " R-HSA-5693554_REACTOME_RESOLUTION_OF_D_LOOP_STR...  0.222222          13  \n",
       " GO:0072331_GOBP_SIGNAL_TRANSDUCTION_BY_P53_CLAS...  0.020942          13  \n",
       " GO:0150076_GOBP_NEUROINFLAMMATORY_RESPONSE          0.166667          13  \n",
       " R-HSA-5693537_REACTOME_RESOLUTION_OF_D_LOOP_STR...  0.166667          13  \n",
       " GO:1901798_GOBP_POSITIVE_REGULATION_OF_SIGNAL_T...  0.142857          13  \n",
       " GO:0010259_GOBP_MULTICELLULAR_ORGANISM_AGING        0.125000          13  ,\n",
       " 'T2D_DE':                         pval      qval    log2fc        mean  zero_mean  \\\n",
       " gene                                                                      \n",
       " ENSMUSG00000051951  0.000011  0.000013  0.307300    0.066066      False   \n",
       " ENSMUSG00000033845  0.000000  0.000000  0.184043    0.861142      False   \n",
       " ENSMUSG00000025903  0.000000  0.000000 -0.455122    0.507621      False   \n",
       " ENSMUSG00000033813  0.000000  0.000000 -0.326792    1.148189      False   \n",
       " ENSMUSG00000033793  0.000000  0.000000 -0.338769    0.883926      False   \n",
       " ...                      ...       ...       ...         ...        ...   \n",
       " ENSMUSG00000064368  0.603101  0.614679  0.014795    0.528503      False   \n",
       " ENSMUSG00000064370  0.000000  0.000000 -0.434244  131.996829      False   \n",
       " ENSMUSG00000051412  0.000000  0.000000  0.239062    1.277671      False   \n",
       " ENSMUSG00000095041  0.000000  0.000000 -0.359222    1.117523      False   \n",
       " ENSMUSG00000063897  0.000000  0.000000  0.463859    0.902928      False   \n",
       " \n",
       "                             grad  coef_mle   coef_sd             ll  \\\n",
       " gene                                                                  \n",
       " ENSMUSG00000051951  4.337062e-08  0.307300  0.069771  -10674.083488   \n",
       " ENSMUSG00000033845  1.112048e-06  0.184043  0.020849  -54567.283849   \n",
       " ENSMUSG00000025903  1.739494e-06 -0.455122  0.024659  -42143.793666   \n",
       " ENSMUSG00000033813  9.631941e-07 -0.326792  0.017958  -62806.836425   \n",
       " ENSMUSG00000033793  1.142498e-06 -0.338769  0.019809  -56385.526102   \n",
       " ...                          ...       ...       ...            ...   \n",
       " ENSMUSG00000064368  6.308996e-07  0.014795  0.028455  -39320.472155   \n",
       " ENSMUSG00000064370  2.304797e-07 -0.434244  0.007934 -242922.730736   \n",
       " ENSMUSG00000051412  6.119591e-07  0.239062  0.018181  -64690.746931   \n",
       " ENSMUSG00000095041  8.585163e-07 -0.359222  0.018486  -62967.233150   \n",
       " ENSMUSG00000063897  3.424977e-07  0.463859  0.021482  -55690.567151   \n",
       " \n",
       "                        gene_symbol   hc  rel_beta_expr  N_PMID  \\\n",
       " gene                                                             \n",
       " ENSMUSG00000051951            Xkr4  NaN       1.000000    11.0   \n",
       " ENSMUSG00000033845          Mrpl15  NaN       0.577643    55.0   \n",
       " ENSMUSG00000025903          Lypla1  NaN       0.755665    42.0   \n",
       " ENSMUSG00000033813           Tcea1  NaN       0.898695    66.0   \n",
       " ENSMUSG00000033793         Atp6v1h  NaN       1.000000    65.0   \n",
       " ...                            ...  ...            ...     ...   \n",
       " ENSMUSG00000064368          mt-Nd6  NaN       1.000000    42.0   \n",
       " ENSMUSG00000064370         mt-Cytb  NaN       1.000000    39.0   \n",
       " ENSMUSG00000051412           Vamp7  NaN       1.000000     0.0   \n",
       " ENSMUSG00000095041      AC149090.1  NaN       1.000000     0.0   \n",
       " ENSMUSG00000063897  CAAA01118383.1  NaN       0.619154     0.0   \n",
       " \n",
       "                     N_PMID_pancreas_notCancerNonendo  mean_expr_in_expr_cells  \n",
       " gene                                                                           \n",
       " ENSMUSG00000051951                               0.0                 0.487778  \n",
       " ENSMUSG00000033845                               0.0                 0.665728  \n",
       " ENSMUSG00000025903                               0.0                 0.617923  \n",
       " ENSMUSG00000033813                               0.0                 0.794732  \n",
       " ENSMUSG00000033793                               1.0                 0.706451  \n",
       " ...                                              ...                      ...  \n",
       " ENSMUSG00000064368                               4.0                 0.684435  \n",
       " ENSMUSG00000064370                               5.0                 4.252210  \n",
       " ENSMUSG00000051412                               0.0                 0.746626  \n",
       " ENSMUSG00000095041                               0.0                 0.812761  \n",
       " ENSMUSG00000063897                               0.0                 0.673011  \n",
       " \n",
       " [11149 rows x 15 columns],\n",
       " 'T2D_DE_down_1_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'T2D_DE_down_2_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  6.000000e-11   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                2.400000e-07   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...  2.500000e-07   \n",
       " GO:0070431_GOBP_NUCLEOTIDE_BINDING_OLIGOMERIZAT...  3.300000e-06   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...  5.400000e-06   \n",
       " ...                                                          ...   \n",
       " GO:0043576_GOBP_REGULATION_OF_RESPIRATORY_GASEO...  3.100000e-03   \n",
       " GO:2001198_GOBP_REGULATION_OF_DENDRITIC_CELL_DI...  3.100000e-03   \n",
       " R-HSA-450341_REACTOME_ACTIVATION_OF_THE_AP_1_FA...  3.100000e-03   \n",
       " GO:0009409_GOBP_RESPONSE_TO_COLD                    3.200000e-03   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  3.300000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  5.500000e-07         95   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                7.600000e-04         95   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...  7.600000e-04         95   \n",
       " GO:0070431_GOBP_NUCLEOTIDE_BINDING_OLIGOMERIZAT...  7.700000e-03         95   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...  7.800000e-03         95   \n",
       " ...                                                          ...        ...   \n",
       " GO:0043576_GOBP_REGULATION_OF_RESPIRATORY_GASEO...  2.400000e-01         95   \n",
       " GO:2001198_GOBP_REGULATION_OF_DENDRITIC_CELL_DI...  2.400000e-01         95   \n",
       " R-HSA-450341_REACTOME_ACTIVATION_OF_THE_AP_1_FA...  2.400000e-01         95   \n",
       " GO:0009409_GOBP_RESPONSE_TO_COLD                    2.500000e-01         95   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  2.500000e-01         95   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...      225       16   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                    174       11   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...       32        6   \n",
       " GO:0070431_GOBP_NUCLEOTIDE_BINDING_OLIGOMERIZAT...       13        4   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...       53        6   \n",
       " ...                                                     ...      ...   \n",
       " GO:0043576_GOBP_REGULATION_OF_RESPIRATORY_GASEO...       10        2   \n",
       " GO:2001198_GOBP_REGULATION_OF_DENDRITIC_CELL_DI...       10        2   \n",
       " R-HSA-450341_REACTOME_ACTIVATION_OF_THE_AP_1_FA...       10        2   \n",
       " GO:0009409_GOBP_RESPONSE_TO_COLD                         35        3   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...       72        4   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...       11149   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                     11149   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...       11149   \n",
       " GO:0070431_GOBP_NUCLEOTIDE_BINDING_OLIGOMERIZAT...       11149   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...       11149   \n",
       " ...                                                        ...   \n",
       " GO:0043576_GOBP_REGULATION_OF_RESPIRATORY_GASEO...       11149   \n",
       " GO:2001198_GOBP_REGULATION_OF_DENDRITIC_CELL_DI...       11149   \n",
       " R-HSA-450341_REACTOME_ACTIVATION_OF_THE_AP_1_FA...       11149   \n",
       " GO:0009409_GOBP_RESPONSE_TO_COLD                         11149   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...       11149   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  Atf3,Cebpb,Ddit3,Egr1,Egr4,Fos,Fosb,Irf1,Jun,J...   \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                Ddit3,Dusp1,Fos,Gadd45b,Gadd45g,Hspa1a,Hspa1b,...   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...                        Arc,Egr1,Egr4,Fos,Fosb,Junb   \n",
       " GO:0070431_GOBP_NUCLEOTIDE_BINDING_OLIGOMERIZAT...                       Hspa1a,Hspa1b,Nfkbia,Tnfaip3   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...                        Arc,Egr1,Egr4,Fos,Fosb,Junb   \n",
       " ...                                                                                               ...   \n",
       " GO:0043576_GOBP_REGULATION_OF_RESPIRATORY_GASEO...                                       Nr4a2,Slc5a3   \n",
       " GO:2001198_GOBP_REGULATION_OF_DENDRITIC_CELL_DI...                                       Cebpb,H2-Q10   \n",
       " R-HSA-450341_REACTOME_ACTIVATION_OF_THE_AP_1_FA...                                            Fos,Jun   \n",
       " GO:0009409_GOBP_RESPONSE_TO_COLD                                                  Fos,Nfkbia,Ppargc1a   \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...                         Hspa1a,Hspa1b,Prr7,Tnfaip3   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  0.071111          95  \n",
       " hsa04010_KEGG_MAPK_SIGNALING_PATHWAY                0.063218          95  \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...  0.187500          95  \n",
       " GO:0070431_GOBP_NUCLEOTIDE_BINDING_OLIGOMERIZAT...  0.307692          95  \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...  0.113208          95  \n",
       " ...                                                      ...         ...  \n",
       " GO:0043576_GOBP_REGULATION_OF_RESPIRATORY_GASEO...  0.200000          95  \n",
       " GO:2001198_GOBP_REGULATION_OF_DENDRITIC_CELL_DI...  0.200000          95  \n",
       " R-HSA-450341_REACTOME_ACTIVATION_OF_THE_AP_1_FA...  0.200000          95  \n",
       " GO:0009409_GOBP_RESPONSE_TO_COLD                    0.085714          95  \n",
       " GO:1903321_GOBP_NEGATIVE_REGULATION_OF_PROTEIN_...  0.055556          95  \n",
       " \n",
       " [120 rows x 9 columns],\n",
       " 'T2D_DE_down_3_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      5.000000e-10  0.000003   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        5.400000e-10  0.000003   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   1.700000e-09  0.000005   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     3.200000e-09  0.000007   \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   5.500000e-09  0.000010   \n",
       " ...                                                          ...       ...   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...  2.100000e-03  0.220000   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY            2.100000e-03  0.230000   \n",
       " GO:0009065_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_CAT...  2.200000e-03  0.230000   \n",
       " GO:0045058_GOBP_T_CELL_SELECTION                    2.200000e-03  0.230000   \n",
       " GO:1901606_GOBP_ALPHA_AMINO_ACID_CATABOLIC_PROCESS  2.300000e-03  0.230000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                            216      295   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS              216      273   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                         216      200   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION           216      166   \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                         216      152   \n",
       " ...                                                       ...      ...   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...        216       47   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY                  216       69   \n",
       " GO:0009065_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_CAT...        216       14   \n",
       " GO:0045058_GOBP_T_CELL_SELECTION                          216       14   \n",
       " GO:1901606_GOBP_ALPHA_AMINO_ACID_CATABOLIC_PROCESS        216       48   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                           25       11149   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS             24       11149   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                        20       11149   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION          18       11149   \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                        17       11149   \n",
       " ...                                                     ...         ...   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...        5       11149   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY                  6       11149   \n",
       " GO:0009065_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_CAT...        3       11149   \n",
       " GO:0045058_GOBP_T_CELL_SELECTION                          3       11149   \n",
       " GO:1901606_GOBP_ALPHA_AMINO_ACID_CATABOLIC_PROCESS        5       11149   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      Agt,G6pc2,Gad1,Gipr,Glud1,Hadh,Hcar2,Kcnj11,Ma...   \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        Agt,Dhrs13,G6pc2,Gipr,Glud1,Hadh,Hcar2,Hsd17b7...   \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   Agt,G6pc2,Gipr,Glud1,Hadh,Hcar2,Kcnj11,Mafa,Mi...   \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     Agt,G6pc2,Gipr,Glud1,Hadh,Hcar2,Kcnj11,Midn,Ox...   \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   G6pc2,Gipr,Glud1,Hadh,Kcnj11,Mafa,Midn,Oxct1,P...   \n",
       " ...                                                                                               ...   \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...                    Hmgcr,Hmgcs1,Idi1,Srebf1,Srebf2   \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY                          Ccnd2,Jak3,Lifr,Stat3,Stat5a,Stat5b   \n",
       " GO:0009065_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_CAT...                                 Aldh4a1,Gad1,Glud1   \n",
       " GO:0045058_GOBP_T_CELL_SELECTION                                                    Itpkb,Ptpn2,Stat3   \n",
       " GO:1901606_GOBP_ALPHA_AMINO_ACID_CATABOLIC_PROCESS                    Acadsb,Aldh4a1,Gad1,Glud1,Mat1a   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0023061_GOBP_SIGNAL_RELEASE                      0.084746         216  \n",
       " GO:0010817_GOBP_REGULATION_OF_HORMONE_LEVELS        0.087912         216  \n",
       " GO:0009914_GOBP_HORMONE_TRANSPORT                   0.100000         216  \n",
       " GO:0046883_GOBP_REGULATION_OF_HORMONE_SECRETION     0.108434         216  \n",
       " GO:0030073_GOBP_INSULIN_SECRETION                   0.111842         216  \n",
       " ...                                                      ...         ...  \n",
       " GO:1902930_GOBP_REGULATION_OF_ALCOHOL_BIOSYNTHE...  0.106383         216  \n",
       " hsa04630_KEGG_JAK_STAT_SIGNALING_PATHWAY            0.086957         216  \n",
       " GO:0009065_GOBP_GLUTAMINE_FAMILY_AMINO_ACID_CAT...  0.214286         216  \n",
       " GO:0045058_GOBP_T_CELL_SELECTION                    0.214286         216  \n",
       " GO:1901606_GOBP_ALPHA_AMINO_ACID_CATABOLIC_PROCESS  0.104167         216  \n",
       " \n",
       " [89 rows x 9 columns],\n",
       " 'T2D_DE_down_4_enrichment':                                                         pval  fdr  signature  \\\n",
       " gene_set                                                                       \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...  0.000035  0.2        200   \n",
       " GO:0032793_GOBP_POSITIVE_REGULATION_OF_CREB_TRA...  0.000044  0.2        200   \n",
       " GO:0000904_GOBP_CELL_MORPHOGENESIS_INVOLVED_IN_...  0.000091  0.2        200   \n",
       " GO:1905603_GOBP_REGULATION_OF_BLOOD_BRAIN_BARRI...  0.000110  0.2        200   \n",
       " GO:0043534_GOBP_BLOOD_VESSEL_ENDOTHELIAL_CELL_M...  0.000170  0.2        200   \n",
       " GO:0002327_GOBP_IMMATURE_B_CELL_DIFFERENTIATION     0.000190  0.2        200   \n",
       " GO:0032990_GOBP_CELL_PART_MORPHOGENESIS             0.000210  0.2        200   \n",
       " GO:0040013_GOBP_NEGATIVE_REGULATION_OF_LOCOMOTION   0.000210  0.2        200   \n",
       " GO:0032989_GOBP_CELLULAR_COMPONENT_MORPHOGENESIS    0.000220  0.2        200   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT         0.000230  0.2        200   \n",
       " GO:0061028_GOBP_ESTABLISHMENT_OF_ENDOTHELIAL_BA...  0.000240  0.2        200   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...      275       16   \n",
       " GO:0032793_GOBP_POSITIVE_REGULATION_OF_CREB_TRA...       12        4   \n",
       " GO:0000904_GOBP_CELL_MORPHOGENESIS_INVOLVED_IN_...      430       20   \n",
       " GO:1905603_GOBP_REGULATION_OF_BLOOD_BRAIN_BARRI...        6        3   \n",
       " GO:0043534_GOBP_BLOOD_VESSEL_ENDOTHELIAL_CELL_M...       66        7   \n",
       " GO:0002327_GOBP_IMMATURE_B_CELL_DIFFERENTIATION           7        3   \n",
       " GO:0032990_GOBP_CELL_PART_MORPHOGENESIS                 423       19   \n",
       " GO:0040013_GOBP_NEGATIVE_REGULATION_OF_LOCOMOTION       167       11   \n",
       " GO:0032989_GOBP_CELLULAR_COMPONENT_MORPHOGENESIS        459       20   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT             116        9   \n",
       " GO:0061028_GOBP_ESTABLISHMENT_OF_ENDOTHELIAL_BA...       32        5   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...       11149   \n",
       " GO:0032793_GOBP_POSITIVE_REGULATION_OF_CREB_TRA...       11149   \n",
       " GO:0000904_GOBP_CELL_MORPHOGENESIS_INVOLVED_IN_...       11149   \n",
       " GO:1905603_GOBP_REGULATION_OF_BLOOD_BRAIN_BARRI...       11149   \n",
       " GO:0043534_GOBP_BLOOD_VESSEL_ENDOTHELIAL_CELL_M...       11149   \n",
       " GO:0002327_GOBP_IMMATURE_B_CELL_DIFFERENTIATION          11149   \n",
       " GO:0032990_GOBP_CELL_PART_MORPHOGENESIS                  11149   \n",
       " GO:0040013_GOBP_NEGATIVE_REGULATION_OF_LOCOMOTION        11149   \n",
       " GO:0032989_GOBP_CELLULAR_COMPONENT_MORPHOGENESIS         11149   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT              11149   \n",
       " GO:0061028_GOBP_ESTABLISHMENT_OF_ENDOTHELIAL_BA...       11149   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...  Alms1,Arrdc3,Grhl1,Homer1,Nr1d2,Per2,Prkab2,Sc...   \n",
       " GO:0032793_GOBP_POSITIVE_REGULATION_OF_CREB_TRA...                           Lrp8,Prkd1,Rps6ka5,Vegfa   \n",
       " GO:0000904_GOBP_CELL_MORPHOGENESIS_INVOLVED_IN_...  Abl1,Efna5,Ezr,Itgav,Lrp8,Myadm,Nefh,Olfm4,Par...   \n",
       " GO:1905603_GOBP_REGULATION_OF_BLOOD_BRAIN_BARRI...                                  Sh3gl2,Tjp1,Vegfa   \n",
       " GO:0043534_GOBP_BLOOD_VESSEL_ENDOTHELIAL_CELL_M...             Abl1,Plcg1,Prkd1,Robo1,Sirt1,Srf,Vegfa   \n",
       " GO:0002327_GOBP_IMMATURE_B_CELL_DIFFERENTIATION                                    Abl1,Fnip1,Irf2bp2   \n",
       " GO:0032990_GOBP_CELL_PART_MORPHOGENESIS             Abl1,Als2,Efna5,Ezr,Grip1,Lrp8,Myo9a,Nefh,Pard...   \n",
       " GO:0040013_GOBP_NEGATIVE_REGULATION_OF_LOCOMOTION   Abhd2,Acvr1c,Arrdc3,Dach1,Dusp10,Nav3,Osbpl8,R...   \n",
       " GO:0032989_GOBP_CELLULAR_COMPONENT_MORPHOGENESIS    Abl1,Als2,Efna5,Ezr,Grip1,Lrp8,Myo9a,Nefh,Pard...   \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT          Ezr,Fosl2,Gsk3a,Il6st,Myadm,Prox1,Tjp1,Vcl,Vegfa   \n",
       " GO:0061028_GOBP_ESTABLISHMENT_OF_ENDOTHELIAL_BA...                           Ezr,Myadm,Tjp1,Vcl,Vegfa   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0048871_GOBP_MULTICELLULAR_ORGANISMAL_HOMEOS...  0.058182         200  \n",
       " GO:0032793_GOBP_POSITIVE_REGULATION_OF_CREB_TRA...  0.333333         200  \n",
       " GO:0000904_GOBP_CELL_MORPHOGENESIS_INVOLVED_IN_...  0.046512         200  \n",
       " GO:1905603_GOBP_REGULATION_OF_BLOOD_BRAIN_BARRI...  0.500000         200  \n",
       " GO:0043534_GOBP_BLOOD_VESSEL_ENDOTHELIAL_CELL_M...  0.106061         200  \n",
       " GO:0002327_GOBP_IMMATURE_B_CELL_DIFFERENTIATION     0.428571         200  \n",
       " GO:0032990_GOBP_CELL_PART_MORPHOGENESIS             0.044917         200  \n",
       " GO:0040013_GOBP_NEGATIVE_REGULATION_OF_LOCOMOTION   0.065868         200  \n",
       " GO:0032989_GOBP_CELLULAR_COMPONENT_MORPHOGENESIS    0.043573         200  \n",
       " GO:0002064_GOBP_EPITHELIAL_CELL_DEVELOPMENT         0.077586         200  \n",
       " GO:0061028_GOBP_ESTABLISHMENT_OF_ENDOTHELIAL_BA...  0.156250         200  ,\n",
       " 'T2D_DE_down_5_enrichment':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   7.400000e-11   \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              4.100000e-10   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           6.600000e-10   \n",
       " GO:0060333_GOBP_INTERFERON_GAMMA_MEDIATED_SIGNA...  5.400000e-09   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          6.800000e-09   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        1.200000e-08   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       1.300000e-08   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  3.000000e-08   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    1.200000e-07   \n",
       " GO:0002831_GOBP_REGULATION_OF_RESPONSE_TO_BIOTI...  2.000000e-06   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  2.300000e-06   \n",
       " R-HSA-1280215_REACTOME_CYTOKINE_SIGNALING_IN_IM...  2.900000e-06   \n",
       " GO:0002684_GOBP_POSITIVE_REGULATION_OF_IMMUNE_S...  3.400000e-05   \n",
       " GO:0002833_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  6.300000e-05   \n",
       " GO:0032101_GOBP_REGULATION_OF_RESPONSE_TO_EXTER...  8.700000e-05   \n",
       " GO:0030224_GOBP_MONOCYTE_DIFFERENTIATION            8.800000e-05   \n",
       " GO:0050778_GOBP_POSITIVE_REGULATION_OF_IMMUNE_R...  1.100000e-04   \n",
       " GO:0050776_GOBP_REGULATION_OF_IMMUNE_RESPONSE       1.400000e-04   \n",
       " GO:0031347_GOBP_REGULATION_OF_DEFENSE_RESPONSE      1.600000e-04   \n",
       " GO:0032103_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  1.600000e-04   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...  2.000000e-04   \n",
       " GO:0045071_GOBP_NEGATIVE_REGULATION_OF_VIRAL_GE...  2.000000e-04   \n",
       " GO:0045088_GOBP_REGULATION_OF_INNATE_IMMUNE_RES...  2.600000e-04   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...  3.200000e-04   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS       3.200000e-04   \n",
       " GO:0002711_GOBP_POSITIVE_REGULATION_OF_T_CELL_M...  3.600000e-04   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                 4.400000e-04   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION        4.400000e-04   \n",
       " GO:0060760_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  5.200000e-04   \n",
       " GO:0002819_GOBP_REGULATION_OF_ADAPTIVE_IMMUNE_R...  5.900000e-04   \n",
       " GO:0019079_GOBP_VIRAL_GENOME_REPLICATION            5.900000e-04   \n",
       " GO:0039530_GOBP_MDA_5_SIGNALING_PATHWAY             5.900000e-04   \n",
       " GO:0002709_GOBP_REGULATION_OF_T_CELL_MEDIATED_I...  6.600000e-04   \n",
       " GO:0002484_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  7.500000e-04   \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION       9.400000e-04   \n",
       " GO:0002456_GOBP_T_CELL_MEDIATED_IMMUNITY            9.700000e-04   \n",
       " GO:0002699_GOBP_POSITIVE_REGULATION_OF_IMMUNE_E...  1.000000e-03   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  1.100000e-03   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  1.100000e-03   \n",
       " GO:0098586_GOBP_CELLULAR_RESPONSE_TO_VIRUS          1.100000e-03   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX         1.100000e-03   \n",
       " GO:0030881_GOMF_BETA_2_MICROGLOBULIN_BINDING        1.100000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   6.800000e-07         52   \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              1.900000e-06         52   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           2.000000e-06         52   \n",
       " GO:0060333_GOBP_INTERFERON_GAMMA_MEDIATED_SIGNA...  1.200000e-05         52   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          1.300000e-05         52   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        1.700000e-05         52   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       1.700000e-05         52   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  3.500000e-05         52   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    1.200000e-04         52   \n",
       " GO:0002831_GOBP_REGULATION_OF_RESPONSE_TO_BIOTI...  1.800000e-03         52   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  1.900000e-03         52   \n",
       " R-HSA-1280215_REACTOME_CYTOKINE_SIGNALING_IN_IM...  2.200000e-03         52   \n",
       " GO:0002684_GOBP_POSITIVE_REGULATION_OF_IMMUNE_S...  2.400000e-02         52   \n",
       " GO:0002833_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  4.200000e-02         52   \n",
       " GO:0032101_GOBP_REGULATION_OF_RESPONSE_TO_EXTER...  5.100000e-02         52   \n",
       " GO:0030224_GOBP_MONOCYTE_DIFFERENTIATION            5.100000e-02         52   \n",
       " GO:0050778_GOBP_POSITIVE_REGULATION_OF_IMMUNE_R...  6.100000e-02         52   \n",
       " GO:0050776_GOBP_REGULATION_OF_IMMUNE_RESPONSE       7.000000e-02         52   \n",
       " GO:0031347_GOBP_REGULATION_OF_DEFENSE_RESPONSE      7.400000e-02         52   \n",
       " GO:0032103_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  7.400000e-02         52   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...  8.600000e-02         52   \n",
       " GO:0045071_GOBP_NEGATIVE_REGULATION_OF_VIRAL_GE...  8.600000e-02         52   \n",
       " GO:0045088_GOBP_REGULATION_OF_INNATE_IMMUNE_RES...  1.000000e-01         52   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...  1.200000e-01         52   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS       1.200000e-01         52   \n",
       " GO:0002711_GOBP_POSITIVE_REGULATION_OF_T_CELL_M...  1.300000e-01         52   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                 1.500000e-01         52   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION        1.500000e-01         52   \n",
       " GO:0060760_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  1.600000e-01         52   \n",
       " GO:0002819_GOBP_REGULATION_OF_ADAPTIVE_IMMUNE_R...  1.700000e-01         52   \n",
       " GO:0019079_GOBP_VIRAL_GENOME_REPLICATION            1.700000e-01         52   \n",
       " GO:0039530_GOBP_MDA_5_SIGNALING_PATHWAY             1.700000e-01         52   \n",
       " GO:0002709_GOBP_REGULATION_OF_T_CELL_MEDIATED_I...  1.900000e-01         52   \n",
       " GO:0002484_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  2.000000e-01         52   \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION       2.500000e-01         52   \n",
       " GO:0002456_GOBP_T_CELL_MEDIATED_IMMUNITY            2.500000e-01         52   \n",
       " GO:0002699_GOBP_POSITIVE_REGULATION_OF_IMMUNE_E...  2.500000e-01         52   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  2.500000e-01         52   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  2.500000e-01         52   \n",
       " GO:0098586_GOBP_CELLULAR_RESPONSE_TO_VIRUS          2.500000e-01         52   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX         2.500000e-01         52   \n",
       " GO:0030881_GOMF_BETA_2_MICROGLOBULIN_BINDING        2.500000e-01         52   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                       197       12   \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE                  419       15   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS               139       10   \n",
       " GO:0060333_GOBP_INTERFERON_GAMMA_MEDIATED_SIGNA...       55        7   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING              129        9   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA             96        8   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON            62        7   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...       41        6   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING         51        6   \n",
       " GO:0002831_GOBP_REGULATION_OF_RESPONSE_TO_BIOTI...      250        9   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...      410       11   \n",
       " R-HSA-1280215_REACTOME_CYTOKINE_SIGNALING_IN_IM...      421       11   \n",
       " GO:0002684_GOBP_POSITIVE_REGULATION_OF_IMMUNE_S...      447       10   \n",
       " GO:0002833_GOBP_POSITIVE_REGULATION_OF_RESPONSE...      149        6   \n",
       " GO:0032101_GOBP_REGULATION_OF_RESPONSE_TO_EXTER...      500       10   \n",
       " GO:0030224_GOBP_MONOCYTE_DIFFERENTIATION                 19        3   \n",
       " GO:0050778_GOBP_POSITIVE_REGULATION_OF_IMMUNE_R...      323        8   \n",
       " GO:0050776_GOBP_REGULATION_OF_IMMUNE_RESPONSE           426        9   \n",
       " GO:0031347_GOBP_REGULATION_OF_DEFENSE_RESPONSE          340        8   \n",
       " GO:0032103_GOBP_POSITIVE_REGULATION_OF_RESPONSE...      253        7   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...      183        6   \n",
       " GO:0045071_GOBP_NEGATIVE_REGULATION_OF_VIRAL_GE...       25        3   \n",
       " GO:0045088_GOBP_REGULATION_OF_INNATE_IMMUNE_RES...      193        6   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...        6        2   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS             6        2   \n",
       " GO:0002711_GOBP_POSITIVE_REGULATION_OF_T_CELL_M...       30        3   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                       7        2   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION              7        2   \n",
       " GO:0060760_GOBP_POSITIVE_REGULATION_OF_RESPONSE...       34        3   \n",
       " GO:0002819_GOBP_REGULATION_OF_ADAPTIVE_IMMUNE_R...       83        4   \n",
       " GO:0019079_GOBP_VIRAL_GENOME_REPLICATION                 83        4   \n",
       " GO:0039530_GOBP_MDA_5_SIGNALING_PATHWAY                   8        2   \n",
       " GO:0002709_GOBP_REGULATION_OF_T_CELL_MEDIATED_I...       37        3   \n",
       " GO:0002484_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...        9        2   \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION            10        2   \n",
       " GO:0002456_GOBP_T_CELL_MEDIATED_IMMUNITY                 42        3   \n",
       " GO:0002699_GOBP_POSITIVE_REGULATION_OF_IMMUNE_E...       96        4   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       11        2   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       11        2   \n",
       " GO:0098586_GOBP_CELLULAR_RESPONSE_TO_VIRUS               44        3   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX              11        2   \n",
       " GO:0030881_GOMF_BETA_2_MICROGLOBULIN_BINDING             11        2   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                        11149   \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE                   11149   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS                11149   \n",
       " GO:0060333_GOBP_INTERFERON_GAMMA_MEDIATED_SIGNA...       11149   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING               11149   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA             11149   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON            11149   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...       11149   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING         11149   \n",
       " GO:0002831_GOBP_REGULATION_OF_RESPONSE_TO_BIOTI...       11149   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...       11149   \n",
       " R-HSA-1280215_REACTOME_CYTOKINE_SIGNALING_IN_IM...       11149   \n",
       " GO:0002684_GOBP_POSITIVE_REGULATION_OF_IMMUNE_S...       11149   \n",
       " GO:0002833_GOBP_POSITIVE_REGULATION_OF_RESPONSE...       11149   \n",
       " GO:0032101_GOBP_REGULATION_OF_RESPONSE_TO_EXTER...       11149   \n",
       " GO:0030224_GOBP_MONOCYTE_DIFFERENTIATION                 11149   \n",
       " GO:0050778_GOBP_POSITIVE_REGULATION_OF_IMMUNE_R...       11149   \n",
       " GO:0050776_GOBP_REGULATION_OF_IMMUNE_RESPONSE            11149   \n",
       " GO:0031347_GOBP_REGULATION_OF_DEFENSE_RESPONSE           11149   \n",
       " GO:0032103_GOBP_POSITIVE_REGULATION_OF_RESPONSE...       11149   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...       11149   \n",
       " GO:0045071_GOBP_NEGATIVE_REGULATION_OF_VIRAL_GE...       11149   \n",
       " GO:0045088_GOBP_REGULATION_OF_INNATE_IMMUNE_RES...       11149   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...       11149   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS            11149   \n",
       " GO:0002711_GOBP_POSITIVE_REGULATION_OF_T_CELL_M...       11149   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                      11149   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION             11149   \n",
       " GO:0060760_GOBP_POSITIVE_REGULATION_OF_RESPONSE...       11149   \n",
       " GO:0002819_GOBP_REGULATION_OF_ADAPTIVE_IMMUNE_R...       11149   \n",
       " GO:0019079_GOBP_VIRAL_GENOME_REPLICATION                 11149   \n",
       " GO:0039530_GOBP_MDA_5_SIGNALING_PATHWAY                  11149   \n",
       " GO:0002709_GOBP_REGULATION_OF_T_CELL_MEDIATED_I...       11149   \n",
       " GO:0002484_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       11149   \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION            11149   \n",
       " GO:0002456_GOBP_T_CELL_MEDIATED_IMMUNITY                 11149   \n",
       " GO:0002699_GOBP_POSITIVE_REGULATION_OF_IMMUNE_E...       11149   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       11149   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...       11149   \n",
       " GO:0098586_GOBP_CELLULAR_RESPONSE_TO_VIRUS               11149   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX              11149   \n",
       " GO:0030881_GOMF_BETA_2_MICROGLOBULIN_BINDING             11149   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   Btbd17,Cxcl10,Ddx60,Herc6,Ifit1,Irf7,Irf9,Irgm...   \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              Ddx60,Gbp7,Herc6,H2-Q7,H2-Q4,Ifit1,Irf7,Irf9,I...   \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           Cxcl10,Ddx60,Herc6,Ifit1,Irf7,Irf9,Plscr1,Rsad...   \n",
       " GO:0060333_GOBP_INTERFERON_GAMMA_MEDIATED_SIGNA...             H2-Q7,H2-Q4,Irf7,Irf9,Irgm2,Mt2,Parp14   \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING           Gbp7,Herc6,H2-Q7,H2-Q4,Irf7,Irf9,Mt2,Rsad2,Stat2   \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA              Gbp7,H2-Q7,H2-Q4,Irf7,Irf9,Irgm2,Mt2,Parp14   \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON                 H2-Q7,H2-Q4,Ifit1,Irf7,Irf9,Rsad2,Stat2   \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...                  H2-Q7,H2-Q4,Irf7,Irf9,Rsad2,Stat2   \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING                       Gbp7,H2-Q7,H2-Q4,Irf7,Irf9,Mt2   \n",
       " GO:0002831_GOBP_REGULATION_OF_RESPONSE_TO_BIOTI...  Ddx60,Herc6,H2-Q7,Ifit1,Irf7,Irgm2,Parp14,Plsc...   \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  Cxcl10,H2-Q7,H2-Q4,Irf7,Irf9,Irgm2,Mt2,Parp14,...   \n",
       " R-HSA-1280215_REACTOME_CYTOKINE_SIGNALING_IN_IM...  Cxcl10,Gbp7,Herc6,H2-Q7,H2-Q4,Irf7,Irf9,Mt2,Ps...   \n",
       " GO:0002684_GOBP_POSITIVE_REGULATION_OF_IMMUNE_S...  Baiap2,Cxcl10,Ddx60,H2-Q7,H2-Q4,Irf7,Irgm2,Pls...   \n",
       " GO:0002833_GOBP_POSITIVE_REGULATION_OF_RESPONSE...                Ddx60,H2-Q7,Irf7,Irgm2,Plscr1,Psmb9   \n",
       " GO:0032101_GOBP_REGULATION_OF_RESPONSE_TO_EXTER...  Cxcl10,Ddx60,Herc6,H2-Q7,Ifit1,Irf7,Irgm2,Parp...   \n",
       " GO:0030224_GOBP_MONOCYTE_DIFFERENTIATION                                                 Irf7,Mt1,Mt2   \n",
       " GO:0050778_GOBP_POSITIVE_REGULATION_OF_IMMUNE_R...   Baiap2,H2-Q7,H2-Q4,Irf7,Irgm2,Plscr1,Psmb9,Rsad2   \n",
       " GO:0050776_GOBP_REGULATION_OF_IMMUNE_RESPONSE       Baiap2,H2-Q7,H2-Q4,Irf7,Irgm2,Parp14,Plscr1,Ps...   \n",
       " GO:0031347_GOBP_REGULATION_OF_DEFENSE_RESPONSE       Herc6,H2-Q7,Ifit1,Irf7,Irgm2,Parp14,Plscr1,Psmb9   \n",
       " GO:0032103_GOBP_POSITIVE_REGULATION_OF_RESPONSE...         Cxcl10,Ddx60,H2-Q7,Irf7,Irgm2,Plscr1,Psmb9   \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...                Ddx60,Herc6,H2-Q7,H2-Q4,Ifit1,Rsad2   \n",
       " GO:0045071_GOBP_NEGATIVE_REGULATION_OF_VIRAL_GE...                                Btbd17,Plscr1,Rsad2   \n",
       " GO:0045088_GOBP_REGULATION_OF_INNATE_IMMUNE_RES...               H2-Q7,Irf7,Irgm2,Parp14,Plscr1,Psmb9   \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...                                            Mt1,Mt2   \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS                                                 Mt1,Mt2   \n",
       " GO:0002711_GOBP_POSITIVE_REGULATION_OF_T_CELL_M...                                  H2-Q7,H2-Q4,Rsad2   \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                                                           Mt1,Mt2   \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION                                                  Mt1,Mt2   \n",
       " GO:0060760_GOBP_POSITIVE_REGULATION_OF_RESPONSE...                                  Irf7,Irgm2,Parp14   \n",
       " GO:0002819_GOBP_REGULATION_OF_ADAPTIVE_IMMUNE_R...                             H2-Q7,H2-Q4,Irf7,Rsad2   \n",
       " GO:0019079_GOBP_VIRAL_GENOME_REPLICATION                                  Btbd17,Plscr1,Rsad2,Smarcb1   \n",
       " GO:0039530_GOBP_MDA_5_SIGNALING_PATHWAY                                                    Ddx60,Irf7   \n",
       " GO:0002709_GOBP_REGULATION_OF_T_CELL_MEDIATED_I...                                  H2-Q7,H2-Q4,Rsad2   \n",
       " GO:0002484_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...                                        H2-Q7,H2-Q4   \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION                                                 Mt1,Mt2   \n",
       " GO:0002456_GOBP_T_CELL_MEDIATED_IMMUNITY                                            H2-Q7,H2-Q4,Rsad2   \n",
       " GO:0002699_GOBP_POSITIVE_REGULATION_OF_IMMUNE_E...                            Ddx60,H2-Q7,H2-Q4,Rsad2   \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...                                        H2-Q7,H2-Q4   \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...                                        H2-Q7,H2-Q4   \n",
       " GO:0098586_GOBP_CELLULAR_RESPONSE_TO_VIRUS                                           Ddx60,Irf7,Irgm2   \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX                                               H2-Q7,H2-Q4   \n",
       " GO:0030881_GOMF_BETA_2_MICROGLOBULIN_BINDING                                              H2-Q7,H2-Q4   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0009615_GOBP_RESPONSE_TO_VIRUS                   0.060914          52  \n",
       " GO:0045087_GOBP_INNATE_IMMUNE_RESPONSE              0.035800          52  \n",
       " GO:0051607_GOBP_DEFENSE_RESPONSE_TO_VIRUS           0.071942          52  \n",
       " GO:0060333_GOBP_INTERFERON_GAMMA_MEDIATED_SIGNA...  0.127273          52  \n",
       " R-HSA-913531_REACTOME_INTERFERON_SIGNALING          0.069767          52  \n",
       " GO:0034341_GOBP_RESPONSE_TO_INTERFERON_GAMMA        0.083333          52  \n",
       " GO:0034340_GOBP_RESPONSE_TO_TYPE_I_INTERFERON       0.112903          52  \n",
       " R-HSA-909733_REACTOME_INTERFERON_ALPHA_BETA_SIG...  0.146341          52  \n",
       " R-HSA-877300_REACTOME_INTERFERON_GAMMA_SIGNALING    0.117647          52  \n",
       " GO:0002831_GOBP_REGULATION_OF_RESPONSE_TO_BIOTI...  0.036000          52  \n",
       " GO:0019221_GOBP_CYTOKINE_MEDIATED_SIGNALING_PAT...  0.026829          52  \n",
       " R-HSA-1280215_REACTOME_CYTOKINE_SIGNALING_IN_IM...  0.026128          52  \n",
       " GO:0002684_GOBP_POSITIVE_REGULATION_OF_IMMUNE_S...  0.022371          52  \n",
       " GO:0002833_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  0.040268          52  \n",
       " GO:0032101_GOBP_REGULATION_OF_RESPONSE_TO_EXTER...  0.020000          52  \n",
       " GO:0030224_GOBP_MONOCYTE_DIFFERENTIATION            0.157895          52  \n",
       " GO:0050778_GOBP_POSITIVE_REGULATION_OF_IMMUNE_R...  0.024768          52  \n",
       " GO:0050776_GOBP_REGULATION_OF_IMMUNE_RESPONSE       0.021127          52  \n",
       " GO:0031347_GOBP_REGULATION_OF_DEFENSE_RESPONSE      0.023529          52  \n",
       " GO:0032103_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  0.027668          52  \n",
       " GO:0002697_GOBP_REGULATION_OF_IMMUNE_EFFECTOR_P...  0.032787          52  \n",
       " GO:0045071_GOBP_NEGATIVE_REGULATION_OF_VIRAL_GE...  0.120000          52  \n",
       " GO:0045088_GOBP_REGULATION_OF_INNATE_IMMUNE_RES...  0.031088          52  \n",
       " GO:0061687_GOBP_DETOXIFICATION_OF_INORGANIC_COM...  0.333333          52  \n",
       " R-HSA-5660526_REACTOME_RESPONSE_TO_METAL_IONS       0.333333          52  \n",
       " GO:0002711_GOBP_POSITIVE_REGULATION_OF_T_CELL_M...  0.100000          52  \n",
       " GO:0042117_GOBP_MONOCYTE_ACTIVATION                 0.285714          52  \n",
       " GO:0097501_GOBP_STRESS_RESPONSE_TO_METAL_ION        0.285714          52  \n",
       " GO:0060760_GOBP_POSITIVE_REGULATION_OF_RESPONSE...  0.088235          52  \n",
       " GO:0002819_GOBP_REGULATION_OF_ADAPTIVE_IMMUNE_R...  0.048193          52  \n",
       " GO:0019079_GOBP_VIRAL_GENOME_REPLICATION            0.048193          52  \n",
       " GO:0039530_GOBP_MDA_5_SIGNALING_PATHWAY             0.250000          52  \n",
       " GO:0002709_GOBP_REGULATION_OF_T_CELL_MEDIATED_I...  0.081081          52  \n",
       " GO:0002484_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.222222          52  \n",
       " GO:0071294_GOBP_CELLULAR_RESPONSE_TO_ZINC_ION       0.200000          52  \n",
       " GO:0002456_GOBP_T_CELL_MEDIATED_IMMUNITY            0.071429          52  \n",
       " GO:0002699_GOBP_POSITIVE_REGULATION_OF_IMMUNE_E...  0.041667          52  \n",
       " GO:0002428_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.181818          52  \n",
       " GO:0002480_GOBP_ANTIGEN_PROCESSING_AND_PRESENTA...  0.181818          52  \n",
       " GO:0098586_GOBP_CELLULAR_RESPONSE_TO_VIRUS          0.068182          52  \n",
       " GO:0042612_GOCC_MHC_CLASS_I_PROTEIN_COMPLEX         0.181818          52  \n",
       " GO:0030881_GOMF_BETA_2_MICROGLOBULIN_BINDING        0.181818          52  ,\n",
       " 'T2D_DE_down_6_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'T2D_DE_down_7_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'T2D_DE_up_1_enrichment':                                                             pval      fdr  \\\n",
       " gene_set                                                                    \n",
       " GO:0007186_GOBP_G_PROTEIN_COUPLED_RECEPTOR_SIGN...  2.400000e-08  0.00022   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING           7.400000e-08  0.00034   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                    1.800000e-07  0.00056   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY       2.000000e-06  0.00440   \n",
       " GO:0007189_GOBP_ADENYLATE_CYCLASE_ACTIVATING_G_...  2.400000e-06  0.00440   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...  2.900000e-06  0.00450   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...  8.600000e-06  0.01100   \n",
       " R-HSA-372790_REACTOME_SIGNALING_BY_GPCR             1.400000e-05  0.01700   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         2.600000e-05  0.02500   \n",
       " GO:0007188_GOBP_ADENYLATE_CYCLASE_MODULATING_G_...  2.700000e-05  0.02500   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                    3.000000e-05  0.02500   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY      9.300000e-05  0.07200   \n",
       " GO:0050819_GOBP_NEGATIVE_REGULATION_OF_COAGULATION  1.700000e-04  0.12000   \n",
       " GO:0051917_GOBP_REGULATION_OF_FIBRINOLYSIS          2.200000e-04  0.13000   \n",
       " GO:0051918_GOBP_NEGATIVE_REGULATION_OF_FIBRINOL...  2.200000e-04  0.13000   \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING         2.200000e-04  0.13000   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...  4.600000e-04  0.24000   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...  4.800000e-04  0.24000   \n",
       " GO:0050818_GOBP_REGULATION_OF_COAGULATION           5.000000e-04  0.24000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0007186_GOBP_G_PROTEIN_COUPLED_RECEPTOR_SIGN...         53      320   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING                  53       78   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                           53       28   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY              53        6   \n",
       " GO:0007189_GOBP_ADENYLATE_CYCLASE_ACTIVATING_G_...         53       46   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...         53       48   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...         53       28   \n",
       " R-HSA-372790_REACTOME_SIGNALING_BY_GPCR                    53      237   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY                53      125   \n",
       " GO:0007188_GOBP_ADENYLATE_CYCLASE_MODULATING_G_...         53       75   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                           53       38   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY             53       19   \n",
       " GO:0050819_GOBP_NEGATIVE_REGULATION_OF_COAGULATION         53       23   \n",
       " GO:0051917_GOBP_REGULATION_OF_FIBRINOLYSIS                 53        5   \n",
       " GO:0051918_GOBP_NEGATIVE_REGULATION_OF_FIBRINOL...         53        5   \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING                53       63   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...         53       32   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...         53       77   \n",
       " GO:0050818_GOBP_REGULATION_OF_COAGULATION                  53       33   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0007186_GOBP_G_PROTEIN_COUPLED_RECEPTOR_SIGN...       12       11149   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING                 7       11149   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                          5       11149   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY             3       11149   \n",
       " GO:0007189_GOBP_ADENYLATE_CYCLASE_ACTIVATING_G_...        5       11149   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...        5       11149   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...        4       11149   \n",
       " R-HSA-372790_REACTOME_SIGNALING_BY_GPCR                   8       11149   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY               6       11149   \n",
       " GO:0007188_GOBP_ADENYLATE_CYCLASE_MODULATING_G_...        5       11149   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                          4       11149   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY            3       11149   \n",
       " GO:0050819_GOBP_NEGATIVE_REGULATION_OF_COAGULATION        3       11149   \n",
       " GO:0051917_GOBP_REGULATION_OF_FIBRINOLYSIS                2       11149   \n",
       " GO:0051918_GOBP_NEGATIVE_REGULATION_OF_FIBRINOL...        2       11149   \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING               4       11149   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...        3       11149   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...        4       11149   \n",
       " GO:0050818_GOBP_REGULATION_OF_COAGULATION                 3       11149   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0007186_GOBP_G_PROTEIN_COUPLED_RECEPTOR_SIGN...  Cartpt,F2,Gcg,Gpr119,Gpr179,Pde11a,Ppy,Ptger3,...   \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING                             F2,Gcg,Ppy,Ptger3,Pyy,Ramp1,Sst   \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                                               Cartpt,Gcg,Ppy,Pyy,Sst   \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY                                          Cartpt,Ppy,Pyy   \n",
       " GO:0007189_GOBP_ADENYLATE_CYCLASE_ACTIVATING_G_...                       Gcg,Gpr119,Ptger3,Ramp1,Rit2   \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...                              F2,Ppy,Ptger3,Pyy,Sst   \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...                                     F2,Ppy,Pyy,Sst   \n",
       " R-HSA-372790_REACTOME_SIGNALING_BY_GPCR                        F2,Gcg,Pde11a,Ppy,Ptger3,Pyy,Ramp1,Sst   \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY                                 Cartpt,F2,Gcg,Ppy,Pyy,Sst   \n",
       " GO:0007188_GOBP_ADENYLATE_CYCLASE_MODULATING_G_...                       Gcg,Gpr119,Ptger3,Ramp1,Rit2   \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                                                   Cartpt,Gcg,Ppy,Pyy   \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY                                         Cartpt,Ppy,Pyy   \n",
       " GO:0050819_GOBP_NEGATIVE_REGULATION_OF_COAGULATION                                     Cpb2,F2,Tspan8   \n",
       " GO:0051917_GOBP_REGULATION_OF_FIBRINOLYSIS                                                    Cpb2,F2   \n",
       " GO:0051918_GOBP_NEGATIVE_REGULATION_OF_FIBRINOL...                                            Cpb2,F2   \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING                                  Cpb2,F2,Tnfrsf12a,Tspan8   \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...                                     Cpb2,F2,Tspan8   \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...                                Aldob,Fbp2,Gcg,Rbp4   \n",
       " GO:0050818_GOBP_REGULATION_OF_COAGULATION                                              Cpb2,F2,Tspan8   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0007186_GOBP_G_PROTEIN_COUPLED_RECEPTOR_SIGN...  0.037500          53  \n",
       " R-HSA-500792_REACTOME_GPCR_LIGAND_BINDING           0.089744          53  \n",
       " GO:0005179_GOMF_HORMONE_ACTIVITY                    0.178571          53  \n",
       " GO:0005184_GOMF_NEUROPEPTIDE_HORMONE_ACTIVITY       0.500000          53  \n",
       " GO:0007189_GOBP_ADENYLATE_CYCLASE_ACTIVATING_G_...  0.108696          53  \n",
       " R-HSA-373076_REACTOME_CLASS_A_1_RHODOPSIN_LIKE_...  0.104167          53  \n",
       " R-HSA-375276_REACTOME_PEPTIDE_LIGAND_BINDING_RE...  0.142857          53  \n",
       " R-HSA-372790_REACTOME_SIGNALING_BY_GPCR             0.033755          53  \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         0.048000          53  \n",
       " GO:0007188_GOBP_ADENYLATE_CYCLASE_MODULATING_G_...  0.066667          53  \n",
       " GO:0007631_GOBP_FEEDING_BEHAVIOR                    0.105263          53  \n",
       " GO:0007218_GOBP_NEUROPEPTIDE_SIGNALING_PATHWAY      0.157895          53  \n",
       " GO:0050819_GOBP_NEGATIVE_REGULATION_OF_COAGULATION  0.130435          53  \n",
       " GO:0051917_GOBP_REGULATION_OF_FIBRINOLYSIS          0.400000          53  \n",
       " GO:0051918_GOBP_NEGATIVE_REGULATION_OF_FIBRINOL...  0.400000          53  \n",
       " GO:0061041_GOBP_REGULATION_OF_WOUND_HEALING         0.063492          53  \n",
       " GO:0061045_GOBP_NEGATIVE_REGULATION_OF_WOUND_HE...  0.093750          53  \n",
       " GO:0046364_GOBP_MONOSACCHARIDE_BIOSYNTHETIC_PRO...  0.051948          53  \n",
       " GO:0050818_GOBP_REGULATION_OF_COAGULATION           0.090909          53  ,\n",
       " 'T2D_DE_up_2_enrichment':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         1.000000e-08  0.000097   \n",
       " GO:0098793_GOCC_PRESYNAPSE                          3.500000e-06  0.011000   \n",
       " GO:0050877_GOBP_NERVOUS_SYSTEM_PROCESS              3.600000e-06  0.011000   \n",
       " GO:0099560_GOBP_SYNAPTIC_MEMBRANE_ADHESION          6.500000e-06  0.015000   \n",
       " GO:0030312_GOCC_EXTERNAL_ENCAPSULATING_STRUCTURE    9.700000e-06  0.018000   \n",
       " ...                                                          ...       ...   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE         1.900000e-03  0.240000   \n",
       " GO:0010977_GOBP_NEGATIVE_REGULATION_OF_NEURON_P...  1.900000e-03  0.240000   \n",
       " GO:0050803_GOBP_REGULATION_OF_SYNAPSE_STRUCTURE...  1.900000e-03  0.240000   \n",
       " GO:0070252_GOBP_ACTIN_MEDIATED_CELL_CONTRACTION     1.900000e-03  0.240000   \n",
       " GO:0098839_GOCC_POSTSYNAPTIC_DENSITY_MEMBRANE       2.100000e-03  0.250000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY               349      125   \n",
       " GO:0098793_GOCC_PRESYNAPSE                                349      305   \n",
       " GO:0050877_GOBP_NERVOUS_SYSTEM_PROCESS                    349      424   \n",
       " GO:0099560_GOBP_SYNAPTIC_MEMBRANE_ADHESION                349       10   \n",
       " GO:0030312_GOCC_EXTERNAL_ENCAPSULATING_STRUCTURE          349      159   \n",
       " ...                                                       ...      ...   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE               349      125   \n",
       " GO:0010977_GOBP_NEGATIVE_REGULATION_OF_NEURON_P...        349       73   \n",
       " GO:0050803_GOBP_REGULATION_OF_SYNAPSE_STRUCTURE...        349      125   \n",
       " GO:0070252_GOBP_ACTIN_MEDIATED_CELL_CONTRACTION           349       57   \n",
       " GO:0098839_GOCC_POSTSYNAPTIC_DENSITY_MEMBRANE             349       43   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY              19       11149   \n",
       " GO:0098793_GOCC_PRESYNAPSE                               26       11149   \n",
       " GO:0050877_GOBP_NERVOUS_SYSTEM_PROCESS                   32       11149   \n",
       " GO:0099560_GOBP_SYNAPTIC_MEMBRANE_ADHESION                5       11149   \n",
       " GO:0030312_GOCC_EXTERNAL_ENCAPSULATING_STRUCTURE         17       11149   \n",
       " ...                                                     ...         ...   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE              11       11149   \n",
       " GO:0010977_GOBP_NEGATIVE_REGULATION_OF_NEURON_P...        8       11149   \n",
       " GO:0050803_GOBP_REGULATION_OF_SYNAPSE_STRUCTURE...       11       11149   \n",
       " GO:0070252_GOBP_ACTIN_MEDIATED_CELL_CONTRACTION           7       11149   \n",
       " GO:0098839_GOCC_POSTSYNAPTIC_DENSITY_MEMBRANE             6       11149   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         Bdnf,Bmp3,Calca,Cck,Ccl28,Crlf1,Fam3c,Fgf1,Fgf...   \n",
       " GO:0098793_GOCC_PRESYNAPSE                          Bdnf,Cadps2,Calb1,Calca,Cck,Clcn5,Gpc4,Hap1,Kc...   \n",
       " GO:0050877_GOBP_NERVOUS_SYSTEM_PROCESS              Aldh1a3,Aqp4,Asic2,Bdnf,Calb1,Calca,Cck,Cst3,D...   \n",
       " GO:0099560_GOBP_SYNAPTIC_MEMBRANE_ADHESION                             Gpc4,Mdga1,Nlgn1,Ptprs,Slitrk3   \n",
       " GO:0030312_GOCC_EXTERNAL_ENCAPSULATING_STRUCTURE    Ache,Bcan,Cd151,Col27a1,Cthrc1,Ctsz,Fgf1,Gpc4,...   \n",
       " ...                                                                                               ...   \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE         Gstm5,Gsto2,Gucy2c,Mgst3,Mt3,Mtarc2,Pon2,Pon3,...   \n",
       " GO:0010977_GOBP_NEGATIVE_REGULATION_OF_NEURON_P...     Itm2c,Mt3,Nlgn1,Ptprg,Ptprs,Sema4g,Spp1,Zfp365   \n",
       " GO:0050803_GOBP_REGULATION_OF_SYNAPSE_STRUCTURE...  Asic2,Bdnf,C1ql3,Dbn1,Dgkb,Gpc4,Mdga1,Nlgn1,Pt...   \n",
       " GO:0070252_GOBP_ACTIN_MEDIATED_CELL_CONTRACTION               Epdr1,Fgf12,Kcnj3,Myl4,Pde4d,Ryr2,Scn3b   \n",
       " GO:0098839_GOCC_POSTSYNAPTIC_DENSITY_MEMBRANE                   Lin7a,Nlgn1,Ptprs,Rgs9,Slitrk3,Sorcs2   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0030545_GOMF_RECEPTOR_REGULATOR_ACTIVITY         0.152000         349  \n",
       " GO:0098793_GOCC_PRESYNAPSE                          0.085246         349  \n",
       " GO:0050877_GOBP_NERVOUS_SYSTEM_PROCESS              0.075472         349  \n",
       " GO:0099560_GOBP_SYNAPTIC_MEMBRANE_ADHESION          0.500000         349  \n",
       " GO:0030312_GOCC_EXTERNAL_ENCAPSULATING_STRUCTURE    0.106918         349  \n",
       " ...                                                      ...         ...  \n",
       " GO:0009636_GOBP_RESPONSE_TO_TOXIC_SUBSTANCE         0.088000         349  \n",
       " GO:0010977_GOBP_NEGATIVE_REGULATION_OF_NEURON_P...  0.109589         349  \n",
       " GO:0050803_GOBP_REGULATION_OF_SYNAPSE_STRUCTURE...  0.088000         349  \n",
       " GO:0070252_GOBP_ACTIN_MEDIATED_CELL_CONTRACTION     0.122807         349  \n",
       " GO:0098839_GOCC_POSTSYNAPTIC_DENSITY_MEMBRANE       0.139535         349  \n",
       " \n",
       " [75 rows x 9 columns],\n",
       " 'T2D_DE_up_3_enrichment': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'sharedT1DT2D_DE_enrichment_up': Empty DataFrame\n",
       " Columns: [pval, fdr, signature, geneset, overlap, background, hits]\n",
       " Index: [],\n",
       " 'sharedT1DT2D_DE_enrichment_down':                                                             pval  \\\n",
       " gene_set                                                           \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  1.700000e-12   \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  1.300000e-10   \n",
       " GO:0003700_GOMF_DNA_BINDING_TRANSCRIPTION_FACTO...  3.400000e-09   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...  6.700000e-09   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...  2.000000e-07   \n",
       " ...                                                          ...   \n",
       " GO:0048514_GOBP_BLOOD_VESSEL_MORPHOGENESIS          6.900000e-03   \n",
       " GO:0007626_GOBP_LOCOMOTORY_BEHAVIOR                 7.000000e-03   \n",
       " GO:0051402_GOBP_NEURON_APOPTOTIC_PROCESS            7.000000e-03   \n",
       " GO:0003682_GOMF_CHROMATIN_BINDING                   7.200000e-03   \n",
       " GO:1990845_GOBP_ADAPTIVE_THERMOGENESIS              7.300000e-03   \n",
       " \n",
       "                                                              fdr  signature  \\\n",
       " gene_set                                                                      \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  1.500000e-08         89   \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  5.600000e-07         89   \n",
       " GO:0003700_GOMF_DNA_BINDING_TRANSCRIPTION_FACTO...  9.900000e-06         89   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...  1.500000e-05         89   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...  3.500000e-04         89   \n",
       " ...                                                          ...        ...   \n",
       " GO:0048514_GOBP_BLOOD_VESSEL_MORPHOGENESIS          2.500000e-01         89   \n",
       " GO:0007626_GOBP_LOCOMOTORY_BEHAVIOR                 2.500000e-01         89   \n",
       " GO:0051402_GOBP_NEURON_APOPTOTIC_PROCESS            2.500000e-01         89   \n",
       " GO:0003682_GOMF_CHROMATIN_BINDING                   2.500000e-01         89   \n",
       " GO:1990845_GOBP_ADAPTIVE_THERMOGENESIS              2.500000e-01         89   \n",
       " \n",
       "                                                     geneset  overlap  \\\n",
       " gene_set                                                               \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...      192       17   \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...      457       22   \n",
       " GO:0003700_GOMF_DNA_BINDING_TRANSCRIPTION_FACTO...      493       21   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...       29        7   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...       46        7   \n",
       " ...                                                     ...      ...   \n",
       " GO:0048514_GOBP_BLOOD_VESSEL_MORPHOGENESIS              238        7   \n",
       " GO:0007626_GOBP_LOCOMOTORY_BEHAVIOR                      81        4   \n",
       " GO:0051402_GOBP_NEURON_APOPTOTIC_PROCESS                128        5   \n",
       " GO:0003682_GOMF_CHROMATIN_BINDING                       366        9   \n",
       " GO:1990845_GOBP_ADAPTIVE_THERMOGENESIS                   82        4   \n",
       " \n",
       "                                                     background  \\\n",
       " gene_set                                                         \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...        9481   \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...        9481   \n",
       " GO:0003700_GOMF_DNA_BINDING_TRANSCRIPTION_FACTO...        9481   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...        9481   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...        9481   \n",
       " ...                                                        ...   \n",
       " GO:0048514_GOBP_BLOOD_VESSEL_MORPHOGENESIS                9481   \n",
       " GO:0007626_GOBP_LOCOMOTORY_BEHAVIOR                       9481   \n",
       " GO:0051402_GOBP_NEURON_APOPTOTIC_PROCESS                  9481   \n",
       " GO:0003682_GOMF_CHROMATIN_BINDING                         9481   \n",
       " GO:1990845_GOBP_ADAPTIVE_THERMOGENESIS                    9481   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  Atf3,Csrnp1,Dbp,Ddit3,Egr1,Egr4,Fos,Fosb,Fosl2...   \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  Atf3,Chd7,Dbp,Ddit3,Egr1,Egr4,Fos,Fosb,Fosl2,J...   \n",
       " GO:0003700_GOMF_DNA_BINDING_TRANSCRIPTION_FACTO...  Atf3,Csrnp1,Dbp,Ddit3,Egr1,Egr4,Fos,Fosb,Fosl2...   \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...                    Arc,Egr1,Egr4,Fos,Fosb,Junb,Srf   \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...                    Arc,Egr1,Egr4,Fos,Fosb,Junb,Srf   \n",
       " ...                                                                                               ...   \n",
       " GO:0048514_GOBP_BLOOD_VESSEL_MORPHOGENESIS                       Chd7,Jun,Junb,Klf4,Nr4a1,Srf,Tnfaip3   \n",
       " GO:0007626_GOBP_LOCOMOTORY_BEHAVIOR                                              Chd7,Egr1,Nr4a2,Usp2   \n",
       " GO:0051402_GOBP_NEURON_APOPTOTIC_PROCESS                                Btg2,Ddit3,Jun,Nr4a2,Ppargc1a   \n",
       " GO:0003682_GOMF_CHROMATIN_BINDING                     Cbx4,Chd7,Egr1,Fos,Fosl2,Klf4,Per1,Ppargc1a,Srf   \n",
       " GO:1990845_GOBP_ADAPTIVE_THERMOGENESIS                                     Ddit3,Per2,Ppargc1a,Prkab2   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0001216_GOMF_DNA_BINDING_TRANSCRIPTION_ACTIV...  0.088542          89  \n",
       " GO:0000987_GOMF_CIS_REGULATORY_REGION_SEQUENCE_...  0.048140          89  \n",
       " GO:0003700_GOMF_DNA_BINDING_TRANSCRIPTION_FACTO...  0.042596          89  \n",
       " R-HSA-9031628_REACTOME_NGF_STIMULATED_TRANSCRIP...  0.241379          89  \n",
       " R-HSA-198725_REACTOME_NUCLEAR_EVENTS_KINASE_AND...  0.152174          89  \n",
       " ...                                                      ...         ...  \n",
       " GO:0048514_GOBP_BLOOD_VESSEL_MORPHOGENESIS          0.029412          89  \n",
       " GO:0007626_GOBP_LOCOMOTORY_BEHAVIOR                 0.049383          89  \n",
       " GO:0051402_GOBP_NEURON_APOPTOTIC_PROCESS            0.039062          89  \n",
       " GO:0003682_GOMF_CHROMATIN_BINDING                   0.024590          89  \n",
       " GO:1990845_GOBP_ADAPTIVE_THERMOGENESIS              0.048780          89  \n",
       " \n",
       " [249 rows x 9 columns]}"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "03c06328",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'DE_diabetes.xlsx',engine='xlsxwriter') \n",
    "for name,res_sub1 in res.items():\n",
    "    res_sub1.to_excel(writer, sheet_name=name,index=True)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "c07b2947",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/DE_diabetes.xlsx'"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'DE_diabetes.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdad73f9",
   "metadata": {},
   "source": [
    "## DE diabetes in endocrine cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "fafd005a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load table parts\n",
    "res={}\n",
    "for de,res_sub in pd.read_excel(path_data+'combined/de/deR_endocrine/endo_summaries.xlsx',\n",
    "                                index_col=0,sheet_name=None).items():\n",
    "    res_sub.index.name='EID'\n",
    "    res[de+'_DE']=res_sub\n",
    "for direction,res_sub in pd.read_excel(\n",
    "    path_data+'combined/de/deR_endocrine/endoADG_shared_enrichment.xlsx', \n",
    "    index_col=0,sheet_name=None).items():\n",
    "    res_sub.index.name='gene_set'\n",
    "    res_sub.drop('label',axis=1,inplace=True)\n",
    "    res[f'sharedADG_DE_enrichment_{direction}']=res_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "05ce1f50",
   "metadata": {
    "lines_to_next_cell": 2,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alpha_T1D_DE':                        logFC    logCPM         LR        PValue          padj  \\\n",
       " EID                                                                             \n",
       " ENSMUSG00000051951 -0.846904  0.801437   5.186254  2.276623e-02  4.469554e-02   \n",
       " ENSMUSG00000033845 -0.059369  4.637682   0.105242  7.456274e-01  8.059962e-01   \n",
       " ENSMUSG00000025903  0.766796  3.963959  25.090379  5.470511e-07  3.351673e-06   \n",
       " ENSMUSG00000033813  0.159291  5.430743   0.711212  3.990418e-01  4.954594e-01   \n",
       " ENSMUSG00000002459 -3.919859  0.723604  57.199570  3.937572e-14  7.295015e-13   \n",
       " ...                      ...       ...        ...           ...           ...   \n",
       " ENSMUSG00000079834  1.041111 -0.750531   1.617811  2.033973e-01  2.847907e-01   \n",
       " ENSMUSG00000079808  0.187118 -0.121416   0.051531  8.204195e-01  8.650111e-01   \n",
       " ENSMUSG00000095041  0.821924  5.692306  13.487351  2.401770e-04  8.287143e-04   \n",
       " ENSMUSG00000063897  0.267417  3.542380   1.510481  2.190658e-01  3.025374e-01   \n",
       " ENSMUSG00000095742 -1.465943 -0.529621   6.686151  9.716475e-03  2.150143e-02   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                0.573295            Xkr4  \n",
       " ENSMUSG00000033845                0.433064          Mrpl15  \n",
       " ENSMUSG00000025903                0.524342          Lypla1  \n",
       " ENSMUSG00000033813                0.614443           Tcea1  \n",
       " ENSMUSG00000002459                0.494008           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.210997           Tmlhe  \n",
       " ENSMUSG00000079808                0.107895      AC168977.1  \n",
       " ENSMUSG00000095041                0.761393      AC149090.1  \n",
       " ENSMUSG00000063897                0.197829  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.277202  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'beta_T1D_DE':                        logFC    logCPM         LR        PValue          padj  \\\n",
       " EID                                                                             \n",
       " ENSMUSG00000051951  0.418180  0.801437   1.557583  2.120195e-01  3.578235e-01   \n",
       " ENSMUSG00000033845  0.571002  4.637682  18.365260  1.823526e-05  1.922792e-04   \n",
       " ENSMUSG00000025903  0.163237  3.963959   2.266459  1.322017e-01  2.529184e-01   \n",
       " ENSMUSG00000033813  0.325152  5.430743   5.964176  1.459942e-02  4.715499e-02   \n",
       " ENSMUSG00000002459  2.866928  0.723604  42.494799  7.087041e-11  3.095535e-09   \n",
       " ...                      ...       ...        ...           ...           ...   \n",
       " ENSMUSG00000079834  0.495714 -0.750531   0.690284  4.060677e-01  5.673103e-01   \n",
       " ENSMUSG00000079808  0.148918 -0.121416   0.116131  7.332693e-01  8.348057e-01   \n",
       " ENSMUSG00000095041 -0.317438  5.692306   3.216047  7.291953e-02  1.623977e-01   \n",
       " ENSMUSG00000063897 -0.704787  3.542380  17.637055  2.673286e-05  2.661383e-04   \n",
       " ENSMUSG00000095742  0.764871 -0.529621   1.906977  1.672995e-01  3.015231e-01   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                1.000000            Xkr4  \n",
       " ENSMUSG00000033845                0.577643          Mrpl15  \n",
       " ENSMUSG00000025903                0.755665          Lypla1  \n",
       " ENSMUSG00000033813                0.898695           Tcea1  \n",
       " ENSMUSG00000002459                0.194879           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.473132           Tmlhe  \n",
       " ENSMUSG00000079808                0.129705      AC168977.1  \n",
       " ENSMUSG00000095041                1.000000      AC149090.1  \n",
       " ENSMUSG00000063897                0.619154  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.250929  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'gamma_T1D_DE':                        logFC    logCPM         LR    PValue      padj  \\\n",
       " EID                                                                     \n",
       " ENSMUSG00000051951 -0.106380  0.801437   0.014431  0.904382  0.946854   \n",
       " ENSMUSG00000033845  0.090762  4.637682   0.148520  0.699954  0.807018   \n",
       " ENSMUSG00000025903  0.773543  3.963959  11.566791  0.000671  0.004086   \n",
       " ENSMUSG00000033813  0.576634  5.430743   6.369515  0.011610  0.040773   \n",
       " ENSMUSG00000002459  1.883934  0.723604   2.005637  0.156715  0.291653   \n",
       " ...                      ...       ...        ...       ...       ...   \n",
       " ENSMUSG00000079834 -2.404896 -0.750531   4.660558  0.030863  0.087233   \n",
       " ENSMUSG00000079808 -0.005022 -0.121416   0.000309  0.985972  0.992745   \n",
       " ENSMUSG00000095041  0.238765  5.692306   0.867837  0.351555  0.513949   \n",
       " ENSMUSG00000063897  0.228104  3.542380   0.451470  0.501638  0.654085   \n",
       " ENSMUSG00000095742 -2.458626 -0.529621   2.228831  0.135456  0.263381   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                0.308394            Xkr4  \n",
       " ENSMUSG00000033845                0.383709          Mrpl15  \n",
       " ENSMUSG00000025903                0.526982          Lypla1  \n",
       " ENSMUSG00000033813                0.494142           Tcea1  \n",
       " ENSMUSG00000002459                0.142759           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.313572           Tmlhe  \n",
       " ENSMUSG00000079808                0.035011      AC168977.1  \n",
       " ENSMUSG00000095041                0.694378      AC149090.1  \n",
       " ENSMUSG00000063897                0.174842  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.130164  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'delta_T1D_DE':                        logFC    logCPM         LR        PValue          padj  \\\n",
       " EID                                                                             \n",
       " ENSMUSG00000051951 -0.395201  0.801437   0.919656  3.375654e-01  4.578181e-01   \n",
       " ENSMUSG00000033845  0.317393  4.637682   2.750785  9.720668e-02  1.715389e-01   \n",
       " ENSMUSG00000025903  0.779738  3.963959  21.986085  2.746344e-06  1.781949e-05   \n",
       " ENSMUSG00000033813  0.563355  5.430743   8.442920  3.664673e-03  1.078552e-02   \n",
       " ENSMUSG00000002459 -4.082015  0.723604  83.358998  6.842662e-20  2.446113e-18   \n",
       " ...                      ...       ...        ...           ...           ...   \n",
       " ENSMUSG00000079834 -1.338799 -0.750531   4.293611  3.825583e-02  7.923263e-02   \n",
       " ENSMUSG00000079808 -1.017652 -0.121416   2.230185  1.353377e-01  2.241636e-01   \n",
       " ENSMUSG00000095041  0.191674  5.692306   0.719054  3.964545e-01  5.170175e-01   \n",
       " ENSMUSG00000063897  0.501833  3.542380   5.052830  2.458575e-02  5.478318e-02   \n",
       " ENSMUSG00000095742  0.514639 -0.529621   0.637427  4.246440e-01  5.453684e-01   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                0.410918            Xkr4  \n",
       " ENSMUSG00000033845                0.405627          Mrpl15  \n",
       " ENSMUSG00000025903                0.540284          Lypla1  \n",
       " ENSMUSG00000033813                0.685363           Tcea1  \n",
       " ENSMUSG00000002459                0.995348           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.387366           Tmlhe  \n",
       " ENSMUSG00000079808                0.201724      AC168977.1  \n",
       " ENSMUSG00000095041                0.854742      AC149090.1  \n",
       " ENSMUSG00000063897                0.160074  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.141650  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'alpha_T2D_DE':                        logFC    logCPM         LR        PValue      padj  \\\n",
       " EID                                                                         \n",
       " ENSMUSG00000051951 -0.897819  0.801437   5.275809  2.162373e-02  0.054727   \n",
       " ENSMUSG00000033845 -0.292148  4.637682   2.060049  1.512053e-01  0.255646   \n",
       " ENSMUSG00000025903  0.846788  3.963959  25.717971  3.951305e-07  0.000004   \n",
       " ENSMUSG00000033813  0.022719  5.430743   0.012007  9.127464e-01  0.944169   \n",
       " ENSMUSG00000002459 -2.865146  0.723604  26.183863  3.104056e-07  0.000004   \n",
       " ...                      ...       ...        ...           ...       ...   \n",
       " ENSMUSG00000079834  0.263212 -0.750531   0.081309  7.755317e-01  0.848473   \n",
       " ENSMUSG00000079808 -1.008518 -0.121416   1.048429  3.058692e-01  0.435480   \n",
       " ENSMUSG00000095041  0.181311  5.692306   0.526075  4.682617e-01  0.595319   \n",
       " ENSMUSG00000063897  0.495750  3.542380   5.238864  2.208769e-02  0.055624   \n",
       " ENSMUSG00000095742 -1.139115 -0.529621   4.841491  2.778287e-02  0.066954   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                0.573295            Xkr4  \n",
       " ENSMUSG00000033845                0.433064          Mrpl15  \n",
       " ENSMUSG00000025903                0.524342          Lypla1  \n",
       " ENSMUSG00000033813                0.614443           Tcea1  \n",
       " ENSMUSG00000002459                0.494008           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.210997           Tmlhe  \n",
       " ENSMUSG00000079808                0.107895      AC168977.1  \n",
       " ENSMUSG00000095041                0.761393      AC149090.1  \n",
       " ENSMUSG00000063897                0.197829  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.277202  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'beta_T2D_DE':                        logFC    logCPM         LR    PValue      padj  \\\n",
       " EID                                                                     \n",
       " ENSMUSG00000051951  0.753116  0.801437   8.706365  0.003171  0.009803   \n",
       " ENSMUSG00000033845  0.501750  4.637682  11.987672  0.000536  0.002118   \n",
       " ENSMUSG00000025903 -0.078750  3.963959   0.498426  0.480193  0.594428   \n",
       " ENSMUSG00000033813 -0.056205  5.430743   0.127532  0.721005  0.801245   \n",
       " ENSMUSG00000002459  0.356405  0.723604   0.380345  0.537419  0.646877   \n",
       " ...                      ...       ...        ...       ...       ...   \n",
       " ENSMUSG00000079834  0.614349 -0.750531   1.834575  0.175588  0.274332   \n",
       " ENSMUSG00000079808  1.832676 -0.121416  11.584936  0.000665  0.002553   \n",
       " ENSMUSG00000095041 -0.077575  5.692306   0.163021  0.686390  0.774056   \n",
       " ENSMUSG00000063897  0.686214  3.542380  22.752868  0.000002  0.000014   \n",
       " ENSMUSG00000095742  0.511711 -0.529621   1.926934  0.165094  0.261279   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                1.000000            Xkr4  \n",
       " ENSMUSG00000033845                0.577643          Mrpl15  \n",
       " ENSMUSG00000025903                0.755665          Lypla1  \n",
       " ENSMUSG00000033813                0.898695           Tcea1  \n",
       " ENSMUSG00000002459                0.194879           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.473132           Tmlhe  \n",
       " ENSMUSG00000079808                0.129705      AC168977.1  \n",
       " ENSMUSG00000095041                1.000000      AC149090.1  \n",
       " ENSMUSG00000063897                0.619154  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.250929  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'gamma_T2D_DE':                        logFC    logCPM        LR    PValue      padj  \\\n",
       " EID                                                                    \n",
       " ENSMUSG00000051951 -0.110950  0.801437  0.015147  0.902049  0.955047   \n",
       " ENSMUSG00000033845  0.035150  4.637682  0.018683  0.891280  0.949513   \n",
       " ENSMUSG00000025903  0.425726  3.963959  2.792515  0.094705  0.257898   \n",
       " ENSMUSG00000033813  0.559294  5.430743  5.129125  0.023528  0.099818   \n",
       " ENSMUSG00000002459  1.876283  0.723604  1.827031  0.176479  0.380410   \n",
       " ...                      ...       ...       ...       ...       ...   \n",
       " ENSMUSG00000079834 -1.671952 -0.750531  2.382583  0.122695  0.302178   \n",
       " ENSMUSG00000079808 -0.731933 -0.121416  0.153278  0.695422  0.836533   \n",
       " ENSMUSG00000095041 -0.178282  5.692306  0.403137  0.525474  0.722862   \n",
       " ENSMUSG00000063897 -0.072887  3.542380  0.050812  0.821656  0.913309   \n",
       " ENSMUSG00000095742 -0.133664 -0.529621  0.013682  0.906883  0.957028   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                0.308394            Xkr4  \n",
       " ENSMUSG00000033845                0.383709          Mrpl15  \n",
       " ENSMUSG00000025903                0.526982          Lypla1  \n",
       " ENSMUSG00000033813                0.494142           Tcea1  \n",
       " ENSMUSG00000002459                0.142759           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.313572           Tmlhe  \n",
       " ENSMUSG00000079808                0.035011      AC168977.1  \n",
       " ENSMUSG00000095041                0.694378      AC149090.1  \n",
       " ENSMUSG00000063897                0.174842  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.130164  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'delta_T2D_DE':                        logFC    logCPM         LR        PValue          padj  \\\n",
       " EID                                                                             \n",
       " ENSMUSG00000051951 -1.100339  0.801437   5.314001  2.115472e-02  6.553590e-02   \n",
       " ENSMUSG00000033845  0.119642  4.637682   0.308614  5.785318e-01  7.228841e-01   \n",
       " ENSMUSG00000025903  0.836678  3.963959  18.698326  1.531168e-05  1.565708e-04   \n",
       " ENSMUSG00000033813  0.322500  5.430743   2.286942  1.304667e-01  2.602015e-01   \n",
       " ENSMUSG00000002459 -3.564656  0.723604  42.206821  8.211314e-11  2.507302e-09   \n",
       " ...                      ...       ...        ...           ...           ...   \n",
       " ENSMUSG00000079834 -2.225505 -0.750531   7.981915  4.724693e-03  1.997546e-02   \n",
       " ENSMUSG00000079808  0.099084 -0.121416   0.021733  8.828001e-01  9.321979e-01   \n",
       " ENSMUSG00000095041 -0.212391  5.692306   0.714927  3.978131e-01  5.674256e-01   \n",
       " ENSMUSG00000063897  0.829244  3.542380  13.023845  3.075497e-04  2.058289e-03   \n",
       " ENSMUSG00000095742 -1.046099 -0.529621   1.917639  1.661171e-01  3.093390e-01   \n",
       " \n",
       "                     relative_ct_expression     gene_symbol  \n",
       " EID                                                         \n",
       " ENSMUSG00000051951                0.410918            Xkr4  \n",
       " ENSMUSG00000033845                0.405627          Mrpl15  \n",
       " ENSMUSG00000025903                0.540284          Lypla1  \n",
       " ENSMUSG00000033813                0.685363           Tcea1  \n",
       " ENSMUSG00000002459                0.995348           Rgs20  \n",
       " ...                                    ...             ...  \n",
       " ENSMUSG00000079834                0.387366           Tmlhe  \n",
       " ENSMUSG00000079808                0.201724      AC168977.1  \n",
       " ENSMUSG00000095041                0.854742      AC149090.1  \n",
       " ENSMUSG00000063897                0.160074  CAAA01118383.1  \n",
       " ENSMUSG00000095742                0.141650  CAAA01147332.1  \n",
       " \n",
       " [13191 rows x 7 columns],\n",
       " 'sharedADG_DE_enrichment_up':                                                         pval    fdr  \\\n",
       " gene_set                                                              \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  0.000005  0.052   \n",
       " R-HSA-1912422_REACTOME_PRE_NOTCH_EXPRESSION_AND...  0.000019  0.094   \n",
       " R-HSA-1912420_REACTOME_PRE_NOTCH_PROCESSING_IN_...  0.000048  0.160   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...         64      267   \n",
       " R-HSA-1912422_REACTOME_PRE_NOTCH_EXPRESSION_AND...         64       68   \n",
       " R-HSA-1912420_REACTOME_PRE_NOTCH_PROCESSING_IN_...         64       15   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...        9       13191   \n",
       " R-HSA-1912422_REACTOME_PRE_NOTCH_EXPRESSION_AND...        5       13191   \n",
       " R-HSA-1912420_REACTOME_PRE_NOTCH_PROCESSING_IN_...        3       13191   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  Aqp11,Atp2a2,Atp2a3,Hyou1,Sdf2l1,Sel1l,Sesn2,T...   \n",
       " R-HSA-1912422_REACTOME_PRE_NOTCH_EXPRESSION_AND...                     Atp2a2,Atp2a3,H2bc6,H4c9,Sel1l   \n",
       " R-HSA-1912420_REACTOME_PRE_NOTCH_PROCESSING_IN_...                                Atp2a2,Atp2a3,Sel1l   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:0034976_GOBP_RESPONSE_TO_ENDOPLASMIC_RETICUL...  0.033708          64  \n",
       " R-HSA-1912422_REACTOME_PRE_NOTCH_EXPRESSION_AND...  0.073529          64  \n",
       " R-HSA-1912420_REACTOME_PRE_NOTCH_PROCESSING_IN_...  0.200000          64  ,\n",
       " 'sharedADG_DE_enrichment_down':                                                             pval       fdr  \\\n",
       " gene_set                                                                     \n",
       " GO:1990351_GOCC_TRANSPORTER_COMPLEX                 6.300000e-09  0.000053   \n",
       " R-HSA-397014_REACTOME_MUSCLE_CONTRACTION            1.100000e-08  0.000053   \n",
       " GO:0034703_GOCC_CATION_CHANNEL_COMPLEX              2.900000e-08  0.000098   \n",
       " GO:0061337_GOBP_CARDIAC_CONDUCTION                  9.600000e-08  0.000240   \n",
       " GO:1903522_GOBP_REGULATION_OF_BLOOD_CIRCULATION     1.200000e-07  0.000240   \n",
       " ...                                                          ...       ...   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY          1.900000e-03  0.240000   \n",
       " GO:0015961_GOBP_DIADENOSINE_POLYPHOSPHATE_CATAB...  2.000000e-03  0.240000   \n",
       " GO:0035502_GOBP_METANEPHRIC_PART_OF_URETERIC_BU...  2.000000e-03  0.240000   \n",
       " GO:0050912_GOBP_DETECTION_OF_CHEMICAL_STIMULUS_...  2.000000e-03  0.240000   \n",
       " GO:0099624_GOBP_ATRIAL_CARDIAC_MUSCLE_CELL_MEMB...  2.000000e-03  0.240000   \n",
       " \n",
       "                                                     signature  geneset  \\\n",
       " gene_set                                                                 \n",
       " GO:1990351_GOCC_TRANSPORTER_COMPLEX                       189      180   \n",
       " R-HSA-397014_REACTOME_MUSCLE_CONTRACTION                  189      116   \n",
       " GO:0034703_GOCC_CATION_CHANNEL_COMPLEX                    189      126   \n",
       " GO:0061337_GOBP_CARDIAC_CONDUCTION                        189       94   \n",
       " GO:1903522_GOBP_REGULATION_OF_BLOOD_CIRCULATION           189      167   \n",
       " ...                                                       ...      ...   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY                189       62   \n",
       " GO:0015961_GOBP_DIADENOSINE_POLYPHOSPHATE_CATAB...        189        5   \n",
       " GO:0035502_GOBP_METANEPHRIC_PART_OF_URETERIC_BU...        189        5   \n",
       " GO:0050912_GOBP_DETECTION_OF_CHEMICAL_STIMULUS_...        189        5   \n",
       " GO:0099624_GOBP_ATRIAL_CARDIAC_MUSCLE_CELL_MEMB...        189        5   \n",
       " \n",
       "                                                     overlap  background  \\\n",
       " gene_set                                                                  \n",
       " GO:1990351_GOCC_TRANSPORTER_COMPLEX                      16       13191   \n",
       " R-HSA-397014_REACTOME_MUSCLE_CONTRACTION                 13       13191   \n",
       " GO:0034703_GOCC_CATION_CHANNEL_COMPLEX                   13       13191   \n",
       " GO:0061337_GOBP_CARDIAC_CONDUCTION                       11       13191   \n",
       " GO:1903522_GOBP_REGULATION_OF_BLOOD_CIRCULATION          14       13191   \n",
       " ...                                                     ...         ...   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY                5       13191   \n",
       " GO:0015961_GOBP_DIADENOSINE_POLYPHOSPHATE_CATAB...        2       13191   \n",
       " GO:0035502_GOBP_METANEPHRIC_PART_OF_URETERIC_BU...        2       13191   \n",
       " GO:0050912_GOBP_DETECTION_OF_CHEMICAL_STIMULUS_...        2       13191   \n",
       " GO:0099624_GOBP_ATRIAL_CARDIAC_MUSCLE_CELL_MEMB...        2       13191   \n",
       " \n",
       "                                                                                                  hits  \\\n",
       " gene_set                                                                                                \n",
       " GO:1990351_GOCC_TRANSPORTER_COMPLEX                 Cacna2d3,Cldn4,Gabrg1,Grik4,Kcna5,Kcnip3,Kcnj3...   \n",
       " R-HSA-397014_REACTOME_MUSCLE_CONTRACTION            Corin,Dmd,Fgf14,Itpr1,Kcnip3,Kcnk3,Myl12a,Myl1...   \n",
       " GO:0034703_GOCC_CATION_CHANNEL_COMPLEX              Cacna2d3,Grik4,Kcna5,Kcnip3,Kcnj3,Mcub,Nlgn1,O...   \n",
       " GO:0061337_GOBP_CARDIAC_CONDUCTION                  Cacna2d3,Corin,Itpr1,Kcna5,Kcnip3,Kcnj3,Kcnk3,...   \n",
       " GO:1903522_GOBP_REGULATION_OF_BLOOD_CIRCULATION     Cacna2d3,Corin,Dmd,Gstm7,Itpr1,Kcna5,Kcnip3,Kc...   \n",
       " ...                                                                                               ...   \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY                             Grik4,Kcna5,Kcnip3,Kcnj3,Kcnk3   \n",
       " GO:0015961_GOBP_DIADENOSINE_POLYPHOSPHATE_CATAB...                                      Nudt10,Nudt11   \n",
       " GO:0035502_GOBP_METANEPHRIC_PART_OF_URETERIC_BU...                                        Basp1,Calb1   \n",
       " GO:0050912_GOBP_DETECTION_OF_CHEMICAL_STIMULUS_...                                         Cst3,Ffar4   \n",
       " GO:0099624_GOBP_ATRIAL_CARDIAC_MUSCLE_CELL_MEMB...                                        Kcna5,Kcnj3   \n",
       " \n",
       "                                                       recall  query_size  \n",
       " gene_set                                                                  \n",
       " GO:1990351_GOCC_TRANSPORTER_COMPLEX                 0.088889         189  \n",
       " R-HSA-397014_REACTOME_MUSCLE_CONTRACTION            0.112069         189  \n",
       " GO:0034703_GOCC_CATION_CHANNEL_COMPLEX              0.103175         189  \n",
       " GO:0061337_GOBP_CARDIAC_CONDUCTION                  0.117021         189  \n",
       " GO:1903522_GOBP_REGULATION_OF_BLOOD_CIRCULATION     0.083832         189  \n",
       " ...                                                      ...         ...  \n",
       " GO:0005267_GOMF_POTASSIUM_CHANNEL_ACTIVITY          0.080645         189  \n",
       " GO:0015961_GOBP_DIADENOSINE_POLYPHOSPHATE_CATAB...  0.400000         189  \n",
       " GO:0035502_GOBP_METANEPHRIC_PART_OF_URETERIC_BU...  0.400000         189  \n",
       " GO:0050912_GOBP_DETECTION_OF_CHEMICAL_STIMULUS_...  0.400000         189  \n",
       " GO:0099624_GOBP_ATRIAL_CARDIAC_MUSCLE_CELL_MEMB...  0.400000         189  \n",
       " \n",
       " [83 rows x 9 columns]}"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "cd620c14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'DE_diabetes_endo.xlsx',engine='xlsxwriter') \n",
    "for name,res_sub1 in res.items():\n",
    "    res_sub1.to_excel(writer, sheet_name=name,index=True)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "c166b379",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/DE_diabetes_endo.xlsx'"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'DE_diabetes_endo.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13eb7a21",
   "metadata": {},
   "source": [
    "## DE sex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "244b28c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load table parts\n",
    "res={}\n",
    "res['P16']=pd.read_table(path_data+'combined/de/de_sexaging_covarSample/maleFemale_Fltp_P16_0.05_summary_sdFiltered.tsv',\n",
    "                         index_col=0)\n",
    "res['aged']=pd.read_table(path_data+'combined/de/de_sexaging_covarSample/deDataClusters.tsv',\n",
    "                         index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "624fb150",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'P16':                             pval      qval    log2fc       mean  zero_mean  \\\n",
       " gene                                                                         \n",
       " ENSMUSG00000033845  2.495902e-01  0.667174  0.035670   0.589588      False   \n",
       " ENSMUSG00000025903  7.400376e-02  0.404821 -0.075601   0.319405      False   \n",
       " ENSMUSG00000033813  4.421132e-01  0.791722 -0.020360   0.868154      False   \n",
       " ENSMUSG00000033793  4.050177e-02  0.312505  0.062705   0.643813      False   \n",
       " ENSMUSG00000025907  6.540125e-01  0.889950 -0.016788   0.398377      False   \n",
       " ...                          ...       ...       ...        ...        ...   \n",
       " ENSMUSG00000064367  3.444915e-01  0.737965  0.014160   4.082488      False   \n",
       " ENSMUSG00000064370  2.497162e-02  0.248728  0.019781  56.947397      False   \n",
       " ENSMUSG00000051412  4.352518e-08  0.000009  0.176885   0.643813      False   \n",
       " ENSMUSG00000095041  5.635760e-02  0.362153  0.076939   0.381339      False   \n",
       " ENSMUSG00000063897  7.408077e-01  0.919520  0.012832   0.413117      False   \n",
       " \n",
       "                             grad  coef_mle   coef_sd            ll  \n",
       " gene                                                                \n",
       " ENSMUSG00000033845  9.550715e-07  0.035670  0.030981  -7374.614009  \n",
       " ENSMUSG00000025903  3.088802e-09 -0.075601  0.042316  -5271.874573  \n",
       " ENSMUSG00000033813  9.242517e-09 -0.020360  0.026488  -9034.756703  \n",
       " ENSMUSG00000033793  4.186349e-09  0.062705  0.030609  -7802.509202  \n",
       " ENSMUSG00000025907  3.065263e-02 -0.016788  0.037457   -886.745994  \n",
       " ...                          ...       ...       ...           ...  \n",
       " ENSMUSG00000064367  5.866631e-07  0.014160  0.014979 -16781.349153  \n",
       " ENSMUSG00000064370  3.629503e-08  0.019781  0.008823 -32588.158870  \n",
       " ENSMUSG00000051412  1.584055e-07  0.176885  0.032302  -7864.709494  \n",
       " ENSMUSG00000095041  4.484279e-08  0.076939  0.040319  -5908.099906  \n",
       " ENSMUSG00000063897  2.002722e-08  0.012832  0.038793  -6173.522018  \n",
       " \n",
       " [9312 rows x 9 columns],\n",
       " 'aged':                             pval          qval    log2fc        mean  \\\n",
       " gene                                                                   \n",
       " ENSMUSG00000051951  0.000000e+00  0.000000e+00  0.729754    0.096120   \n",
       " ENSMUSG00000033845  8.174452e-07  1.503167e-06  0.141079    0.452256   \n",
       " ENSMUSG00000025903  2.833213e-08  5.743214e-08  0.150064    0.546622   \n",
       " ENSMUSG00000033813  0.000000e+00  0.000000e+00  0.159760    1.343226   \n",
       " ENSMUSG00000033793  1.288747e-10  3.040068e-10  0.129402    1.082158   \n",
       " ...                          ...           ...       ...         ...   \n",
       " ENSMUSG00000064368  0.000000e+00  0.000000e+00  0.283973    0.562759   \n",
       " ENSMUSG00000064370  3.596830e-04  5.384486e-04  0.032460  190.210342   \n",
       " ENSMUSG00000051412  0.000000e+00  0.000000e+00  0.634761    1.314179   \n",
       " ENSMUSG00000095041  1.332268e-15  4.086019e-15  0.137913    1.797236   \n",
       " ENSMUSG00000063897  0.000000e+00  0.000000e+00  0.437492    0.643584   \n",
       " \n",
       "                     zero_mean          grad  coef_mle   coef_sd            ll  \\\n",
       " gene                                                                            \n",
       " ENSMUSG00000051951      False  7.940045e-08  0.729754  0.063924  -4714.605539   \n",
       " ENSMUSG00000033845      False  1.538251e-06  0.141079  0.028610 -12475.058608   \n",
       " ENSMUSG00000025903      False  1.498951e-06  0.150064  0.027032 -14285.212961   \n",
       " ENSMUSG00000033813      False  1.621427e-06  0.159760  0.018504 -22047.210763   \n",
       " ENSMUSG00000033793      False  1.753754e-06  0.129402  0.020129 -20026.859932   \n",
       " ...                       ...           ...       ...       ...           ...   \n",
       " ENSMUSG00000064368      False  7.021447e-08  0.283973  0.033230 -14590.166850   \n",
       " ENSMUSG00000064370      False  1.024241e-06  0.032460  0.009097 -84573.885796   \n",
       " ENSMUSG00000051412      False  2.930141e-07  0.634761  0.019659 -21393.940985   \n",
       " ENSMUSG00000095041      False  7.513909e-07  0.137913  0.017252 -25295.638814   \n",
       " ENSMUSG00000063897      False  7.889379e-07  0.437492  0.026059 -15469.932733   \n",
       " \n",
       "                      hc     gene_symbol  rel_beta_expr  \\\n",
       " gene                                                     \n",
       " ENSMUSG00000051951  NaN            Xkr4       1.000000   \n",
       " ENSMUSG00000033845  NaN          Mrpl15       0.577643   \n",
       " ENSMUSG00000025903  NaN          Lypla1       0.755665   \n",
       " ENSMUSG00000033813  NaN           Tcea1       0.898695   \n",
       " ENSMUSG00000033793  NaN         Atp6v1h       1.000000   \n",
       " ...                 ...             ...            ...   \n",
       " ENSMUSG00000064368  NaN          mt-Nd6       1.000000   \n",
       " ENSMUSG00000064370  NaN         mt-Cytb       1.000000   \n",
       " ENSMUSG00000051412  NaN           Vamp7       1.000000   \n",
       " ENSMUSG00000095041  NaN      AC149090.1       1.000000   \n",
       " ENSMUSG00000063897  NaN  CAAA01118383.1       0.619154   \n",
       " \n",
       "                     mean_expr_in_expr_cells  N_PMID  \\\n",
       " gene                                                  \n",
       " ENSMUSG00000051951                 0.487778    11.0   \n",
       " ENSMUSG00000033845                 0.665728    55.0   \n",
       " ENSMUSG00000025903                 0.617923    42.0   \n",
       " ENSMUSG00000033813                 0.794732    66.0   \n",
       " ENSMUSG00000033793                 0.706451    65.0   \n",
       " ...                                     ...     ...   \n",
       " ENSMUSG00000064368                 0.684435    42.0   \n",
       " ENSMUSG00000064370                 4.252210    39.0   \n",
       " ENSMUSG00000051412                 0.746626     0.0   \n",
       " ENSMUSG00000095041                 0.812761     0.0   \n",
       " ENSMUSG00000063897                 0.673011     0.0   \n",
       " \n",
       "                     N_PMID_pancreas_notCancerNonendo  \n",
       " gene                                                  \n",
       " ENSMUSG00000051951                               0.0  \n",
       " ENSMUSG00000033845                               0.0  \n",
       " ENSMUSG00000025903                               0.0  \n",
       " ENSMUSG00000033813                               0.0  \n",
       " ENSMUSG00000033793                               1.0  \n",
       " ...                                              ...  \n",
       " ENSMUSG00000064368                               4.0  \n",
       " ENSMUSG00000064370                               5.0  \n",
       " ENSMUSG00000051412                               0.0  \n",
       " ENSMUSG00000095041                               0.0  \n",
       " ENSMUSG00000063897                               0.0  \n",
       " \n",
       " [10258 rows x 15 columns]}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a120ae3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save single excel file\n",
    "writer = pd.ExcelWriter(path_tables+'DE_sex_ages.xlsx',engine='xlsxwriter') \n",
    "for name,res_sub1 in res.items():\n",
    "    res_sub1.to_excel(writer, sheet_name=name,index=True)   \n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "49d7970b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/lustre/groups/ml01/workspace/karin.hrovatin/data/pancreas/tables/paper/DE_sex_ages.xlsx'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_tables+'DE_sex_ages.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efd4dd3f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py:percent"
  },
  "kernelspec": {
   "display_name": "Python rpy2_3",
   "language": "python",
   "name": "rpy2_3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
